<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta name="generator" content="bookdown 0.16 and GitBook 2.6.7" />

  <meta property="og:title" content="4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/" />
  <meta property="og:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />
  
  <meta name="github-repo" content="broadinstitute/KrumlovSingleCellWorkship2020" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  
  <meta name="twitter:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />

<meta name="author" content="Kirk Gosik" />


<meta name="date" content="2020-01-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="expression-qc-and-normalization-1.html"/>
<link rel="next" href="identifying-cell-populations-1.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#introduction-1"><i class="fa fa-check"></i><b>1.5.1</b> Introduction</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#transcriptome-quantification"><i class="fa fa-check"></i><b>1.5.2</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#expression-qc-and-normalization"><i class="fa fa-check"></i><b>1.5.3</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#data-wrangling-scrnaseq"><i class="fa fa-check"></i><b>1.5.4</b> Data Wrangling scRNAseq</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#identifying-cell-populations"><i class="fa fa-check"></i><b>1.5.5</b> Identifying Cell Populations</a></li>
<li class="chapter" data-level="1.5.6" data-path="index.html"><a href="index.html#feature-selection-and-cluster-analysis"><i class="fa fa-check"></i><b>1.5.6</b> Feature Selection and Cluster Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html"><i class="fa fa-check"></i><b>2</b> Transcriptome Quantification</a><ul>
<li class="chapter" data-level="2.1" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html#google-slides"><i class="fa fa-check"></i><b>2.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html"><i class="fa fa-check"></i><b>3</b> Expression QC and Normalization</a><ul>
<li class="chapter" data-level="3.1" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html#google-slides-1"><i class="fa fa-check"></i><b>3.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html"><i class="fa fa-check"></i><b>4</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#goal"><i class="fa fa-check"></i><b>4.1</b> Goal</a></li>
<li class="chapter" data-level="4.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#introduction-2"><i class="fa fa-check"></i><b>4.2</b> Introduction</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#load-necessary-packages"><i class="fa fa-check"></i><b>4.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>4.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>4.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>4.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>4.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>4.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>4.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>4.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#beginning-with-seurat"><i class="fa fa-check"></i><b>4.4</b> Beginning with Seurat:</a><ul>
<li class="chapter" data-level="4.4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>4.4.1</b> Creating a seurat object</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>4.5</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="4.6" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>4.6</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="4.6.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>4.6.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>4.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="4.8" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>4.8</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html"><i class="fa fa-check"></i><b>5</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="5.1" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html#google-slides-2"><i class="fa fa-check"></i><b>5.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html"><i class="fa fa-check"></i><b>6</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#abstract"><i class="fa fa-check"></i><b>6.1</b> Abstract</a></li>
<li class="chapter" data-level="6.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#seurat-tutorial-redo"><i class="fa fa-check"></i><b>6.2</b> Seurat Tutorial Redo</a><ul>
<li class="chapter" data-level="6.2.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#preprocessing-steps"><i class="fa fa-check"></i><b>6.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="6.2.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>6.2.2</b> Start of Identifying Cell Types</a></li>
<li class="chapter" data-level="6.2.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#run-non-linear-dimensional-reduction-umaptsne"><i class="fa fa-check"></i><b>6.2.3</b> Run non-linear dimensional reduction (UMAP/tSNE)</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#feature-selection"><i class="fa fa-check"></i><b>6.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="6.3.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="6.3.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#check-clusters"><i class="fa fa-check"></i><b>6.3.2</b> Check Clusters</a></li>
<li class="chapter" data-level="6.3.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#view-entire-object-structure"><i class="fa fa-check"></i><b>6.3.3</b> View Entire Object Structure</a></li>
<li class="chapter" data-level="6.3.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>6.3.4</b> Practice Visualizing/Embedding</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#other-options-for-analysis"><i class="fa fa-check"></i><b>6.4</b> Other Options For Analysis</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-wrangling-scrnaseq-1" class="section level1">
<h1><span class="header-section-number">4</span> Data Wrangling scRNAseq</h1>
<div id="goal" class="section level2">
<h2><span class="header-section-number">4.1</span> Goal</h2>
<ul>
<li>To give you experience with the analysis of single cell RNA sequencing (scRNA-seq) including performing quality control and identifying cell type subsets.</li>
<li>To introduce you to scRNA-seq analysis using the Seurat package.</li>
</ul>
</div>
<div id="introduction-2" class="section level2">
<h2><span class="header-section-number">4.2</span> Introduction</h2>
<p>Data produced in a single cell RNA-seq experiment has several interesting characteristics that make it distinct from data produced in a bulk population RNA-seq experiment. Two characteristics that are important to keep in mind when working with scRNA-Seq are drop-out (the excessive amount of zeros due to limiting mRNA) and the potential for quality control (QC) metrics to be confounded with biology. This combined with the ability to measure heterogeniety from cells in samples has shifted the field away from the typical analysis in population-based RNA-Seq. Here we demonstrate some approaches to quality control, followed by identifying and analyzing cell subsets.</p>
<p>For this tutorial, we will be analyzing the a dataset of Non-Small Cell Lung Cancer Cells (NSCLC) freely available from 10X Genomics (<a href="https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex" class="uri">https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex</a>), using the Seurat R package (<a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a>), a popular and powerful set of tools to conduct scRNA-seq analysis in R. In this dataset, there are 7802 single cells that were sequenced on the Illumina NovaSeq 6000. Please note this tutorial borrows heavily from Seurat’s tutorials, so feel free to go through them in more detail.</p>
<div id="load-necessary-packages" class="section level3">
<h3><span class="header-section-number">4.2.1</span> Load necessary packages</h3>
<p>When loading libraries, we are asking R to load code for us written by someone else. It is a convenient way to leverage and reproduce methodology developed by others.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="data-wrangling-scrnaseq-1.html#cb1-1"></a>knitr<span class="op">::</span>opts_knit<span class="op">$</span><span class="kw">set</span>(<span class="dt">progress=</span><span class="ot">FALSE</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</span>
<span id="cb1-2"><a href="data-wrangling-scrnaseq-1.html#cb1-2"></a></span>
<span id="cb1-3"><a href="data-wrangling-scrnaseq-1.html#cb1-3"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb1-4"><a href="data-wrangling-scrnaseq-1.html#cb1-4"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb1-5"><a href="data-wrangling-scrnaseq-1.html#cb1-5"></a><span class="kw">library</span>(Matrix)</span></code></pre></div>
</div>
<div id="read-in-nsclc-counts-matrix." class="section level3">
<h3><span class="header-section-number">4.2.2</span> Read in NSCLC counts matrix.</h3>
<p>The data for Non-Small Cell Lung Cancer Cells (NSCLC) is freely available from 10X Genomics (<a href="https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex" class="uri">https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex</a>). We start by reading in the counts matrix generated by the Cell Ranger count program.</p>
<p>Task: Change the directory name to mydir/ where you saved your data</p>
<p><strong>Task: Check the <code>dirname</code> to dir where you saved your data</strong></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="data-wrangling-scrnaseq-1.html#cb2-1"></a><span class="co">## dirname &lt;- &quot;/home/genomics/workshop_materials/scrna/&quot;</span></span>
<span id="cb2-2"><a href="data-wrangling-scrnaseq-1.html#cb2-2"></a>dirname &lt;-<span class="st"> &quot;/Users/kgosik/Documents/data/cellranger/vdj-nsclc/&quot;</span></span>
<span id="cb2-3"><a href="data-wrangling-scrnaseq-1.html#cb2-3"></a>counts_matrix_filename =<span class="st"> </span><span class="kw">paste0</span>(dirname, <span class="st">&quot;filtered_gene_bc_matrices/GRCh38/&quot;</span>)</span>
<span id="cb2-4"><a href="data-wrangling-scrnaseq-1.html#cb2-4"></a>counts &lt;-<span class="st"> </span><span class="kw">Read10X</span>(<span class="dt">data.dir =</span> counts_matrix_filename)  <span class="co"># Seurat function to read in 10x count data</span></span></code></pre></div>
</div>
<div id="lets-examine-the-sparse-counts-matrix" class="section level3">
<h3><span class="header-section-number">4.2.3</span> Let’s examine the sparse counts matrix</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="data-wrangling-scrnaseq-1.html#cb3-1"></a>counts[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## 10 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
##               AAACCTGAGCTAGTCT AAACCTGAGGGCACTA AAACCTGAGTACGTTC
## RP11-34P13.3                 .                .                .
## FAM138A                      .                .                .
## OR4F5                        .                .                .
## RP11-34P13.7                 .                .                .
## RP11-34P13.8                 .                .                .
## RP11-34P13.14                .                .                .
## RP11-34P13.9                 .                .                .
## FO538757.3                   .                .                .
## FO538757.2                   .                .                .
## AP006222.2                   .                .                .</code></pre>
<p>Here we see the upper left corner of the sparse matrix. The columns are indexed by 10x cell barcodes (each 16 nt long), and the rows are the gene names. We mentioned these matrices are sparse, here we see only zeroes (indicated by the “.” symbol); this is the most common value in these sparse matrices. Next, let us look at the dimensions of this matrix.</p>
</div>
<div id="how-big-is-the-matrix" class="section level3">
<h3><span class="header-section-number">4.2.4</span> How big is the matrix?</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="data-wrangling-scrnaseq-1.html#cb5-1"></a><span class="kw">dim</span>(counts) <span class="co"># report number of genes (rows) and number of cells (columns)</span></span></code></pre></div>
<pre><code>## [1] 33694  7802</code></pre>
<p>Here we see the counts matrix has 33694 genes and 7802 cells.</p>
</div>
<div id="how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix" class="section level3">
<h3><span class="header-section-number">4.2.5</span> How much memory does a sparse matrix take up relative to a dense matrix?</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="data-wrangling-scrnaseq-1.html#cb7-1"></a><span class="kw">object.size</span>(counts) <span class="co"># size in bytes</span></span></code></pre></div>
<pre><code>## 169457000 bytes</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="data-wrangling-scrnaseq-1.html#cb9-1"></a><span class="kw">object.size</span>(<span class="kw">as.matrix</span>(counts)) <span class="co"># size in bytes</span></span></code></pre></div>
<pre><code>## 2106010424 bytes</code></pre>
<p>We see here that the sparse matrix takes 225 Mb in memory while storing the matrix in a dense format (where all count values including zeros are stored) takes almost 10 times as much memory! This memory saving is very important, especially as data sets are now being created that are beyond a million cells. These matrices can become unmanageable without special computing resources.</p>
<p>In the sparse representation, we assume that the majority of count values in a matrix are zero. We only store the non-zero values. This is implemented in the Matrix package using a dgTMatrix object.</p>
</div>
</div>
<div id="filtering-low-quality-cells" class="section level2">
<h2><span class="header-section-number">4.3</span> Filtering low-quality cells</h2>
<p>You can learn a lot about your scRNA-seq data’s quality with simple plotting. Let’s do some plotting to look at the number of reads per cell, reads per genes, expressed genes per cell (often called complexity), and rarity of genes (cells expressing genes).</p>
<div id="look-at-the-summary-counts-for-genes-and-cells" class="section level3">
<h3><span class="header-section-number">4.3.1</span> Look at the summary counts for genes and cells</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="data-wrangling-scrnaseq-1.html#cb11-1"></a>counts_per_cell &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(counts)</span>
<span id="cb11-2"><a href="data-wrangling-scrnaseq-1.html#cb11-2"></a><span class="kw">cat</span>(<span class="st">&quot;counts per cell: &quot;</span>, counts_per_cell[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>], <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>) <span class="co">## counts for first 5 cells</span></span></code></pre></div>
<pre><code>## counts per cell:  3605 3828 6457 3075 9400</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="data-wrangling-scrnaseq-1.html#cb13-1"></a>counts_per_gene &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">rowSums</span>(counts)</span>
<span id="cb13-2"><a href="data-wrangling-scrnaseq-1.html#cb13-2"></a><span class="kw">cat</span>(<span class="st">&quot;counts per gene: &quot;</span>, counts_per_gene[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>], <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)  <span class="co">## counts for first 5 genes</span></span></code></pre></div>
<pre><code>## counts per gene:  2 0 0 2 0</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="data-wrangling-scrnaseq-1.html#cb15-1"></a>genes_per_cell &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="co"># count gene only if it has non-zero reads mapped.</span></span>
<span id="cb15-2"><a href="data-wrangling-scrnaseq-1.html#cb15-2"></a><span class="kw">cat</span>(<span class="st">&quot;counts for non-zero genes: &quot;</span>, genes_per_cell[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])  <span class="co">## counts for first 5 genes</span></span></code></pre></div>
<pre><code>## counts for non-zero genes:  1184 1387 1784 1092 2626</code></pre>
<p><strong>Task: In a similar way, can you calculate cells per genes? replace the ‘?’ in the command below</strong></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="data-wrangling-scrnaseq-1.html#cb17-1"></a><span class="co">#### cells_per_gene &lt;- Matrix::?(counts&gt;?) # only count cells where the gene is expressed</span></span>
<span id="cb17-2"><a href="data-wrangling-scrnaseq-1.html#cb17-2"></a>cells_per_gene &lt;-<span class="st">  </span><span class="co"># only count cells where the gene is expressed</span></span>
<span id="cb17-3"><a href="data-wrangling-scrnaseq-1.html#cb17-3"></a><span class="kw">cat</span>(<span class="st">&quot;count of cells with expressed genes: &quot;</span>, cells_per_gene)</span></code></pre></div>
<p>colSums and rowSums are functions that work on each row or column in a matrix and return the column sums or row sums as a vector. If this is true counts_per_cell should have 1 entry per cell. Let’s make sure the length of the returned vector matches the matrix dimension for column. How would you do that? ( Hint:length() ).</p>
<p>Notes:
1. Matrix::colSums is a way to force functions from the Matrix library to be used. There are many libraries that implement colSums, we are forcing the one from the Matrix library to be used here to make sure it handles the dgTmatrix (sparse matrix) correctly. This is good practice.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="data-wrangling-scrnaseq-1.html#cb18-1"></a><span class="kw">hist</span>(<span class="kw">log10</span>(counts_per_cell<span class="op">+</span><span class="dv">1</span>),<span class="dt">main=</span><span class="st">&#39;counts per cell&#39;</span>,<span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_counts-1.png" width="672" /></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="data-wrangling-scrnaseq-1.html#cb19-1"></a><span class="kw">hist</span>(<span class="kw">log10</span>(genes_per_cell<span class="op">+</span><span class="dv">1</span>), <span class="dt">main=</span><span class="st">&#39;genes per cell&#39;</span>, <span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_counts-2.png" width="672" /></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="data-wrangling-scrnaseq-1.html#cb20-1"></a><span class="kw">plot</span>(counts_per_cell, genes_per_cell, <span class="dt">log=</span><span class="st">&#39;xy&#39;</span>, <span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span>
<span id="cb20-2"><a href="data-wrangling-scrnaseq-1.html#cb20-2"></a><span class="kw">title</span>(<span class="st">&#39;counts vs genes per cell&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_counts-3.png" width="672" /></p>
<p>Here we see examples of plotting a new plot, the histogram. R makes this really easy with the hist function. We are also transforming the values to log10 before plotting, this is done with the log10 method. When logging count data, the + 1 is used to avoid log10(0) which is not defined.</p>
<p>Can you make a histogram of counts per gene in log10 scale?</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="data-wrangling-scrnaseq-1.html#cb21-1"></a><span class="kw">hist</span>(<span class="kw">log10</span>(counts_per_gene<span class="op">+</span><span class="dv">1</span>), <span class="dt">main=</span><span class="st">&#39;counts per gene&#39;</span>, <span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/hist_genes-1.png" width="672" /></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="data-wrangling-scrnaseq-1.html#cb22-1"></a><span class="co">### hist(?(?+1), main=&#39;counts per gene&#39;, col=&#39;wheat&#39;)</span></span></code></pre></div>
</div>
<div id="plot-cells-ranked-by-their-number-of-detected-genes." class="section level3">
<h3><span class="header-section-number">4.3.2</span> Plot cells ranked by their number of detected genes.</h3>
<p>Here we rank each cell by its library complexity, ie the number of genes detected per cell. This is a very useful plot as it shows the distribution of library complexity in the sequencing run. One can use this plot to investigate observations (potential cells) that are actually failed libraries (lower end outliers) or observations that are cell doublets (higher end outliers).</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="data-wrangling-scrnaseq-1.html#cb23-1"></a><span class="kw">plot</span>(<span class="kw">sort</span>(genes_per_cell), <span class="dt">xlab=</span><span class="st">&#39;cell&#39;</span>, <span class="dt">log=</span><span class="st">&#39;y&#39;</span>, <span class="dt">main=</span><span class="st">&#39;genes per cell (ordered)&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/rank_cells-1.png" width="672" /></p>
</div>
</div>
<div id="beginning-with-seurat" class="section level2">
<h2><span class="header-section-number">4.4</span> Beginning with Seurat:</h2>
<p><a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a></p>
<div id="creating-a-seurat-object" class="section level3">
<h3><span class="header-section-number">4.4.1</span> Creating a seurat object</h3>
<p>To analyze our single cell data we will use a seurat object. Can you create an Seurat object with the 10x data and save it in an object called ‘seurat’? hint: CreateSeuratObject().
Can you include only genes that are are expressed in 3 or more cells and cells with complexity of 350 genes or more?
How many genes are you left with? How many cells?</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="data-wrangling-scrnaseq-1.html#cb24-1"></a><span class="co">### seurat&lt;-CreateSeuratObject(raw.data = counts, ? = 3, ? = 350, project = &quot;10X_NSCLC&quot;)</span></span>
<span id="cb24-2"><a href="data-wrangling-scrnaseq-1.html#cb24-2"></a>seurat &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(<span class="dt">counts =</span> counts, <span class="dt">min.cells =</span> <span class="dv">3</span>, <span class="dt">min.features =</span> <span class="dv">350</span>, <span class="dt">project =</span> <span class="st">&quot;10X_NSCLC&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Feature names cannot have underscores (&#39;_&#39;), replacing
## with dashes (&#39;-&#39;)</code></pre>
<div id="object-slots" class="section level4">
<h4><span class="header-section-number">4.4.1.1</span> Object Slots</h4>
<p><a href="https://github.com/satijalab/seurat/wiki/Seurat#slots">reference</a></p>
<p>Almost all our analysis will be on the single object, of class Seurat. This object contains various “slots” (designated by <a href="mailto:seurat@slotname" class="email">seurat@slotname</a>) that will store not only the raw count data, but also the results from various computations below. This has the advantage that we do not need to keep track of inidividual variables of interest - they can all be collapsed into a single object as long as these slots are pre-defined as:</p>
<ul>
<li>assays - A list of assays within this object</li>
<li>meta.data - Cell-level meta data</li>
<li>active.assay - Name of active, or default, assay</li>
<li>active.ident - Identity classes for the current object</li>
<li>graphs - A list of nearest neighbor graphs</li>
<li>reductions - A list of DimReduc objects</li>
<li>project.name - User-defined project name (optional)</li>
<li>tools - Empty list. Tool developers can store any internal data from their methods here</li>
<li>misc - Empty slot. User can store additional information here
version Seurat version used when creating the object</li>
</ul>
<p><code>seurat@assays$RNA@counts</code> is a slot that stores the original gene count matrix. We can view the first 10 rows (genes) and the first 10 columns (cells).</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="data-wrangling-scrnaseq-1.html#cb26-1"></a>seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>counts[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;</code></pre>
<pre><code>##    [[ suppressing 10 column names &#39;AAACCTGAGCTAGTCT&#39;, &#39;AAACCTGAGGGCACTA&#39;, &#39;AAACCTGAGTACGTTC&#39; ... ]]</code></pre>
<pre><code>##                                  
## FO538757.2    . . . . 1 2 . . . .
## AP006222.2    . . . . . . . . . .
## RP4-669L17.10 . . . . . . . . . .
## RP11-206L10.9 . . . . . . . . . .
## LINC00115     . . . . . . . . . .
## FAM41C        . . . . . . . . . .
## SAMD11        . . . . . 1 . . . .
## NOC2L         . . . . 2 4 . 1 . .
## KLHL17        . . . . . . . . . .
## PLEKHN1       . . . . . . . . . .</code></pre>
</div>
<div id="object-information" class="section level4">
<h4><span class="header-section-number">4.4.1.2</span> Object Information</h4>
<p>Summary information about Seurat objects can be had quickly and easily using standard R functions. Object shape/dimensions can be found using the dim, ncol, and nrow functions; cell and feature names can be found using the colnames and rownames functions, respectively, or the dimnames function. A vector of names of Assay, DimReduc, and Graph objects contained in a Seurat object can be had by using names.</p>
<p><strong>The following examples use the PBMC 3k dataset</strong></p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="data-wrangling-scrnaseq-1.html#cb30-1"></a>seurat</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 20213 features across 7109 samples within 1 assay 
## Active assay: RNA (20213 features)</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="data-wrangling-scrnaseq-1.html#cb32-1"></a><span class="co"># nrow and ncol provide the number of features and cells in the active assay, respectively</span></span>
<span id="cb32-2"><a href="data-wrangling-scrnaseq-1.html#cb32-2"></a><span class="co"># dim provides both nrow and ncol at the same time</span></span>
<span id="cb32-3"><a href="data-wrangling-scrnaseq-1.html#cb32-3"></a><span class="kw">dim</span>(<span class="dt">x =</span> seurat)</span></code></pre></div>
<pre><code>## [1] 20213  7109</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="data-wrangling-scrnaseq-1.html#cb34-1"></a><span class="co"># In addtion to rownames and colnames, one can use dimnames</span></span>
<span id="cb34-2"><a href="data-wrangling-scrnaseq-1.html#cb34-2"></a><span class="co"># which provides a two-length list with both rownames and colnames</span></span>
<span id="cb34-3"><a href="data-wrangling-scrnaseq-1.html#cb34-3"></a><span class="kw">head</span>(<span class="dt">x =</span> <span class="kw">rownames</span>(<span class="dt">x =</span> seurat))</span></code></pre></div>
<pre><code>## [1] &quot;FO538757.2&quot;    &quot;AP006222.2&quot;    &quot;RP4-669L17.10&quot;
## [4] &quot;RP11-206L10.9&quot; &quot;LINC00115&quot;     &quot;FAM41C&quot;</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="data-wrangling-scrnaseq-1.html#cb36-1"></a><span class="kw">head</span>(<span class="dt">x =</span> <span class="kw">colnames</span>(<span class="dt">x =</span> seurat))</span></code></pre></div>
<pre><code>## [1] &quot;AAACCTGAGCTAGTCT&quot; &quot;AAACCTGAGGGCACTA&quot; &quot;AAACCTGAGTACGTTC&quot;
## [4] &quot;AAACCTGAGTCCGGTC&quot; &quot;AAACCTGCACCAGGTC&quot; &quot;AAACCTGCACCTCGTT&quot;</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="data-wrangling-scrnaseq-1.html#cb38-1"></a><span class="co"># A vector of names of associated objects can be had with the names function</span></span>
<span id="cb38-2"><a href="data-wrangling-scrnaseq-1.html#cb38-2"></a><span class="co"># These can be passed to the double [[ extract operator to pull them from the Seurat object</span></span>
<span id="cb38-3"><a href="data-wrangling-scrnaseq-1.html#cb38-3"></a><span class="kw">names</span>(<span class="dt">x =</span> seurat)</span></code></pre></div>
<pre><code>## [1] &quot;RNA&quot;</code></pre>
<p>You can also see a print out of the entire structure of the object by using the <code>str</code> commmand.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="data-wrangling-scrnaseq-1.html#cb40-1"></a><span class="kw">str</span>(seurat)</span></code></pre></div>
<pre><code>## Formal class &#39;Seurat&#39; [package &quot;Seurat&quot;] with 12 slots
##   ..@ assays      :List of 1
##   .. ..$ RNA:Formal class &#39;Assay&#39; [package &quot;Seurat&quot;] with 8 slots
##   .. .. .. ..@ counts       :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. .. .. .. .. ..@ i       : int [1:13732088] 22 37 100 102 105 132 146 171 177 185 ...
##   .. .. .. .. .. ..@ p       : int [1:7110] 0 1184 2571 4355 5447 8072 13674 15861 17504 18605 ...
##   .. .. .. .. .. ..@ Dim     : int [1:2] 20213 7109
##   .. .. .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. .. .. ..$ : chr [1:20213] &quot;FO538757.2&quot; &quot;AP006222.2&quot; &quot;RP4-669L17.10&quot; &quot;RP11-206L10.9&quot; ...
##   .. .. .. .. .. .. ..$ : chr [1:7109] &quot;AAACCTGAGCTAGTCT&quot; &quot;AAACCTGAGGGCACTA&quot; &quot;AAACCTGAGTACGTTC&quot; &quot;AAACCTGAGTCCGGTC&quot; ...
##   .. .. .. .. .. ..@ x       : num [1:13732088] 1 2 7 1 1 3 1 1 1 1 ...
##   .. .. .. .. .. ..@ factors : list()
##   .. .. .. ..@ data         :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. .. .. .. .. ..@ i       : int [1:13732088] 22 37 100 102 105 132 146 171 177 185 ...
##   .. .. .. .. .. ..@ p       : int [1:7110] 0 1184 2571 4355 5447 8072 13674 15861 17504 18605 ...
##   .. .. .. .. .. ..@ Dim     : int [1:2] 20213 7109
##   .. .. .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. .. .. ..$ : chr [1:20213] &quot;FO538757.2&quot; &quot;AP006222.2&quot; &quot;RP4-669L17.10&quot; &quot;RP11-206L10.9&quot; ...
##   .. .. .. .. .. .. ..$ : chr [1:7109] &quot;AAACCTGAGCTAGTCT&quot; &quot;AAACCTGAGGGCACTA&quot; &quot;AAACCTGAGTACGTTC&quot; &quot;AAACCTGAGTCCGGTC&quot; ...
##   .. .. .. .. .. ..@ x       : num [1:13732088] 1 2 7 1 1 3 1 1 1 1 ...
##   .. .. .. .. .. ..@ factors : list()
##   .. .. .. ..@ scale.data   : num[0 , 0 ] 
##   .. .. .. ..@ key          : chr &quot;rna_&quot;
##   .. .. .. ..@ assay.orig   : NULL
##   .. .. .. ..@ var.features : logi(0) 
##   .. .. .. ..@ meta.features:&#39;data.frame&#39;:   20213 obs. of  0 variables
##   .. .. .. ..@ misc         : NULL
##   ..@ meta.data   :&#39;data.frame&#39;: 7109 obs. of  3 variables:
##   .. ..$ orig.ident  : Factor w/ 1 level &quot;10X_NSCLC&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ nCount_RNA  : num [1:7109] 3605 3828 6457 3075 9399 ...
##   .. ..$ nFeature_RNA: int [1:7109] 1184 1387 1784 1092 2625 5602 2187 1643 1101 2717 ...
##   ..@ active.assay: chr &quot;RNA&quot;
##   ..@ active.ident: Factor w/ 1 level &quot;10X_NSCLC&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..- attr(*, &quot;names&quot;)= chr [1:7109] &quot;AAACCTGAGCTAGTCT&quot; &quot;AAACCTGAGGGCACTA&quot; &quot;AAACCTGAGTACGTTC&quot; &quot;AAACCTGAGTCCGGTC&quot; ...
##   ..@ graphs      : list()
##   ..@ neighbors   : list()
##   ..@ reductions  : list()
##   ..@ project.name: chr &quot;10X_NSCLC&quot;
##   ..@ misc        : list()
##   ..@ version     :Classes &#39;package_version&#39;, &#39;numeric_version&#39;  hidden list of 1
##   .. ..$ : int [1:3] 3 1 2
##   ..@ commands    : list()
##   ..@ tools       : list()</code></pre>
</div>
<div id="data-access" class="section level4">
<h4><span class="header-section-number">4.4.1.3</span> Data Access</h4>
<p>Pulling specific Assay, DimReduc, or Graph objects can be done with the double [[ extract operator. Adding new objects to a Seurat object is also done with the double [[ extract operator; Seurat will figure out where in the Seurat object a new associated object belongs.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="data-wrangling-scrnaseq-1.html#cb42-1"></a>seurat[[<span class="st">&#39;RNA&#39;</span>]]</span></code></pre></div>
<pre><code>## Assay data with 20213 features for 7109 cells
## First 10 features:
##  FO538757.2, AP006222.2, RP4-669L17.10, RP11-206L10.9,
## LINC00115, FAM41C, SAMD11, NOC2L, KLHL17, PLEKHN1</code></pre>
<p>Accessing data from an Seurat object is done with the GetAssayData function. Adding expression data to either the counts, data, or scale.data slots can be done with SetAssayData. New data must have the same cells in the same order as the current expression data. Data added to counts or data must have the same features as the current expression data.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="data-wrangling-scrnaseq-1.html#cb44-1"></a><span class="co">## same as above but using object Method</span></span>
<span id="cb44-2"><a href="data-wrangling-scrnaseq-1.html#cb44-2"></a><span class="kw">GetAssayData</span>(<span class="dt">object =</span> seurat, <span class="dt">slot =</span> <span class="st">&#39;data&#39;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;</code></pre>
<pre><code>##    [[ suppressing 10 column names &#39;AAACCTGAGCTAGTCT&#39;, &#39;AAACCTGAGGGCACTA&#39;, &#39;AAACCTGAGTACGTTC&#39; ... ]]</code></pre>
<pre><code>##                                  
## FO538757.2    . . . . 1 2 . . . .
## AP006222.2    . . . . . . . . . .
## RP4-669L17.10 . . . . . . . . . .
## RP11-206L10.9 . . . . . . . . . .
## LINC00115     . . . . . . . . . .
## FAM41C        . . . . . . . . . .
## SAMD11        . . . . . 1 . . . .
## NOC2L         . . . . 2 4 . 1 . .
## KLHL17        . . . . . . . . . .
## PLEKHN1       . . . . . . . . . .</code></pre>
<p>Cell-level meta data can be accessed with the single [ extract operator or using the <code>$</code> sign. Pulling with the <code>$</code> sign means only one bit of meta data can be pulled at a time, though tab-autocompletion has been enabled for it, making it ideal for interactive use. Adding cell-level meta data can be set using the single [ extract operator as well, or by using AddMetaData.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="data-wrangling-scrnaseq-1.html#cb48-1"></a><span class="co"># Cell-level meta data is stored as a data frame</span></span>
<span id="cb48-2"><a href="data-wrangling-scrnaseq-1.html#cb48-2"></a><span class="co"># Standard data frame functions work on the meta data data frame</span></span>
<span id="cb48-3"><a href="data-wrangling-scrnaseq-1.html#cb48-3"></a><span class="kw">colnames</span>(<span class="dt">x =</span> seurat[])</span></code></pre></div>
<pre><code>##    [1] &quot;AAACCTGAGCTAGTCT&quot; &quot;AAACCTGAGGGCACTA&quot; &quot;AAACCTGAGTACGTTC&quot;
##    [4] &quot;AAACCTGAGTCCGGTC&quot; &quot;AAACCTGCACCAGGTC&quot; &quot;AAACCTGCACCTCGTT&quot;
##    [7] &quot;AAACCTGCAGCTGTGC&quot; &quot;AAACCTGGTAAATGAC&quot; &quot;AAACCTGGTACTTAGC&quot;
##   [10] &quot;AAACCTGTCCTCAATT&quot; &quot;AAACCTGTCCTGCAGG&quot; &quot;AAACGGGAGCTTATCG&quot;
##   [13] &quot;AAACGGGAGGCCCTTG&quot; &quot;AAACGGGCAAGTAATG&quot; &quot;AAACGGGCACCAGATT&quot;
##   [16] &quot;AAACGGGGTAATCACC&quot; &quot;AAACGGGGTAGCTGCC&quot; &quot;AAACGGGGTGGTGTAG&quot;
##   [19] &quot;AAACGGGTCACCTCGT&quot; &quot;AAACGGGTCCCAAGTA&quot; &quot;AAAGATGAGAAACCTA&quot;
##   [22] &quot;AAAGATGAGACTTTCG&quot; &quot;AAAGATGAGATATGGT&quot; &quot;AAAGATGAGATCCGAG&quot;
##   [25] &quot;AAAGATGAGATGAGAG&quot; &quot;AAAGATGAGCCAGTAG&quot; &quot;AAAGATGCAATCCAAC&quot;
##   [28] &quot;AAAGATGCACGAAATA&quot; &quot;AAAGATGCACGTAAGG&quot; &quot;AAAGATGCAGAGTGTG&quot;
##   [31] &quot;AAAGATGCAGCCAGAA&quot; &quot;AAAGATGCAGGACCCT&quot; &quot;AAAGATGGTCATACTG&quot;
##   [34] &quot;AAAGATGGTCCGTGAC&quot; &quot;AAAGATGGTCGAAAGC&quot; &quot;AAAGATGGTGCCTGGT&quot;
##   [37] &quot;AAAGATGGTTACCGAT&quot; &quot;AAAGATGGTTGGTAAA&quot; &quot;AAAGATGTCAGCGATT&quot;
##   [40] &quot;AAAGATGTCCCAAGTA&quot; &quot;AAAGATGTCCGAATGT&quot; &quot;AAAGATGTCCTAGTGA&quot;
##   [43] &quot;AAAGATGTCGAACTGT&quot; &quot;AAAGATGTCTGTCTCG&quot; &quot;AAAGCAAAGAATTGTG&quot;
##   [46] &quot;AAAGCAAAGAGACGAA&quot; &quot;AAAGCAAAGAGTTGGC&quot; &quot;AAAGCAACAAGGTTCT&quot;
##   [49] &quot;AAAGCAACACGCCAGT&quot; &quot;AAAGCAAGTAGCTGCC&quot; &quot;AAAGCAAGTGCCTGTG&quot;
##   [52] &quot;AAAGCAAGTTAGTGGG&quot; &quot;AAAGCAATCACTATTC&quot; &quot;AAAGCAATCATCGGAT&quot;
##   [55] &quot;AAAGCAATCATCTGCC&quot; &quot;AAAGCAATCCTAGAAC&quot; &quot;AAAGCAATCGAACGGA&quot;
##   [58] &quot;AAAGCAATCGTACGGC&quot; &quot;AAAGCAATCTCAAGTG&quot; &quot;AAAGTAGAGAGTAATC&quot;
##   [61] &quot;AAAGTAGAGCCGATTT&quot; &quot;AAAGTAGAGGATTCGG&quot; &quot;AAAGTAGCAATGTAAG&quot;
##   [64] &quot;AAAGTAGCACGACTCG&quot; &quot;AAAGTAGGTAATTGGA&quot; &quot;AAAGTAGGTATCAGTC&quot;
##   [67] &quot;AAAGTAGGTCGCGAAA&quot; &quot;AAAGTAGGTGGTCCGT&quot; &quot;AAAGTAGGTTGCCTCT&quot;
##   [70] &quot;AAAGTAGTCAGGCAAG&quot; &quot;AAAGTAGTCCAAAGTC&quot; &quot;AAAGTAGTCTTGTCAT&quot;
##   [73] &quot;AAATGCCAGACGACGT&quot; &quot;AAATGCCAGGCAAAGA&quot; &quot;AAATGCCAGGGTGTTG&quot;
##   [76] &quot;AAATGCCAGGTAAACT&quot; &quot;AAATGCCCAATGTAAG&quot; &quot;AAATGCCCACACATGT&quot;
##   [79] &quot;AAATGCCCAGCTCGAC&quot; &quot;AAATGCCCATCTATGG&quot; &quot;AAATGCCGTACTTCTT&quot;
##   [82] &quot;AAATGCCGTCTTGCGG&quot; &quot;AAATGCCGTGTAACGG&quot; &quot;AAATGCCTCCAAACAC&quot;
##   [85] &quot;AAATGCCTCCGCGGTA&quot; &quot;AAATGCCTCGTCGTTC&quot; &quot;AACACGTAGCTAGTCT&quot;
##   [88] &quot;AACACGTAGGATTCGG&quot; &quot;AACACGTAGTACCGGA&quot; &quot;AACACGTCACAGAGGT&quot;
##   [91] &quot;AACACGTCACGACTCG&quot; &quot;AACACGTCAGTCAGCC&quot; &quot;AACACGTGTAATCGTC&quot;
##   [94] &quot;AACACGTGTACCAGTT&quot; &quot;AACACGTGTTGGGACA&quot; &quot;AACACGTTCACCACCT&quot;
##   [97] &quot;AACACGTTCACCGGGT&quot; &quot;AACACGTTCCAGAGGA&quot; &quot;AACACGTTCCTGCAGG&quot;
##  [100] &quot;AACACGTTCGCATGGC&quot; &quot;AACACGTTCTTACCGC&quot; &quot;AACACGTTCTTGCATT&quot;
##  [103] &quot;AACCATGAGTAGTGCG&quot; &quot;AACCATGCAAAGGCGT&quot; &quot;AACCATGCAAGCGAGT&quot;
##  [106] &quot;AACCATGCAGCATACT&quot; &quot;AACCATGCATGTTGAC&quot; &quot;AACCATGCATTCTCAT&quot;
##  [109] &quot;AACCATGGTCCGAAGA&quot; &quot;AACCATGGTGCCTGCA&quot; &quot;AACCATGGTGTTGAGG&quot;
##  [112] &quot;AACCATGGTTGATTCG&quot; &quot;AACCATGTCAATCTCT&quot; &quot;AACCATGTCCGAAGAG&quot;
##  [115] &quot;AACCATGTCTCCGGTT&quot; &quot;AACCGCGAGAGGTTGC&quot; &quot;AACCGCGAGGGTCTCC&quot;
##  [118] &quot;AACCGCGCAAGGTGTG&quot; &quot;AACCGCGCAGGACGTA&quot; &quot;AACCGCGCAGGGTTAG&quot;
##  [121] &quot;AACCGCGCAGTAACGG&quot; &quot;AACCGCGCATTCCTGC&quot; &quot;AACCGCGGTGACTCAT&quot;
##  [124] &quot;AACCGCGGTGGGTCAA&quot; &quot;AACCGCGTCAATCTCT&quot; &quot;AACCGCGTCCTTGGTC&quot;
##  [127] &quot;AACCGCGTCGCACTCT&quot; &quot;AACCGCGTCGGCCGAT&quot; &quot;AACCGCGTCTTGCCGT&quot;
##  [130] &quot;AACGTTGAGAACTCGG&quot; &quot;AACGTTGAGGGTGTGT&quot; &quot;AACGTTGCAGGTCGTC&quot;
##  [133] &quot;AACGTTGCAGTCGTGC&quot; &quot;AACGTTGCATATACGC&quot; &quot;AACGTTGTCATACGGT&quot;
##  [136] &quot;AACGTTGTCCACGACG&quot; &quot;AACGTTGTCCGTCATC&quot; &quot;AACGTTGTCTGGTATG&quot;
##  [139] &quot;AACTCAGAGACTAAGT&quot; &quot;AACTCAGAGCGTAATA&quot; &quot;AACTCAGAGCGTCTAT&quot;
##  [142] &quot;AACTCAGAGCTGAAAT&quot; &quot;AACTCAGAGTGGACGT&quot; &quot;AACTCAGCACATAACC&quot;
##  [145] &quot;AACTCAGGTGTTCGAT&quot; &quot;AACTCAGGTTCTCATT&quot; &quot;AACTCAGTCACCCTCA&quot;
##  [148] &quot;AACTCAGTCAGCCTAA&quot; &quot;AACTCAGTCCTATGTT&quot; &quot;AACTCAGTCTAACCGA&quot;
##  [151] &quot;AACTCAGTCTCGTATT&quot; &quot;AACTCCCAGACAAGCC&quot; &quot;AACTCCCAGACTAGGC&quot;
##  [154] &quot;AACTCCCAGGATCGCA&quot; &quot;AACTCCCAGGCAGGTT&quot; &quot;AACTCCCAGTACGTTC&quot;
##  [157] &quot;AACTCCCAGTCGTTTG&quot; &quot;AACTCCCAGTGACTCT&quot; &quot;AACTCCCCACAAGTAA&quot;
##  [160] &quot;AACTCCCCACGCATCG&quot; &quot;AACTCCCCAGTATCTG&quot; &quot;AACTCCCCATCCTAGA&quot;
##  [163] &quot;AACTCCCCATCGTCGG&quot; &quot;AACTCCCGTCGAACAG&quot; &quot;AACTCCCGTGCACGAA&quot;
##  [166] &quot;AACTCCCGTTCACCTC&quot; &quot;AACTCCCTCAGTTTGG&quot; &quot;AACTCCCTCTGGTGTA&quot;
##  [169] &quot;AACTCTTCAAGAAGAG&quot; &quot;AACTCTTCAAGTCTGT&quot; &quot;AACTCTTCAATGTTGC&quot;
##  [172] &quot;AACTCTTCATCTCCCA&quot; &quot;AACTCTTCATTTCAGG&quot; &quot;AACTCTTGTGCGGTAA&quot;
##  [175] &quot;AACTCTTGTGTCCTCT&quot; &quot;AACTCTTGTTTGGGCC&quot; &quot;AACTCTTTCAACGAAA&quot;
##  [178] &quot;AACTCTTTCCTTGCCA&quot; &quot;AACTCTTTCTCGTTTA&quot; &quot;AACTGGTAGAGGTACC&quot;
##  [181] &quot;AACTGGTAGGGAAACA&quot; &quot;AACTGGTAGGTGTGGT&quot; &quot;AACTGGTAGTACGACG&quot;
##  [184] &quot;AACTGGTCAAACTGCT&quot; &quot;AACTGGTCACGAGGTA&quot; &quot;AACTGGTGTACATGTC&quot;
##  [187] &quot;AACTGGTGTGGCAAAC&quot; &quot;AACTGGTGTTACCAGT&quot; &quot;AACTGGTGTTGAGGTG&quot;
##  [190] &quot;AACTGGTGTTGGAGGT&quot; &quot;AACTGGTGTTTGACAC&quot; &quot;AACTGGTGTTTGTTTC&quot;
##  [193] &quot;AACTGGTTCAATCACG&quot; &quot;AACTGGTTCACAAACC&quot; &quot;AACTGGTTCACCCGAG&quot;
##  [196] &quot;AACTGGTTCAGCACAT&quot; &quot;AACTGGTTCCTTTACA&quot; &quot;AACTGGTTCGGGAGTA&quot;
##  [199] &quot;AACTGGTTCGTCACGG&quot; &quot;AACTGGTTCTCTTATG&quot; &quot;AACTTTCAGGCAGTCA&quot;
##  [202] &quot;AACTTTCAGGTTACCT&quot; &quot;AACTTTCCACCAGGCT&quot; &quot;AACTTTCCAGGTGCCT&quot;
##  [205] &quot;AACTTTCCATGAACCT&quot; &quot;AACTTTCCATGTAAGA&quot; &quot;AACTTTCCATGTCCTC&quot;
##  [208] &quot;AACTTTCGTAAGTGTA&quot; &quot;AACTTTCGTACACCGC&quot; &quot;AACTTTCGTATAGGTA&quot;
##  [211] &quot;AACTTTCGTCTAGTCA&quot; &quot;AACTTTCGTTAAAGAC&quot; &quot;AACTTTCGTTACGGAG&quot;
##  [214] &quot;AACTTTCGTTGTGGAG&quot; &quot;AACTTTCTCACTCCTG&quot; &quot;AACTTTCTCGAACTGT&quot;
##  [217] &quot;AACTTTCTCGCAAACT&quot; &quot;AACTTTCTCTTGCAAG&quot; &quot;AAGACCTAGAGGTAGA&quot;
##  [220] &quot;AAGACCTAGGATATAC&quot; &quot;AAGACCTAGGCTATCT&quot; &quot;AAGACCTAGTCCAGGA&quot;
##  [223] &quot;AAGACCTAGTCGTTTG&quot; &quot;AAGACCTAGTGACATA&quot; &quot;AAGACCTCAAGCCTAT&quot;
##  [226] &quot;AAGACCTCAATGAATG&quot; &quot;AAGACCTCATACAGCT&quot; &quot;AAGACCTCATGGGAAC&quot;
##  [229] &quot;AAGACCTCATTTGCCC&quot; &quot;AAGACCTGTAATAGCA&quot; &quot;AAGACCTGTAGCTAAA&quot;
##  [232] &quot;AAGACCTGTCCGAGTC&quot; &quot;AAGACCTGTCTAGAGG&quot; &quot;AAGACCTGTGAAAGAG&quot;
##  [235] &quot;AAGACCTGTTGACGTT&quot; &quot;AAGACCTTCACCGTAA&quot; &quot;AAGACCTTCCGGCACA&quot;
##  [238] &quot;AAGACCTTCTCCAGGG&quot; &quot;AAGCCGCAGAGTCGGT&quot; &quot;AAGCCGCAGCACGCCT&quot;
##  [241] &quot;AAGCCGCAGCCAGGAT&quot; &quot;AAGCCGCAGCTGCGAA&quot; &quot;AAGCCGCCACGACGAA&quot;
##  [244] &quot;AAGCCGCCATGGTTGT&quot; &quot;AAGCCGCCATTCCTCG&quot; &quot;AAGCCGCGTACGCACC&quot;
##  [247] &quot;AAGCCGCGTCCTCCAT&quot; &quot;AAGCCGCGTTACCAGT&quot; &quot;AAGCCGCGTTGACGTT&quot;
##  [250] &quot;AAGCCGCTCATGCAAC&quot; &quot;AAGCCGCTCCAGAGGA&quot; &quot;AAGCCGCTCCTATTCA&quot;
##  [253] &quot;AAGCCGCTCCTCCTAG&quot; &quot;AAGCCGCTCGGATGGA&quot; &quot;AAGCCGCTCGTAGGAG&quot;
##  [256] &quot;AAGCCGCTCTAACTGG&quot; &quot;AAGGAGCAGATGGCGT&quot; &quot;AAGGAGCAGCCCAATT&quot;
##  [259] &quot;AAGGAGCAGCGGCTTC&quot; &quot;AAGGAGCAGGATATAC&quot; &quot;AAGGAGCAGTAACCCT&quot;
##  [262] &quot;AAGGAGCCAGACGCTC&quot; &quot;AAGGAGCCAGACGTAG&quot; &quot;AAGGAGCCAGCCTATA&quot;
##  [265] &quot;AAGGAGCGTCTCAACA&quot; &quot;AAGGAGCTCAATCACG&quot; &quot;AAGGAGCTCAGGCAAG&quot;
##  [268] &quot;AAGGAGCTCGGAAACG&quot; &quot;AAGGAGCTCTAGCACA&quot; &quot;AAGGAGCTCTCAAACG&quot;
##  [271] &quot;AAGGAGCTCTTGAGAC&quot; &quot;AAGGCAGAGGATGGAA&quot; &quot;AAGGCAGAGTGACATA&quot;
##  [274] &quot;AAGGCAGCAAGTAATG&quot; &quot;AAGGCAGCACACATGT&quot; &quot;AAGGCAGCACGAGGTA&quot;
##  [277] &quot;AAGGCAGCATACTACG&quot; &quot;AAGGCAGCATCAGTAC&quot; &quot;AAGGCAGCATTCACTT&quot;
##  [280] &quot;AAGGCAGGTAGCCTAT&quot; &quot;AAGGCAGGTGATGTGG&quot; &quot;AAGGCAGGTGTTTGGT&quot;
##  [283] &quot;AAGGCAGGTTGTGGCC&quot; &quot;AAGGCAGTCAATAAGG&quot; &quot;AAGGCAGTCTGCTGTC&quot;
##  [286] &quot;AAGGTTCAGAACAATC&quot; &quot;AAGGTTCAGAATCTCC&quot; &quot;AAGGTTCAGCGTTGCC&quot;
##  [289] &quot;AAGGTTCAGGCCGAAT&quot; &quot;AAGGTTCAGGCTATCT&quot; &quot;AAGGTTCAGGGTGTGT&quot;
##  [292] &quot;AAGGTTCAGTCAAGCG&quot; &quot;AAGGTTCAGTTACCCA&quot; &quot;AAGGTTCCAACACCCG&quot;
##  [295] &quot;AAGGTTCCACAGCGTC&quot; &quot;AAGGTTCCACTAAGTC&quot; &quot;AAGGTTCCACTTGGAT&quot;
##  [298] &quot;AAGGTTCCATAAGACA&quot; &quot;AAGGTTCCATCACAAC&quot; &quot;AAGGTTCCATCGGACC&quot;
##  [301] &quot;AAGGTTCGTATAGTAG&quot; &quot;AAGGTTCGTGATGATA&quot; &quot;AAGGTTCGTTATGCGT&quot;
##  [304] &quot;AAGGTTCGTTCCGGCA&quot; &quot;AAGGTTCTCCCTTGTG&quot; &quot;AAGTCTGAGATGTGGC&quot;
##  [307] &quot;AAGTCTGAGGATCGCA&quot; &quot;AAGTCTGCAAGCCGCT&quot; &quot;AAGTCTGGTTCCTCCA&quot;
##  [310] &quot;AAGTCTGTCAACCATG&quot; &quot;AAGTCTGTCATAGCAC&quot; &quot;AAGTCTGTCCAAGCCG&quot;
##  [313] &quot;AAGTCTGTCTAACTTC&quot; &quot;AATCCAGAGCTGATAA&quot; &quot;AATCCAGAGGCTCAGA&quot;
##  [316] &quot;AATCCAGAGGGATCTG&quot; &quot;AATCCAGAGGTGTTAA&quot; &quot;AATCCAGCACATCTTT&quot;
##  [319] &quot;AATCCAGCACGAAACG&quot; &quot;AATCCAGCACGGATAG&quot; &quot;AATCCAGCAGACGCTC&quot;
##  [322] &quot;AATCCAGGTCAGAGGT&quot; &quot;AATCCAGGTCGGCTCA&quot; &quot;AATCCAGGTTCCCGAG&quot;
##  [325] &quot;AATCCAGGTTCGGGCT&quot; &quot;AATCCAGGTTGGTGGA&quot; &quot;AATCCAGTCAGAGCTT&quot;
##  [328] &quot;AATCCAGTCCATTCTA&quot; &quot;AATCCAGTCCTGCAGG&quot; &quot;AATCGGTAGCCACGTC&quot;
##  [331] &quot;AATCGGTAGGGATACC&quot; &quot;AATCGGTCAACGCACC&quot; &quot;AATCGGTCACCATCCT&quot;
##  [334] &quot;AATCGGTCAGCGAACA&quot; &quot;AATCGGTCATGTCCTC&quot; &quot;AATCGGTCATTATCTC&quot;
##  [337] &quot;AATCGGTGTCGGATCC&quot; &quot;AATCGGTGTCTGGTCG&quot; &quot;AATCGGTGTTTGCATG&quot;
##  [340] &quot;AATCGGTTCAAACGGG&quot; &quot;AATCGGTTCAACGCTA&quot; &quot;AATCGGTTCCACGCAG&quot;
##  [343] &quot;AATCGGTTCCCGGATG&quot; &quot;AATCGGTTCGTGGACC&quot; &quot;AATCGGTTCTACCAGA&quot;
##  [346] &quot;AATCGGTTCTAGCACA&quot; &quot;ACACCAAAGCGATATA&quot; &quot;ACACCAAAGCTCCCAG&quot;
##  [349] &quot;ACACCAAAGTGGTAGC&quot; &quot;ACACCAAAGTGTCCAT&quot; &quot;ACACCAACAAGTCATC&quot;
##  [352] &quot;ACACCAACACACGCTG&quot; &quot;ACACCAACACAGCCCA&quot; &quot;ACACCAACAGGTCTCG&quot;
##  [355] &quot;ACACCAACATCGGAAG&quot; &quot;ACACCAAGTCAAACTC&quot; &quot;ACACCAAGTCGGCTCA&quot;
##  [358] &quot;ACACCAAGTCTGGTCG&quot; &quot;ACACCAAGTGACAAAT&quot; &quot;ACACCAAGTGCTTCTC&quot;
##  [361] &quot;ACACCAAGTGGACGAT&quot; &quot;ACACCAATCATCGGAT&quot; &quot;ACACCAATCCGTAGTA&quot;
##  [364] &quot;ACACCAATCGTCTGAA&quot; &quot;ACACCAATCGTTGCCT&quot; &quot;ACACCCTAGACCTAGG&quot;
##  [367] &quot;ACACCCTAGCCAGTAG&quot; &quot;ACACCCTAGTAGCGGT&quot; &quot;ACACCCTCAACTGCTA&quot;
##  [370] &quot;ACACCCTCAGCCAGAA&quot; &quot;ACACCCTCATCGGTTA&quot; &quot;ACACCCTCATGCCACG&quot;
##  [373] &quot;ACACCCTCATGTCGAT&quot; &quot;ACACCCTCATTATCTC&quot; &quot;ACACCCTGTAGGACAC&quot;
##  [376] &quot;ACACCCTTCAAACAAG&quot; &quot;ACACCCTTCATATCGG&quot; &quot;ACACCCTTCATTATCC&quot;
##  [379] &quot;ACACCCTTCCAGGGCT&quot; &quot;ACACCCTTCCCTAATT&quot; &quot;ACACCCTTCCGTTGCT&quot;
##  [382] &quot;ACACCCTTCGCGCCAA&quot; &quot;ACACCCTTCGTAGGTT&quot; &quot;ACACCCTTCGTTACGA&quot;
##  [385] &quot;ACACCCTTCTGGGCCA&quot; &quot;ACACCGGAGAGCTATA&quot; &quot;ACACCGGAGCACCGCT&quot;
##  [388] &quot;ACACCGGAGGCATGGT&quot; &quot;ACACCGGAGGCTATCT&quot; &quot;ACACCGGAGTCATCCA&quot;
##  [391] &quot;ACACCGGAGTGAAGAG&quot; &quot;ACACCGGAGTGGGATC&quot; &quot;ACACCGGCAAGCCGTC&quot;
##  [394] &quot;ACACCGGCAGGTCTCG&quot; &quot;ACACCGGGTGACTACT&quot; &quot;ACACCGGGTTCGCTAA&quot;
##  [397] &quot;ACACCGGTCGCCTGTT&quot; &quot;ACACCGGTCTTAGAGC&quot; &quot;ACACTGAAGACAGACC&quot;
##  [400] &quot;ACACTGAAGTACTTGC&quot; &quot;ACACTGAAGTCTTGCA&quot; &quot;ACACTGACAACTGCGC&quot;
##  [403] &quot;ACACTGACAAGAGGCT&quot; &quot;ACACTGACAAGCCGCT&quot; &quot;ACACTGACACCCAGTG&quot;
##  [406] &quot;ACACTGACACTCAGGC&quot; &quot;ACACTGAGTACACCGC&quot; &quot;ACACTGAGTAGGGTAC&quot;
##  [409] &quot;ACACTGAGTCGCATAT&quot; &quot;ACACTGAGTCTCTCTG&quot; &quot;ACACTGAGTGGCAAAC&quot;
##  [412] &quot;ACACTGAGTGTCCTCT&quot; &quot;ACACTGAGTTGAGTTC&quot; &quot;ACACTGATCAGTGCAT&quot;
##  [415] &quot;ACAGCCGAGACCACGA&quot; &quot;ACAGCCGAGACTAGGC&quot; &quot;ACAGCCGAGCTACCGC&quot;
##  [418] &quot;ACAGCCGAGTTCGCGC&quot; &quot;ACAGCCGCAACGATCT&quot; &quot;ACAGCCGCACCAGGTC&quot;
##  [421] &quot;ACAGCCGCACCATCCT&quot; &quot;ACAGCCGCATGGGACA&quot; &quot;ACAGCCGGTACCGGCT&quot;
##  [424] &quot;ACAGCCGGTAGCGTAG&quot; &quot;ACAGCCGGTTCAGGCC&quot; &quot;ACAGCCGGTTCCACTC&quot;
##  [427] &quot;ACAGCCGGTTTCGCTC&quot; &quot;ACAGCCGTCAGGCCCA&quot; &quot;ACAGCCGTCCCATTAT&quot;
##  [430] &quot;ACAGCCGTCGTGGACC&quot; &quot;ACAGCCGTCTGATTCT&quot; &quot;ACAGCTAAGATGGCGT&quot;
##  [433] &quot;ACAGCTAAGCGTAGTG&quot; &quot;ACAGCTACAATAGCAA&quot; &quot;ACAGCTATCACATGCA&quot;
##  [436] &quot;ACAGCTATCCAACCAA&quot; &quot;ACAGCTATCGTAGGAG&quot; &quot;ACAGCTATCTCGGACG&quot;
##  [439] &quot;ACATACGAGAATTGTG&quot; &quot;ACATACGAGAGTCTGG&quot; &quot;ACATACGCAAGCCGTC&quot;
##  [442] &quot;ACATACGCAGACGTAG&quot; &quot;ACATACGCATCGGACC&quot; &quot;ACATACGGTCATTAGC&quot;
##  [445] &quot;ACATACGGTCGGCACT&quot; &quot;ACATACGGTCTAGTCA&quot; &quot;ACATACGTCATGCTCC&quot;
##  [448] &quot;ACATCAGAGTCCCACG&quot; &quot;ACATCAGAGTTATCGC&quot; &quot;ACATCAGCACACCGCA&quot;
##  [451] &quot;ACATCAGCACAGTCGC&quot; &quot;ACATCAGCACGCGAAA&quot; &quot;ACATCAGCATCCCACT&quot;
##  [454] &quot;ACATCAGCATTCTTAC&quot; &quot;ACATCAGCATTTGCTT&quot; &quot;ACATCAGGTCGAACAG&quot;
##  [457] &quot;ACATCAGGTCTCGTTC&quot; &quot;ACATCAGGTGAAAGAG&quot; &quot;ACATCAGGTGCATCTA&quot;
##  [460] &quot;ACATCAGTCACCGTAA&quot; &quot;ACATCAGTCAGTCAGT&quot; &quot;ACATCAGTCGAACGGA&quot;
##  [463] &quot;ACATCAGTCGACAGCC&quot; &quot;ACATCAGTCGCACTCT&quot; &quot;ACATCAGTCTTTACGT&quot;
##  [466] &quot;ACATGGTAGACTTGAA&quot; &quot;ACATGGTAGTAGTGCG&quot; &quot;ACATGGTCAAAGCGGT&quot;
##  [469] &quot;ACATGGTCAATAAGCA&quot; &quot;ACATGGTCAATGAATG&quot; &quot;ACATGGTCATGCAATC&quot;
##  [472] &quot;ACATGGTCATGGTCAT&quot; &quot;ACATGGTGTGTGCCTG&quot; &quot;ACATGGTGTTGATTCG&quot;
##  [475] &quot;ACATGGTTCAATCTCT&quot; &quot;ACATGGTTCCGCGTTT&quot; &quot;ACATGGTTCCTAAGTG&quot;
##  [478] &quot;ACATGGTTCGTACGGC&quot; &quot;ACATGGTTCGTTTATC&quot; &quot;ACATGGTTCTACTATC&quot;
##  [481] &quot;ACATGGTTCTGCCAGG&quot; &quot;ACCAGTAAGAGAGCTC&quot; &quot;ACCAGTACACGGTGTC&quot;
##  [484] &quot;ACCAGTACATGCAATC&quot; &quot;ACCAGTAGTCTCTCGT&quot; &quot;ACCAGTAGTGCCTGCA&quot;
##  [487] &quot;ACCAGTAGTTCCCGAG&quot; &quot;ACCAGTAGTTTGGCGC&quot; &quot;ACCAGTATCCAATGGT&quot;
##  [490] &quot;ACCAGTATCCATGAAC&quot; &quot;ACCCACTAGAATTCCC&quot; &quot;ACCCACTAGCGGATCA&quot;
##  [493] &quot;ACCCACTAGCTCCTCT&quot; &quot;ACCCACTAGGCTCATT&quot; &quot;ACCCACTCAAAGCAAT&quot;
##  [496] &quot;ACCCACTCAACACCCG&quot; &quot;ACCCACTCAGATCGGA&quot; &quot;ACCCACTCATGGTAGG&quot;
##  [499] &quot;ACCCACTGTCATTAGC&quot; &quot;ACCCACTGTCCTCTTG&quot; &quot;ACCCACTGTCTCTTAT&quot;
##  [502] &quot;ACCCACTGTGGCAAAC&quot; &quot;ACCCACTGTTCCGGCA&quot; &quot;ACCCACTGTTGTTTGG&quot;
##  [505] &quot;ACCCACTTCCTTGACC&quot; &quot;ACCCACTTCGGTGTTA&quot; &quot;ACCGTAAAGAGTGAGA&quot;
##  [508] &quot;ACCGTAAAGGTAAACT&quot; &quot;ACCGTAACACACGCTG&quot; &quot;ACCGTAACAGCTCGAC&quot;
##  [511] &quot;ACCGTAACAGCTGTAT&quot; &quot;ACCGTAACATCCTAGA&quot; &quot;ACCGTAACATGGGAAC&quot;
##  [514] &quot;ACCGTAACATTTCAGG&quot; &quot;ACCGTAAGTTCGCGAC&quot; &quot;ACCGTAATCCTATGTT&quot;
##  [517] &quot;ACCGTAATCGGTGTTA&quot; &quot;ACCGTAATCGTCACGG&quot; &quot;ACCGTAATCTGGCGTG&quot;
##  [520] &quot;ACCTTTAAGAGCTTCT&quot; &quot;ACCTTTAAGTTGTCGT&quot; &quot;ACCTTTACAAGCCCAC&quot;
##  [523] &quot;ACCTTTACATGTCCTC&quot; &quot;ACCTTTAGTAGAGTGC&quot; &quot;ACCTTTATCATCTGTT&quot;
##  [526] &quot;ACCTTTATCCCTCTTT&quot; &quot;ACCTTTATCCCTGACT&quot; &quot;ACCTTTATCCTGTAGA&quot;
##  [529] &quot;ACCTTTATCGCCAGCA&quot; &quot;ACCTTTATCTGCCCTA&quot; &quot;ACCTTTATCTTGACGA&quot;
##  [532] &quot;ACGAGCCAGAGCCTAG&quot; &quot;ACGAGCCAGGTGATAT&quot; &quot;ACGAGCCCAAGCCGCT&quot;
##  [535] &quot;ACGAGCCCAGGCAGTA&quot; &quot;ACGAGCCGTACCGCTG&quot; &quot;ACGAGCCGTAGAAAGG&quot;
##  [538] &quot;ACGAGCCGTTGTCGCG&quot; &quot;ACGAGCCGTTTCGCTC&quot; &quot;ACGAGCCTCACTTCAT&quot;
##  [541] &quot;ACGAGCCTCCGTAGGC&quot; &quot;ACGAGCCTCGTCTGAA&quot; &quot;ACGAGCCTCGTGGACC&quot;
##  [544] &quot;ACGAGCCTCTTCATGT&quot; &quot;ACGAGGAAGAAGAAGC&quot; &quot;ACGAGGAAGGTGATTA&quot;
##  [547] &quot;ACGAGGACACATAACC&quot; &quot;ACGAGGACATGGATGG&quot; &quot;ACGAGGAGTTATCACG&quot;
##  [550] &quot;ACGAGGAGTTCGTTGA&quot; &quot;ACGAGGATCCAGGGCT&quot; &quot;ACGAGGATCCCTGACT&quot;
##  [553] &quot;ACGAGGATCGAACTGT&quot; &quot;ACGATACAGAAAGTGG&quot; &quot;ACGATACAGACCTAGG&quot;
##  [556] &quot;ACGATACAGACTAGAT&quot; &quot;ACGATACCAAACAACA&quot; &quot;ACGATACCAAGCTGGA&quot;
##  [559] &quot;ACGATACCACAGGCCT&quot; &quot;ACGATACCAGCCTATA&quot; &quot;ACGATACGTATATCCG&quot;
##  [562] &quot;ACGATACGTCAATACC&quot; &quot;ACGATACGTCTTCTCG&quot; &quot;ACGATACGTGCCTTGG&quot;
##  [565] &quot;ACGATACGTGCGATAG&quot; &quot;ACGATACGTGGAAAGA&quot; &quot;ACGATACGTTCCAACA&quot;
##  [568] &quot;ACGATACTCACATGCA&quot; &quot;ACGATACTCAGCGATT&quot; &quot;ACGATACTCCCAGGTG&quot;
##  [571] &quot;ACGATGTAGAGGTACC&quot; &quot;ACGATGTAGATATGCA&quot; &quot;ACGATGTCAAGCGTAG&quot;
##  [574] &quot;ACGATGTCAATAGCAA&quot; &quot;ACGATGTCACCAACCG&quot; &quot;ACGATGTCAGGAACGT&quot;
##  [577] &quot;ACGATGTCATGTCGAT&quot; &quot;ACGATGTGTAGTAGTA&quot; &quot;ACGATGTGTGACTACT&quot;
##  [580] &quot;ACGATGTGTGCACTTA&quot; &quot;ACGATGTGTGCCTGGT&quot; &quot;ACGATGTGTGGTACAG&quot;
##  [583] &quot;ACGATGTGTTCAGGCC&quot; &quot;ACGATGTTCCACGAAT&quot; &quot;ACGATGTTCGGTCTAA&quot;
##  [586] &quot;ACGATGTTCGTCCGTT&quot; &quot;ACGATGTTCGTTTATC&quot; &quot;ACGATGTTCTCGCTTG&quot;
##  [589] &quot;ACGATGTTCTGACCTC&quot; &quot;ACGATGTTCTGTCAAG&quot; &quot;ACGATGTTCTTACCGC&quot;
##  [592] &quot;ACGCAGCAGATCTGAA&quot; &quot;ACGCAGCAGATGTGGC&quot; &quot;ACGCAGCAGTGTCTCA&quot;
##  [595] &quot;ACGCAGCCAACGATGG&quot; &quot;ACGCAGCCACACGCTG&quot; &quot;ACGCAGCCAGCTGTTA&quot;
##  [598] &quot;ACGCAGCCAGGGTTAG&quot; &quot;ACGCAGCGTACTCAAC&quot; &quot;ACGCAGCGTAGGAGTC&quot;
##  [601] &quot;ACGCAGCGTATATCCG&quot; &quot;ACGCAGCGTCACCTAA&quot; &quot;ACGCAGCGTCCTCCAT&quot;
##  [604] &quot;ACGCAGCGTCTTCTCG&quot; &quot;ACGCAGCGTGTGGTTT&quot; &quot;ACGCAGCGTTCCTCCA&quot;
##  [607] &quot;ACGCAGCGTTGTCTTT&quot; &quot;ACGCAGCTCAGGATCT&quot; &quot;ACGCAGCTCGACGGAA&quot;
##  [610] &quot;ACGCAGCTCGCCAAAT&quot; &quot;ACGCAGCTCTGATACG&quot; &quot;ACGCCAGAGAACAACT&quot;
##  [613] &quot;ACGCCAGAGATGGGTC&quot; &quot;ACGCCAGAGATGTAAC&quot; &quot;ACGCCAGAGCACACAG&quot;
##  [616] &quot;ACGCCAGAGGATGTAT&quot; &quot;ACGCCAGAGTAGGTGC&quot; &quot;ACGCCAGAGTCAAGCG&quot;
##  [619] &quot;ACGCCAGAGTCTCAAC&quot; &quot;ACGCCAGCAATGCCAT&quot; &quot;ACGCCAGCACACCGCA&quot;
##  [622] &quot;ACGCCAGCACGTGAGA&quot; &quot;ACGCCAGCAGCGTCCA&quot; &quot;ACGCCAGGTCAATGTC&quot;
##  [625] &quot;ACGCCAGGTCCGAGTC&quot; &quot;ACGCCAGGTCGACTAT&quot; &quot;ACGCCAGGTTCGCTAA&quot;
##  [628] &quot;ACGCCAGTCACTCCTG&quot; &quot;ACGCCAGTCAGAGCTT&quot; &quot;ACGCCAGTCATATCGG&quot;
##  [631] &quot;ACGCCAGTCCCACTTG&quot; &quot;ACGCCAGTCCTGCCAT&quot; &quot;ACGCCAGTCCTTTCGG&quot;
##  [634] &quot;ACGCCAGTCGCAGGCT&quot; &quot;ACGCCAGTCGTTTATC&quot; &quot;ACGCCGAAGACTAGGC&quot;
##  [637] &quot;ACGCCGAAGAGGTTAT&quot; &quot;ACGCCGAAGTGTTAGA&quot; &quot;ACGCCGAAGTTACGGG&quot;
##  [640] &quot;ACGCCGACAATGTAAG&quot; &quot;ACGCCGACAGACGCCT&quot; &quot;ACGCCGACAGGTTTCA&quot;
##  [643] &quot;ACGCCGACATCGGTTA&quot; &quot;ACGCCGAGTCTAACGT&quot; &quot;ACGCCGAGTCTCTTTA&quot;
##  [646] &quot;ACGCCGAGTCTTCTCG&quot; &quot;ACGCCGAGTGTAATGA&quot; &quot;ACGCCGAGTTGCGCAC&quot;
##  [649] &quot;ACGCCGATCAGGTAAA&quot; &quot;ACGCCGATCATGGTCA&quot; &quot;ACGCCGATCCTACAGA&quot;
##  [652] &quot;ACGCCGATCTCGATGA&quot; &quot;ACGCCGATCTTAGAGC&quot; &quot;ACGCCGATCTTCAACT&quot;
##  [655] &quot;ACGGAGAAGAGGACGG&quot; &quot;ACGGAGAAGTTCGCGC&quot; &quot;ACGGAGACAAGTTAAG&quot;
##  [658] &quot;ACGGAGACAAGTTCTG&quot; &quot;ACGGAGACAATCTACG&quot; &quot;ACGGAGACAGACTCGC&quot;
##  [661] &quot;ACGGAGAGTACACCGC&quot; &quot;ACGGAGAGTACTTGAC&quot; &quot;ACGGAGAGTGCTCTTC&quot;
##  [664] &quot;ACGGAGAGTGGGTCAA&quot; &quot;ACGGAGAGTGTGACGA&quot; &quot;ACGGAGAGTTCCCGAG&quot;
##  [667] &quot;ACGGAGAGTTCGGCAC&quot; &quot;ACGGAGATCAAACAAG&quot; &quot;ACGGAGATCACATACG&quot;
##  [670] &quot;ACGGAGATCAGCGACC&quot; &quot;ACGGAGATCATCTGCC&quot; &quot;ACGGAGATCGAACGGA&quot;
##  [673] &quot;ACGGCCAAGACCTAGG&quot; &quot;ACGGCCAAGCGAAGGG&quot; &quot;ACGGCCAAGCTGAAAT&quot;
##  [676] &quot;ACGGCCAAGGATCGCA&quot; &quot;ACGGCCACAACGATCT&quot; &quot;ACGGCCACACACTGCG&quot;
##  [679] &quot;ACGGCCACACGAGAGT&quot; &quot;ACGGCCACACGGATAG&quot; &quot;ACGGCCACACTCTGTC&quot;
##  [682] &quot;ACGGCCACAGTAAGAT&quot; &quot;ACGGCCAGTCAACATC&quot; &quot;ACGGCCAGTGAGTGAC&quot;
##  [685] &quot;ACGGCCATCACTTATC&quot; &quot;ACGGCCATCATAAAGG&quot; &quot;ACGGCCATCCAAAGTC&quot;
##  [688] &quot;ACGGCCATCCGAGCCA&quot; &quot;ACGGCCATCGGAATCT&quot; &quot;ACGGCCATCGTCACGG&quot;
##  [691] &quot;ACGGCCATCTGCCCTA&quot; &quot;ACGGGCTAGCCCAACC&quot; &quot;ACGGGCTAGGGTATCG&quot;
##  [694] &quot;ACGGGCTAGGGTGTTG&quot; &quot;ACGGGCTCACACCGAC&quot; &quot;ACGGGCTCACCCATTC&quot;
##  [697] &quot;ACGGGCTCATGTAGTC&quot; &quot;ACGGGCTCATGTCGAT&quot; &quot;ACGGGCTGTTCAACCA&quot;
##  [700] &quot;ACGGGCTTCAACCAAC&quot; &quot;ACGGGCTTCAACCATG&quot; &quot;ACGGGCTTCTCTTGAT&quot;
##  [703] &quot;ACGGGTCAGATGCCAG&quot; &quot;ACGGGTCAGCACACAG&quot; &quot;ACGGGTCCAGCTGTGC&quot;
##  [706] &quot;ACGGGTCGTATAGGTA&quot; &quot;ACGGGTCGTTGGAGGT&quot; &quot;ACGGGTCTCATGCATG&quot;
##  [709] &quot;ACGGGTCTCATGGTCA&quot; &quot;ACGGGTCTCTCAAACG&quot; &quot;ACGGGTCTCTGGCGTG&quot;
##  [712] &quot;ACGTCAAAGCAATCTC&quot; &quot;ACGTCAAAGTCTTGCA&quot; &quot;ACGTCAAAGTGGGTTG&quot;
##  [715] &quot;ACGTCAACAATCGGTT&quot; &quot;ACGTCAAGTCGAATCT&quot; &quot;ACGTCAAGTGAGTGAC&quot;
##  [718] &quot;ACGTCAAGTGCACCAC&quot; &quot;ACGTCAATCACGAAGG&quot; &quot;ACGTCAATCAGTGCAT&quot;
##  [721] &quot;ACGTCAATCAGTTGAC&quot; &quot;ACGTCAATCATCGGAT&quot; &quot;ACGTCAATCCAAAGTC&quot;
##  [724] &quot;ACGTCAATCCGAAGAG&quot; &quot;ACGTCAATCTAACTTC&quot; &quot;ACGTCAATCTGATACG&quot;
##  [727] &quot;ACGTCAATCTGCCCTA&quot; &quot;ACTATCTAGCAACGGT&quot; &quot;ACTATCTAGTAGCGGT&quot;
##  [730] &quot;ACTATCTCAATGCCAT&quot; &quot;ACTATCTCAGATAATG&quot; &quot;ACTATCTCATAGACTC&quot;
##  [733] &quot;ACTATCTCATTCGACA&quot; &quot;ACTATCTGTGTAATGA&quot; &quot;ACTATCTGTTGGTGGA&quot;
##  [736] &quot;ACTATCTTCACGCATA&quot; &quot;ACTATCTTCAGCACAT&quot; &quot;ACTATCTTCATGTCTT&quot;
##  [739] &quot;ACTGAACAGAAGGACA&quot; &quot;ACTGAACAGGGTCTCC&quot; &quot;ACTGAACCAATGAAAC&quot;
##  [742] &quot;ACTGAACCACGCCAGT&quot; &quot;ACTGAACCACTTCTGC&quot; &quot;ACTGAACCAGCTGCAC&quot;
##  [745] &quot;ACTGAACCAGGCTCAC&quot; &quot;ACTGAACGTACTTAGC&quot; &quot;ACTGAACGTATGGTTC&quot;
##  [748] &quot;ACTGAACGTTCGCGAC&quot; &quot;ACTGAACGTTTGCATG&quot; &quot;ACTGAACTCACAAACC&quot;
##  [751] &quot;ACTGAACTCCATGCTC&quot; &quot;ACTGAGTAGAAGAAGC&quot; &quot;ACTGAGTAGCACCGCT&quot;
##  [754] &quot;ACTGAGTAGGAGTACC&quot; &quot;ACTGAGTAGGCATGGT&quot; &quot;ACTGAGTCAACTGGCC&quot;
##  [757] &quot;ACTGAGTCACACCGAC&quot; &quot;ACTGAGTCATAGAAAC&quot; &quot;ACTGAGTCATATGGTC&quot;
##  [760] &quot;ACTGAGTCATCCGTGG&quot; &quot;ACTGAGTGTCAACTGT&quot; &quot;ACTGAGTGTCAGAAGC&quot;
##  [763] &quot;ACTGAGTGTGGACGAT&quot; &quot;ACTGAGTGTGTGACCC&quot; &quot;ACTGAGTTCGGTCCGA&quot;
##  [766] &quot;ACTGATGAGAAACGAG&quot; &quot;ACTGATGAGACAAGCC&quot; &quot;ACTGATGAGCGATGAC&quot;
##  [769] &quot;ACTGATGAGGTAAACT&quot; &quot;ACTGATGAGTAACCCT&quot; &quot;ACTGATGCACTCGACG&quot;
##  [772] &quot;ACTGATGCAGGTCTCG&quot; &quot;ACTGATGGTCTCTCGT&quot; &quot;ACTGATGGTTCACCTC&quot;
##  [775] &quot;ACTGATGGTTCTGAAC&quot; &quot;ACTGATGTCCAGGGCT&quot; &quot;ACTGATGTCCGCGGTA&quot;
##  [778] &quot;ACTGATGTCTCGCATC&quot; &quot;ACTGCTCAGTTACGGG&quot; &quot;ACTGCTCCAAACTGTC&quot;
##  [781] &quot;ACTGCTCCAAGTTCTG&quot; &quot;ACTGCTCCACCTCGTT&quot; &quot;ACTGCTCCACTTGGAT&quot;
##  [784] &quot;ACTGCTCCAGCAGTTT&quot; &quot;ACTGCTCGTGGTCCGT&quot; &quot;ACTGCTCTCAGAGGTG&quot;
##  [787] &quot;ACTGCTCTCCCGGATG&quot; &quot;ACTGCTCTCCGCGCAA&quot; &quot;ACTGCTCTCTGAGTGT&quot;
##  [790] &quot;ACTGTCCAGACAGACC&quot; &quot;ACTGTCCAGAGATGAG&quot; &quot;ACTGTCCAGGGAACGG&quot;
##  [793] &quot;ACTGTCCAGTTGTCGT&quot; &quot;ACTGTCCCACCTATCC&quot; &quot;ACTGTCCCAGCATGAG&quot;
##  [796] &quot;ACTGTCCGTCAGCTAT&quot; &quot;ACTGTCCGTCCGTGAC&quot; &quot;ACTGTCCTCATTGCCC&quot;
##  [799] &quot;ACTGTCCTCGAACTGT&quot; &quot;ACTGTCCTCTAACTGG&quot; &quot;ACTTACTAGAGGTAGA&quot;
##  [802] &quot;ACTTACTAGTCTTGCA&quot; &quot;ACTTACTAGTGTCCAT&quot; &quot;ACTTACTCAAGCTGAG&quot;
##  [805] &quot;ACTTACTCACATCCAA&quot; &quot;ACTTACTCACGGTAAG&quot; &quot;ACTTACTGTAAACGCG&quot;
##  [808] &quot;ACTTACTGTAGCCTAT&quot; &quot;ACTTACTGTCAAAGAT&quot; &quot;ACTTACTGTTTGACAC&quot;
##  [811] &quot;ACTTACTTCACAGTAC&quot; &quot;ACTTACTTCAGGCGAA&quot; &quot;ACTTACTTCATCACCC&quot;
##  [814] &quot;ACTTGTTAGCGCTCCA&quot; &quot;ACTTGTTAGGCACATG&quot; &quot;ACTTGTTAGGTGCAAC&quot;
##  [817] &quot;ACTTGTTCAATGGTCT&quot; &quot;ACTTGTTCACGCTTTC&quot; &quot;ACTTGTTCACGTCTCT&quot;
##  [820] &quot;ACTTGTTCAGACGCTC&quot; &quot;ACTTGTTCAGTCAGAG&quot; &quot;ACTTGTTCATCCCACT&quot;
##  [823] &quot;ACTTGTTCATCCCATC&quot; &quot;ACTTGTTCATGAACCT&quot; &quot;ACTTGTTGTCCGAGTC&quot;
##  [826] &quot;ACTTGTTGTCGAATCT&quot; &quot;ACTTGTTGTCGAGATG&quot; &quot;ACTTGTTGTCTCTCTG&quot;
##  [829] &quot;ACTTGTTGTGCTGTAT&quot; &quot;ACTTGTTGTTGGGACA&quot; &quot;ACTTGTTGTTGTCTTT&quot;
##  [832] &quot;ACTTGTTTCCACTCCA&quot; &quot;ACTTTCAAGCAGCGTA&quot; &quot;ACTTTCAAGCGTTTAC&quot;
##  [835] &quot;ACTTTCAAGCTAGGCA&quot; &quot;ACTTTCAAGCTCCTTC&quot; &quot;ACTTTCACAAACGCGA&quot;
##  [838] &quot;ACTTTCACAGACTCGC&quot; &quot;ACTTTCACAGTCAGCC&quot; &quot;ACTTTCACATTGTGCA&quot;
##  [841] &quot;ACTTTCAGTACTTAGC&quot; &quot;ACTTTCAGTACTTCTT&quot; &quot;ACTTTCAGTCTAAACC&quot;
##  [844] &quot;ACTTTCAGTGTAACGG&quot; &quot;ACTTTCATCAACACGT&quot; &quot;ACTTTCATCCACGTTC&quot;
##  [847] &quot;AGAATAGAGATGTTAG&quot; &quot;AGAATAGAGCCAGGAT&quot; &quot;AGAATAGAGCTCAACT&quot;
##  [850] &quot;AGAATAGAGGAGTACC&quot; &quot;AGAATAGAGGTAAACT&quot; &quot;AGAATAGAGGTCGGAT&quot;
##  [853] &quot;AGAATAGAGTGGGTTG&quot; &quot;AGAATAGCAACGCACC&quot; &quot;AGAATAGCAAGGTGTG&quot;
##  [856] &quot;AGAATAGCACATGACT&quot; &quot;AGAATAGCACCCAGTG&quot; &quot;AGAATAGCATAAAGGT&quot;
##  [859] &quot;AGAATAGCATAGAAAC&quot; &quot;AGAATAGCATGCCTTC&quot; &quot;AGAATAGGTACTTAGC&quot;
##  [862] &quot;AGAATAGGTTCAGACT&quot; &quot;AGAATAGTCACCTCGT&quot; &quot;AGAATAGTCAGCTCTC&quot;
##  [865] &quot;AGAATAGTCCTATTCA&quot; &quot;AGACGTTAGACTGGGT&quot; &quot;AGACGTTAGGCCCTCA&quot;
##  [868] &quot;AGACGTTAGGTAAACT&quot; &quot;AGACGTTAGTGTTTGC&quot; &quot;AGACGTTCAGCTCCGA&quot;
##  [871] &quot;AGACGTTGTAGGACAC&quot; &quot;AGACGTTGTGGTACAG&quot; &quot;AGACGTTTCAGTTAGC&quot;
##  [874] &quot;AGACGTTTCCGTACAA&quot; &quot;AGACGTTTCTGCAAGT&quot; &quot;AGACGTTTCTTCCTTC&quot;
##  [877] &quot;AGAGCGAAGATCCCGC&quot; &quot;AGAGCGAAGATCCTGT&quot; &quot;AGAGCGAAGCAGATCG&quot;
##  [880] &quot;AGAGCGAAGTCCGGTC&quot; &quot;AGAGCGACAAAGCAAT&quot; &quot;AGAGCGACAATGGAAT&quot;
##  [883] &quot;AGAGCGACATATGCTG&quot; &quot;AGAGCGACATCACAAC&quot; &quot;AGAGCGACATGGAATA&quot;
##  [886] &quot;AGAGCGAGTACGACCC&quot; &quot;AGAGCGAGTCATGCAT&quot; &quot;AGAGCGAGTTGTACAC&quot;
##  [889] &quot;AGAGCGATCACCGTAA&quot; &quot;AGAGCGATCAGCTTAG&quot; &quot;AGAGCGATCGGCGGTT&quot;
##  [892] &quot;AGAGCGATCGGCTTGG&quot; &quot;AGAGCGATCTGTCTCG&quot; &quot;AGAGCTTAGACTAGGC&quot;
##  [895] &quot;AGAGCTTAGCCACGTC&quot; &quot;AGAGCTTAGTCGTTTG&quot; &quot;AGAGCTTAGTGGGCTA&quot;
##  [898] &quot;AGAGCTTCAAGTACCT&quot; &quot;AGAGCTTCACAGCCCA&quot; &quot;AGAGCTTCACCCAGTG&quot;
##  [901] &quot;AGAGCTTCACGCTTTC&quot; &quot;AGAGCTTCAGTATCTG&quot; &quot;AGAGCTTCATAGTAAG&quot;
##  [904] &quot;AGAGCTTGTCTAAACC&quot; &quot;AGAGCTTGTGCTAGCC&quot; &quot;AGAGCTTGTGGCAAAC&quot;
##  [907] &quot;AGAGCTTGTGGTACAG&quot; &quot;AGAGCTTTCACTATTC&quot; &quot;AGAGCTTTCAGAAATG&quot;
##  [910] &quot;AGAGCTTTCAGCACAT&quot; &quot;AGAGCTTTCCTCGCAT&quot; &quot;AGAGTGGAGAGGTAGA&quot;
##  [913] &quot;AGAGTGGAGCCCGAAA&quot; &quot;AGAGTGGAGTCATGCT&quot; &quot;AGAGTGGAGTGGTCCC&quot;
##  [916] &quot;AGAGTGGAGTGTACGG&quot; &quot;AGAGTGGCACATAACC&quot; &quot;AGAGTGGCACCGAAAG&quot;
##  [919] &quot;AGAGTGGCAGCGTTCG&quot; &quot;AGAGTGGCAGGACCCT&quot; &quot;AGAGTGGCATCGATTG&quot;
##  [922] &quot;AGAGTGGGTAGTGAAT&quot; &quot;AGAGTGGGTATAATGG&quot; &quot;AGAGTGGTCCCGACTT&quot;
##  [925] &quot;AGAGTGGTCTACTTAC&quot; &quot;AGAGTGGTCTGTCAAG&quot; &quot;AGATCTGAGAGGTAGA&quot;
##  [928] &quot;AGATCTGAGCAGGTCA&quot; &quot;AGATCTGAGCCAACAG&quot; &quot;AGATCTGAGGATCGCA&quot;
##  [931] &quot;AGATCTGAGGGATCTG&quot; &quot;AGATCTGAGGGCTCTC&quot; &quot;AGATCTGAGGTGATAT&quot;
##  [934] &quot;AGATCTGAGTACGTAA&quot; &quot;AGATCTGAGTTCGCAT&quot; &quot;AGATCTGCAAACAACA&quot;
##  [937] &quot;AGATCTGCACATGACT&quot; &quot;AGATCTGGTAAGAGGA&quot; &quot;AGATCTGTCAGTTCGA&quot;
##  [940] &quot;AGATCTGTCGCCATAA&quot; &quot;AGATCTGTCGCTTAGA&quot; &quot;AGATTGCAGAAACCAT&quot;
##  [943] &quot;AGATTGCAGAGGTACC&quot; &quot;AGATTGCAGATAGTCA&quot; &quot;AGATTGCCACGGTTTA&quot;
##  [946] &quot;AGATTGCCAGGGTATG&quot; &quot;AGATTGCCATCACGTA&quot; &quot;AGATTGCGTACATGTC&quot;
##  [949] &quot;AGATTGCGTCCAGTAT&quot; &quot;AGATTGCGTCCGTGAC&quot; &quot;AGATTGCTCAGTGCAT&quot;
##  [952] &quot;AGATTGCTCATGCATG&quot; &quot;AGATTGCTCGGCGGTT&quot; &quot;AGCAGCCAGAAACCAT&quot;
##  [955] &quot;AGCAGCCAGACCTTTG&quot; &quot;AGCAGCCAGACGACGT&quot; &quot;AGCAGCCAGATGTTAG&quot;
##  [958] &quot;AGCAGCCAGCACCGTC&quot; &quot;AGCAGCCAGTGGTAAT&quot; &quot;AGCAGCCCAAACCCAT&quot;
##  [961] &quot;AGCAGCCCAGATGGCA&quot; &quot;AGCAGCCCATCACGAT&quot; &quot;AGCAGCCGTACTTAGC&quot;
##  [964] &quot;AGCAGCCGTCCGAAGA&quot; &quot;AGCAGCCGTCGCTTCT&quot; &quot;AGCAGCCGTGCAGTAG&quot;
##  [967] &quot;AGCAGCCGTTCCCGAG&quot; &quot;AGCAGCCGTTTAGGAA&quot; &quot;AGCAGCCGTTTGTTTC&quot;
##  [970] &quot;AGCAGCCTCCTTGCCA&quot; &quot;AGCAGCCTCGTCGTTC&quot; &quot;AGCAGCCTCTGGAGCC&quot;
##  [973] &quot;AGCATACAGTTCGATC&quot; &quot;AGCATACCACCACGTG&quot; &quot;AGCATACCACGGATAG&quot;
##  [976] &quot;AGCATACCAGCTATTG&quot; &quot;AGCATACGTCCTCTTG&quot; &quot;AGCATACGTTAAGGGC&quot;
##  [979] &quot;AGCATACGTTACTGAC&quot; &quot;AGCATACGTTAGATGA&quot; &quot;AGCATACTCATCGCTC&quot;
##  [982] &quot;AGCATACTCCGCATCT&quot; &quot;AGCATACTCGAACTGT&quot; &quot;AGCATACTCTAACTTC&quot;
##  [985] &quot;AGCATACTCTCGGACG&quot; &quot;AGCCTAAAGCCCAATT&quot; &quot;AGCCTAAAGGACATTA&quot;
##  [988] &quot;AGCCTAAAGTACGTTC&quot; &quot;AGCCTAAAGTCGATAA&quot; &quot;AGCCTAAAGTCTTGCA&quot;
##  [991] &quot;AGCCTAACAAGCCTAT&quot; &quot;AGCCTAACACGTCTCT&quot; &quot;AGCCTAACAGCCTGTG&quot;
##  [994] &quot;AGCCTAACATCAGTCA&quot; &quot;AGCCTAAGTATCTGCA&quot; &quot;AGCCTAAGTCGGCTCA&quot;
##  [997] &quot;AGCCTAAGTTATCCGA&quot; &quot;AGCCTAATCAGTACGT&quot; &quot;AGCCTAATCCCTTGTG&quot;
## [1000] &quot;AGCCTAATCGTTTGCC&quot;
##  [ reached getOption(&quot;max.print&quot;) -- omitted 6109 entries ]</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="data-wrangling-scrnaseq-1.html#cb50-1"></a><span class="co"># One can pull multiple values from the data frame at any time</span></span>
<span id="cb50-2"><a href="data-wrangling-scrnaseq-1.html#cb50-2"></a><span class="kw">head</span>(<span class="dt">x =</span> seurat<span class="op">@</span>meta.data[<span class="kw">c</span>(<span class="st">&#39;nCount_RNA&#39;</span>, <span class="st">&#39;nFeature_RNA&#39;</span>)])</span></code></pre></div>
<pre><code>##                  nCount_RNA nFeature_RNA
## AAACCTGAGCTAGTCT       3605         1184
## AAACCTGAGGGCACTA       3828         1387
## AAACCTGAGTACGTTC       6457         1784
## AAACCTGAGTCCGGTC       3075         1092
## AAACCTGCACCAGGTC       9399         2625
## AAACCTGCACCTCGTT      48838         5602</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="data-wrangling-scrnaseq-1.html#cb52-1"></a><span class="co"># The $ sigil can only pull bit of meta data at a time; however, tab-autocompletion</span></span>
<span id="cb52-2"><a href="data-wrangling-scrnaseq-1.html#cb52-2"></a><span class="co"># has been enabled for the $ sigil, making it ideal for interactive use</span></span>
<span id="cb52-3"><a href="data-wrangling-scrnaseq-1.html#cb52-3"></a></span>
<span id="cb52-4"><a href="data-wrangling-scrnaseq-1.html#cb52-4"></a><span class="kw">head</span>(<span class="dt">x =</span> seurat<span class="op">@</span>meta.data<span class="op">$</span>nFeature_RNA)</span></code></pre></div>
<pre><code>## [1] 1184 1387 1784 1092 2625 5602</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="data-wrangling-scrnaseq-1.html#cb54-1"></a><span class="co"># Passing `drop = TRUE` will turn the meta data into a names vector</span></span>
<span id="cb54-2"><a href="data-wrangling-scrnaseq-1.html#cb54-2"></a><span class="co"># with each entry being named for the cell it corresponds to</span></span>
<span id="cb54-3"><a href="data-wrangling-scrnaseq-1.html#cb54-3"></a><span class="kw">head</span>(<span class="dt">x =</span> seurat<span class="op">@</span>meta.data[<span class="st">&#39;nFeature_RNA&#39;</span>, <span class="dt">drop =</span> <span class="ot">TRUE</span>])</span></code></pre></div>
<pre><code>## Warning in `[.data.frame`(seurat@meta.data, &quot;nFeature_RNA&quot;, drop =
## TRUE): &#39;drop&#39; argument will be ignored</code></pre>
<pre><code>##                  nFeature_RNA
## AAACCTGAGCTAGTCT         1184
## AAACCTGAGGGCACTA         1387
## AAACCTGAGTACGTTC         1784
## AAACCTGAGTCCGGTC         1092
## AAACCTGCACCAGGTC         2625
## AAACCTGCACCTCGTT         5602</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="data-wrangling-scrnaseq-1.html#cb57-1"></a><span class="co"># Add meta data example</span></span></code></pre></div>
</div>
<div id="methods" class="section level4">
<h4><span class="header-section-number">4.4.1.4</span> Methods</h4>
<p>Methods for the Seurat class can be found with the following:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="data-wrangling-scrnaseq-1.html#cb58-1"></a>utils<span class="op">::</span><span class="kw">methods</span>(<span class="dt">class =</span> <span class="st">&#39;Seurat&#39;</span>)</span></code></pre></div>
<pre><code>##  [1] [                       [[                     
##  [3] [[&lt;-                    $                      
##  [5] $&lt;-                     AddMetaData            
##  [7] as.CellDataSet          as.loom                
##  [9] as.SingleCellExperiment colMeans               
## [11] colSums                 Command                
## [13] DefaultAssay            DefaultAssay&lt;-         
## [15] dim                     dimnames               
## [17] droplevels              Embeddings             
## [19] FindClusters            FindMarkers            
## [21] FindNeighbors           FindVariableFeatures   
## [23] GetAssay                GetAssayData           
## [25] HVFInfo                 Idents                 
## [27] Idents&lt;-                Key                    
## [29] levels                  levels&lt;-               
## [31] Loadings                merge                  
## [33] Misc                    Misc&lt;-                 
## [35] names                   NormalizeData          
## [37] OldWhichCells           Project                
## [39] Project&lt;-               RenameCells            
## [41] RenameIdents            ReorderIdent           
## [43] rowMeans                rowSums                
## [45] RunALRA                 RunCCA                 
## [47] RunICA                  RunLSI                 
## [49] RunPCA                  RunTSNE                
## [51] RunUMAP                 ScaleData              
## [53] ScoreJackStraw          SetAssayData           
## [55] SetIdent                show                   
## [57] StashIdent              Stdev                  
## [59] subset                  SubsetData             
## [61] Tool                    Tool&lt;-                 
## [63] VariableFeatures        VariableFeatures&lt;-     
## [65] WhichCells             
## see &#39;?methods&#39; for accessing help and source code</code></pre>
<p>You can also list all functions in the Seurat package by using:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="data-wrangling-scrnaseq-1.html#cb60-1"></a><span class="kw">ls</span>(<span class="st">&quot;package:Seurat&quot;</span>)</span></code></pre></div>
<pre><code>##   [1] &quot;AddMetaData&quot;                    
##   [2] &quot;AddModuleScore&quot;                 
##   [3] &quot;ALRAChooseKPlot&quot;                
##   [4] &quot;as.CellDataSet&quot;                 
##   [5] &quot;as.Graph&quot;                       
##   [6] &quot;as.loom&quot;                        
##   [7] &quot;as.Seurat&quot;                      
##   [8] &quot;as.SingleCellExperiment&quot;        
##   [9] &quot;as.sparse&quot;                      
##  [10] &quot;Assays&quot;                         
##  [11] &quot;AugmentPlot&quot;                    
##  [12] &quot;AverageExpression&quot;              
##  [13] &quot;BarcodeInflectionsPlot&quot;         
##  [14] &quot;BlackAndWhite&quot;                  
##  [15] &quot;BlueAndRed&quot;                     
##  [16] &quot;BoldTitle&quot;                      
##  [17] &quot;BuildClusterTree&quot;               
##  [18] &quot;CalculateBarcodeInflections&quot;    
##  [19] &quot;CaseMatch&quot;                      
##  [20] &quot;cc.genes&quot;                       
##  [21] &quot;cc.genes.updated.2019&quot;          
##  [22] &quot;CellCycleScoring&quot;               
##  [23] &quot;Cells&quot;                          
##  [24] &quot;CellsByIdentities&quot;              
##  [25] &quot;CellScatter&quot;                    
##  [26] &quot;CellSelector&quot;                   
##  [27] &quot;CollapseEmbeddingOutliers&quot;      
##  [28] &quot;CollapseSpeciesExpressionMatrix&quot;
##  [29] &quot;ColorDimSplit&quot;                  
##  [30] &quot;CombinePlots&quot;                   
##  [31] &quot;Command&quot;                        
##  [32] &quot;CreateAssayObject&quot;              
##  [33] &quot;CreateDimReducObject&quot;           
##  [34] &quot;CreateGeneActivityMatrix&quot;       
##  [35] &quot;CreateSeuratObject&quot;             
##  [36] &quot;CustomDistance&quot;                 
##  [37] &quot;CustomPalette&quot;                  
##  [38] &quot;DarkTheme&quot;                      
##  [39] &quot;DefaultAssay&quot;                   
##  [40] &quot;DefaultAssay&lt;-&quot;                 
##  [41] &quot;DietSeurat&quot;                     
##  [42] &quot;DimHeatmap&quot;                     
##  [43] &quot;DimPlot&quot;                        
##  [44] &quot;DoHeatmap&quot;                      
##  [45] &quot;DotPlot&quot;                        
##  [46] &quot;ElbowPlot&quot;                      
##  [47] &quot;Embeddings&quot;                     
##  [48] &quot;ExpMean&quot;                        
##  [49] &quot;ExportToCellbrowser&quot;            
##  [50] &quot;ExpSD&quot;                          
##  [51] &quot;ExpVar&quot;                         
##  [52] &quot;FeatureLocator&quot;                 
##  [53] &quot;FeaturePlot&quot;                    
##  [54] &quot;FeatureScatter&quot;                 
##  [55] &quot;FetchData&quot;                      
##  [56] &quot;FindAllMarkers&quot;                 
##  [57] &quot;FindClusters&quot;                   
##  [58] &quot;FindConservedMarkers&quot;           
##  [59] &quot;FindIntegrationAnchors&quot;         
##  [60] &quot;FindMarkers&quot;                    
##  [61] &quot;FindNeighbors&quot;                  
##  [62] &quot;FindTransferAnchors&quot;            
##  [63] &quot;FindVariableFeatures&quot;           
##  [64] &quot;FontSize&quot;                       
##  [65] &quot;GeneSymbolThesarus&quot;             
##  [66] &quot;GetAssay&quot;                       
##  [67] &quot;GetAssayData&quot;                   
##  [68] &quot;GetIntegrationData&quot;             
##  [69] &quot;GetResidual&quot;                    
##  [70] &quot;HoverLocator&quot;                   
##  [71] &quot;HTODemux&quot;                       
##  [72] &quot;HTOHeatmap&quot;                     
##  [73] &quot;HVFInfo&quot;                        
##  [74] &quot;Idents&quot;                         
##  [75] &quot;Idents&lt;-&quot;                       
##  [76] &quot;IntegrateData&quot;                  
##  [77] &quot;IsGlobal&quot;                       
##  [78] &quot;JackStraw&quot;                      
##  [79] &quot;JackStrawPlot&quot;                  
##  [80] &quot;JS&quot;                             
##  [81] &quot;JS&lt;-&quot;                           
##  [82] &quot;Key&quot;                            
##  [83] &quot;Key&lt;-&quot;                          
##  [84] &quot;L2CCA&quot;                          
##  [85] &quot;L2Dim&quot;                          
##  [86] &quot;LabelClusters&quot;                  
##  [87] &quot;LabelPoints&quot;                    
##  [88] &quot;Loadings&quot;                       
##  [89] &quot;Loadings&lt;-&quot;                     
##  [90] &quot;LocalStruct&quot;                    
##  [91] &quot;LogNormalize&quot;                   
##  [92] &quot;LogSeuratCommand&quot;               
##  [93] &quot;LogVMR&quot;                         
##  [94] &quot;MetaFeature&quot;                    
##  [95] &quot;MinMax&quot;                         
##  [96] &quot;Misc&quot;                           
##  [97] &quot;Misc&lt;-&quot;                         
##  [98] &quot;MixingMetric&quot;                   
##  [99] &quot;MULTIseqDemux&quot;                  
## [100] &quot;NoAxes&quot;                         
## [101] &quot;NoGrid&quot;                         
## [102] &quot;NoLegend&quot;                       
## [103] &quot;NormalizeData&quot;                  
## [104] &quot;OldWhichCells&quot;                  
## [105] &quot;pbmc_small&quot;                     
## [106] &quot;PCAPlot&quot;                        
## [107] &quot;PCASigGenes&quot;                    
## [108] &quot;PCHeatmap&quot;                      
## [109] &quot;PercentageFeatureSet&quot;           
## [110] &quot;PlotClusterTree&quot;                
## [111] &quot;PolyDimPlot&quot;                    
## [112] &quot;PolyFeaturePlot&quot;                
## [113] &quot;PrepSCTIntegration&quot;             
## [114] &quot;Project&quot;                        
## [115] &quot;Project&lt;-&quot;                      
## [116] &quot;ProjectDim&quot;                     
## [117] &quot;PurpleAndYellow&quot;                
## [118] &quot;Read10X&quot;                        
## [119] &quot;Read10X_h5&quot;                     
## [120] &quot;ReadAlevin&quot;                     
## [121] &quot;ReadAlevinCsv&quot;                  
## [122] &quot;ReadH5AD&quot;                       
## [123] &quot;Reductions&quot;                     
## [124] &quot;RegroupIdents&quot;                  
## [125] &quot;RelativeCounts&quot;                 
## [126] &quot;RenameAssays&quot;                   
## [127] &quot;RenameCells&quot;                    
## [128] &quot;RenameIdents&quot;                   
## [129] &quot;ReorderIdent&quot;                   
## [130] &quot;RestoreLegend&quot;                  
## [131] &quot;RidgePlot&quot;                      
## [132] &quot;RotatedAxis&quot;                    
## [133] &quot;RowMergeSparseMatrices&quot;         
## [134] &quot;RunALRA&quot;                        
## [135] &quot;RunCCA&quot;                         
## [136] &quot;RunICA&quot;                         
## [137] &quot;RunLSI&quot;                         
## [138] &quot;RunPCA&quot;                         
## [139] &quot;RunTSNE&quot;                        
## [140] &quot;RunUMAP&quot;                        
## [141] &quot;SampleUMI&quot;                      
## [142] &quot;ScaleData&quot;                      
## [143] &quot;ScoreJackStraw&quot;                 
## [144] &quot;SCTransform&quot;                    
## [145] &quot;SelectIntegrationFeatures&quot;      
## [146] &quot;SetAssayData&quot;                   
## [147] &quot;SetIdent&quot;                       
## [148] &quot;SetIntegrationData&quot;             
## [149] &quot;SeuratAxes&quot;                     
## [150] &quot;SeuratTheme&quot;                    
## [151] &quot;SpatialTheme&quot;                   
## [152] &quot;SplitObject&quot;                    
## [153] &quot;StashIdent&quot;                     
## [154] &quot;Stdev&quot;                          
## [155] &quot;StopCellbrowser&quot;                
## [156] &quot;SubsetByBarcodeInflections&quot;     
## [157] &quot;SubsetData&quot;                     
## [158] &quot;TF.IDF&quot;                         
## [159] &quot;Tool&quot;                           
## [160] &quot;Tool&lt;-&quot;                         
## [161] &quot;TopCells&quot;                       
## [162] &quot;TopFeatures&quot;                    
## [163] &quot;TransferData&quot;                   
## [164] &quot;TSNEPlot&quot;                       
## [165] &quot;UMAPPlot&quot;                       
## [166] &quot;UpdateSeuratObject&quot;             
## [167] &quot;UpdateSymbolList&quot;               
## [168] &quot;VariableFeaturePlot&quot;            
## [169] &quot;VariableFeatures&quot;               
## [170] &quot;VariableFeatures&lt;-&quot;             
## [171] &quot;VizDimLoadings&quot;                 
## [172] &quot;VlnPlot&quot;                        
## [173] &quot;WhichCells&quot;                     
## [174] &quot;WhiteBackground&quot;</code></pre>
</div>
</div>
</div>
<div id="preprocessing-step-1-filter-out-low-quality-cells" class="section level2">
<h2><span class="header-section-number">4.5</span> Preprocessing step 1 : Filter out low-quality cells</h2>
<p>The Seurat object initialization step above only considered cells that expressed at least 350 genes. Additionally, we would like to exclude cells that are damaged. A common metric to judge this (although by no means the only one) is the relative expression of mitochondrially derived genes. When the cells apoptose due to stress, their mitochondria becomes leaky and there is widespread RNA degradation. Thus a relative enrichment of mitochondrially derived genes can be a tell-tale sign of cell stress. Here, we compute the proportion of transcripts that are of mitochondrial origin for every cell (<code>percent.mt</code>), and visualize its distribution as a violin plot. We also use the <code>FeatureScatter</code> function to observe how <code>percent.mt</code> correlates with other metrics.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="data-wrangling-scrnaseq-1.html#cb62-1"></a><span class="co"># The number of genes and UMIs (nFeature_RNA nCount_RNA) are automatically calculated</span></span>
<span id="cb62-2"><a href="data-wrangling-scrnaseq-1.html#cb62-2"></a><span class="co"># for every object by Seurat.  For non-UMI data, nCount_RNA represents the sum of</span></span>
<span id="cb62-3"><a href="data-wrangling-scrnaseq-1.html#cb62-3"></a><span class="co"># the non-normalized values within a cell We calculate the percentage of</span></span>
<span id="cb62-4"><a href="data-wrangling-scrnaseq-1.html#cb62-4"></a><span class="co"># mitochondrial genes here and store it in percent.mt using AddMetaData.</span></span>
<span id="cb62-5"><a href="data-wrangling-scrnaseq-1.html#cb62-5"></a><span class="co"># We use GetAssayData(object = seurat, slot = &#39;data&#39;) since this represents non-transformed and</span></span>
<span id="cb62-6"><a href="data-wrangling-scrnaseq-1.html#cb62-6"></a><span class="co"># non-log-normalized counts The % of UMI mapping to MT-genes is a common</span></span>
<span id="cb62-7"><a href="data-wrangling-scrnaseq-1.html#cb62-7"></a><span class="co"># scRNA-seq QC metric.</span></span>
<span id="cb62-8"><a href="data-wrangling-scrnaseq-1.html#cb62-8"></a></span>
<span id="cb62-9"><a href="data-wrangling-scrnaseq-1.html#cb62-9"></a></span>
<span id="cb62-10"><a href="data-wrangling-scrnaseq-1.html#cb62-10"></a><span class="co"># The [[ operator can add columns to object metadata. This is a great place to stash QC stats</span></span>
<span id="cb62-11"><a href="data-wrangling-scrnaseq-1.html#cb62-11"></a>seurat[[<span class="st">&quot;percent.mt&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">PercentageFeatureSet</span>(<span class="dt">object =</span> seurat, <span class="dt">pattern =</span> <span class="st">&quot;^MT-&quot;</span>)</span>
<span id="cb62-12"><a href="data-wrangling-scrnaseq-1.html#cb62-12"></a></span>
<span id="cb62-13"><a href="data-wrangling-scrnaseq-1.html#cb62-13"></a><span class="co"># PercentageFeatureSet adds columns to object@meta.data, and is a great place to stash QC stats.</span></span>
<span id="cb62-14"><a href="data-wrangling-scrnaseq-1.html#cb62-14"></a><span class="co"># This also allows us to plot the metadata values using the Seurat&#39;s VlnPlot().</span></span>
<span id="cb62-15"><a href="data-wrangling-scrnaseq-1.html#cb62-15"></a><span class="kw">head</span>(seurat<span class="op">@</span>meta.data) <span class="co"># Before adding</span></span></code></pre></div>
<pre><code>##                  orig.ident nCount_RNA nFeature_RNA percent.mt
## AAACCTGAGCTAGTCT  10X_NSCLC       3605         1184  20.832178
## AAACCTGAGGGCACTA  10X_NSCLC       3828         1387   3.448276
## AAACCTGAGTACGTTC  10X_NSCLC       6457         1784   2.865108
## AAACCTGAGTCCGGTC  10X_NSCLC       3075         1092   6.276423
## AAACCTGCACCAGGTC  10X_NSCLC       9399         2625   3.202468
## AAACCTGCACCTCGTT  10X_NSCLC      48838         5602   2.321962</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="data-wrangling-scrnaseq-1.html#cb64-1"></a><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;percent.mt&quot;</span>), <span class="dt">ncol =</span> <span class="dv">3</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/vln_plot-1.png" width="672" /></p>
<p>Here we calculated the percent mitochondrial reads and added it to the Seurat object in the slot named meta.data. This allowed us to plot using the violin plot function provided by Seurat.</p>
<p>A third metric we use is the number of house keeping genes expressed in a cell. These genes reflect commomn processes active in a cell and hence are a good global quality measure. They are also abundant and are usually steadliy expressed in cells, thus less sensitive to the high dropout.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="data-wrangling-scrnaseq-1.html#cb65-1"></a><span class="co"># Load the the list of house keeping genes</span></span>
<span id="cb65-2"><a href="data-wrangling-scrnaseq-1.html#cb65-2"></a>hkgenes &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste0</span>(dirname, <span class="st">&quot;housekeepers.txt&quot;</span>), <span class="dt">skip =</span> <span class="dv">2</span>)</span>
<span id="cb65-3"><a href="data-wrangling-scrnaseq-1.html#cb65-3"></a>hkgenes &lt;-<span class="st"> </span><span class="kw">as.vector</span>(hkgenes<span class="op">$</span>V1)</span>
<span id="cb65-4"><a href="data-wrangling-scrnaseq-1.html#cb65-4"></a></span>
<span id="cb65-5"><a href="data-wrangling-scrnaseq-1.html#cb65-5"></a><span class="co"># remove hkgenes that were not found</span></span>
<span id="cb65-6"><a href="data-wrangling-scrnaseq-1.html#cb65-6"></a>hkgenes.found &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">toupper</span>(<span class="kw">rownames</span>(seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>data)) <span class="op">%in%</span><span class="st"> </span>hkgenes)</span></code></pre></div>
<p>Possible task: Feel like challenging yourself? write the code to do the following:
1. Sum the number of detected house keeping genes for each cell
2. Add this information as meta data to seurat
3. plot all metrics: “nFeature_RNA”, “nCount_RNA”, “percent.mt”,“n.exp.hkgenes” using VlnPlot
4. Scroll down to see if you got it!</p>
<p>If you feel like going through a more guided version, scroll down and follow the instructions.</p>
<p><strong>Alternative task:</strong> Sum the number of detected house keeping genes for each cell, then add this to the meta data</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="data-wrangling-scrnaseq-1.html#cb66-1"></a><span class="co">#### n.expressed.hkgenes &lt;- ?(seurat@data[hkgenes.found, ] &gt; ?)</span></span>
<span id="cb66-2"><a href="data-wrangling-scrnaseq-1.html#cb66-2"></a><span class="co">#### seurat &lt;- AddMetaData(object = ?, ? = ?, col.name = &quot;n.exp.hkgenes&quot;)</span></span>
<span id="cb66-3"><a href="data-wrangling-scrnaseq-1.html#cb66-3"></a>n.expressed.hkgenes &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>data[hkgenes.found, ] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</span>
<span id="cb66-4"><a href="data-wrangling-scrnaseq-1.html#cb66-4"></a>seurat &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> seurat, <span class="dt">metadata =</span> n.expressed.hkgenes, <span class="dt">col.name =</span> <span class="st">&quot;n.exp.hkgenes&quot;</span>)</span>
<span id="cb66-5"><a href="data-wrangling-scrnaseq-1.html#cb66-5"></a><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;percent.mt&quot;</span>, <span class="st">&quot;n.exp.hkgenes&quot;</span>), <span class="dt">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/add_number_of_house_keeping_genes-1.png" width="672" /></p>
<p>Is there a correlation between the measurements? For example, number of UMIs with number of genes?
Possible task: Feel like challenging yourself? write the code to do the following:
Can you plot the nGene vs nUMI (hint:GenePlot)? What is the correlation? Do you see a strange subpopulation? What do you think happened with these cells?</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="data-wrangling-scrnaseq-1.html#cb67-1"></a>seurat</span></code></pre></div>
<p>Scroll down to see the command</p>
<p><strong>Alternative task:</strong>
Can you plot the nFeature_RNA vs nCount_RNA (hint:FeatureScatter)? What is the correlation? Do you see a strange subpopulation? What do you think happened with these cells?</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="data-wrangling-scrnaseq-1.html#cb68-1"></a><span class="co"># FeatureScatter is typically used to visualize feature-feature relationships, but can be used</span></span>
<span id="cb68-2"><a href="data-wrangling-scrnaseq-1.html#cb68-2"></a><span class="co"># for anything calculated by the object, i.e. columns in object metadata, PC scores etc.</span></span>
<span id="cb68-3"><a href="data-wrangling-scrnaseq-1.html#cb68-3"></a></span>
<span id="cb68-4"><a href="data-wrangling-scrnaseq-1.html#cb68-4"></a>plot1 &lt;-<span class="st"> </span><span class="kw">FeatureScatter</span>(seurat, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;percent.mt&quot;</span>)</span>
<span id="cb68-5"><a href="data-wrangling-scrnaseq-1.html#cb68-5"></a>plot2 &lt;-<span class="st"> </span><span class="kw">FeatureScatter</span>(seurat, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;nFeature_RNA&quot;</span>)</span>
<span id="cb68-6"><a href="data-wrangling-scrnaseq-1.html#cb68-6"></a><span class="kw">CombinePlots</span>(<span class="dt">plots =</span> <span class="kw">list</span>(plot1, plot2))</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/alt_task-1.png" width="672" /></p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="data-wrangling-scrnaseq-1.html#cb69-1"></a><span class="co">### FeatureScatter(object = seurat, feature1 = ?, feature2 = ?)  ## could be genes or meta data columns</span></span>
<span id="cb69-2"><a href="data-wrangling-scrnaseq-1.html#cb69-2"></a><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> seurat, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;nFeature_RNA&quot;</span>)</span></code></pre></div>
</div>
<div id="examine-contents-of-seurat-object" class="section level2">
<h2><span class="header-section-number">4.6</span> Examine contents of Seurat object</h2>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="data-wrangling-scrnaseq-1.html#cb70-1"></a><span class="kw">str</span>(seurat)</span></code></pre></div>
<p>These are the slots in the Seurat object. Some of the slots are automatically updated by Seurat as you move through analysis. Take a moment to look through the information, knowing the slots allow you to leverage work Seurat has already done for you.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="data-wrangling-scrnaseq-1.html#cb71-1"></a><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>), <span class="dt">group.by =</span> <span class="kw">c</span>(<span class="st">&#39;orig.ident&#39;</span>))</span></code></pre></div>
<p>Here we plot the number of genes per cell by what Seurat calls orig.ident. Identity is a concept that is used in the Seurat object to refer to the cell identity. In this case, the cell identity is 10X_NSCLC, but after we cluster the cells, the cell identity will be whatever cluster the cell belongs to. We will see how identity updates as we go throught the analysis.</p>
<p>Next, let’s filter the cells based on the quality control metrics.
Filter based on:
1. nFeature_RNA
2. percent.mt
3. n.exp.hkgenes</p>
<p><strong>Task: Change the thresholds to what you think they should be according to the violin plots</strong></p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="data-wrangling-scrnaseq-1.html#cb72-1"></a><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>,<span class="st">&quot;percent.mt&quot;</span>,<span class="st">&quot;n.exp.hkgenes&quot;</span>), <span class="dt">ncol =</span> <span class="dv">3</span>)</span></code></pre></div>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="data-wrangling-scrnaseq-1.html#cb73-1"></a><span class="co"># seurat &lt;- SubsetData(object = seurat, subset.names = c(&quot;nFeature_RNA&quot;, &quot;percent.mito&quot;,&quot;n.exp.hkgenes&quot;), low.thresholds = c(350, -Inf,55), high.thresholds = c(5000, 0.1, Inf))</span></span>
<span id="cb73-2"><a href="data-wrangling-scrnaseq-1.html#cb73-2"></a></span>
<span id="cb73-3"><a href="data-wrangling-scrnaseq-1.html#cb73-3"></a>seurat &lt;-<span class="st"> </span><span class="kw">subset</span>(seurat, nFeature_RNA <span class="op">&lt;</span><span class="st"> </span><span class="dv">5000</span>)</span>
<span id="cb73-4"><a href="data-wrangling-scrnaseq-1.html#cb73-4"></a>seurat &lt;-<span class="st"> </span><span class="kw">subset</span>(seurat, nFeature_RNA <span class="op">&gt;</span><span class="st"> </span><span class="dv">350</span>)</span>
<span id="cb73-5"><a href="data-wrangling-scrnaseq-1.html#cb73-5"></a>seuart &lt;-<span class="st"> </span><span class="kw">subset</span>(seurat, n.exp.hkgenes <span class="op">&gt;</span><span class="st"> </span><span class="dv">55</span>)</span>
<span id="cb73-6"><a href="data-wrangling-scrnaseq-1.html#cb73-6"></a>seuart &lt;-<span class="st"> </span><span class="kw">subset</span>(seurat, percent.mt <span class="op">&lt;</span><span class="st"> </span><span class="dv">10</span>)</span></code></pre></div>
<p>How many cells are you left with?</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="data-wrangling-scrnaseq-1.html#cb74-1"></a>seurat</span></code></pre></div>
<div id="preprocessing-step-2-expression-normalization" class="section level3">
<h3><span class="header-section-number">4.6.1</span> Preprocessing step 2 : Expression normalization</h3>
<p>After removing unwanted genes cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. There have been many methods to normalize the data, but this is the simplest and the most intuitive. The division by total expression is done to change all expression counts to a relative measure, since experience has suggested that technical factors (e.g. capture rate, efficiency of reverse transcription) are largely responsible for the variation in the number of molecules per cell, although genuine biological factors (e.g. cell cycle stage, cell size) also play a smaller, but non-negligible role. The log-transformation is a commonly used transformation that has many desirable properties, such as variance stabilization (can you think of others?).</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="data-wrangling-scrnaseq-1.html#cb75-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> seurat, <span class="dt">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="dt">scale.factor =</span> <span class="fl">1e4</span>)</span></code></pre></div>
<p>Well there you have it! A filtered and normalized gene-expression data set. A great accomplishment for your first dive into scRNA-Seq analysis. Well done!</p>
</div>
</div>
<div id="detection-of-variable-genes-across-the-single-cells" class="section level2">
<h2><span class="header-section-number">4.7</span> Detection of variable genes across the single cells</h2>
<p>Seurat calculates highly variable genes and focuses on these for downstream analysis. <strong><code>FindVariableFeatures</code></strong> calculates the average expression and dispersion for each gene, places these genes into bins, and then calculates a z-score for dispersion within each bin. This helps control for the relationship between variability and average expression. This function is unchanged from (Macosko <em>et al</em>.), but new methods for variable gene expression identification are coming soon. We suggest that users set these parameters to mark visual outliers on the dispersion plot, but the exact parameter settings may vary based on the data type, heterogeneity in the sample, and normalization strategy. The parameters here identify ~3,000 variable genes, and represent typical parameter settings for UMI data that is normalized to a total of 1e4 molecules.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="data-wrangling-scrnaseq-1.html#cb76-1"></a><span class="co">### seurat &lt;- FindVariableFeatures(object = seurat, selection.method = ?, nfeatures = ?) </span></span>
<span id="cb76-2"><a href="data-wrangling-scrnaseq-1.html#cb76-2"></a><span class="co">### Task: ?FindVariableFeatures into the console to read about different selection methods</span></span>
<span id="cb76-3"><a href="data-wrangling-scrnaseq-1.html#cb76-3"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> seurat, <span class="dt">selection.method =</span> <span class="st">&quot;vst&quot;</span>, <span class="dt">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb76-4"><a href="data-wrangling-scrnaseq-1.html#cb76-4"></a></span>
<span id="cb76-5"><a href="data-wrangling-scrnaseq-1.html#cb76-5"></a><span class="co"># Identify the 10 most highly variable genes</span></span>
<span id="cb76-6"><a href="data-wrangling-scrnaseq-1.html#cb76-6"></a>top10 &lt;-<span class="st"> </span><span class="kw">head</span>(<span class="kw">VariableFeatures</span>(seurat), <span class="dv">10</span>)</span></code></pre></div>
<p>We can see the Seurat object slots have updated for the FindVariableFetures section. Let’s use the slot to see how many variable genes we found.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="data-wrangling-scrnaseq-1.html#cb77-1"></a><span class="kw">str</span>(seurat)</span>
<span id="cb77-2"><a href="data-wrangling-scrnaseq-1.html#cb77-2"></a><span class="kw">length</span>(<span class="dt">x =</span> seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>var.features)</span></code></pre></div>
<p><strong>Task: how does changing the parameters for find variable genes function changes the number of the found genes?</strong>
Play with the parameters - what makes the function find more variable genes? less?</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="data-wrangling-scrnaseq-1.html#cb78-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> seurat, <span class="dt">selection.method =</span> <span class="st">&quot;vst&quot;</span>, <span class="dt">nfeatures =</span> <span class="dv">2000</span>, </span>
<span id="cb78-2"><a href="data-wrangling-scrnaseq-1.html#cb78-2"></a>                               <span class="dt">mean.function =</span> ExpMean, <span class="dt">dispersion.function =</span> LogVMR, </span>
<span id="cb78-3"><a href="data-wrangling-scrnaseq-1.html#cb78-3"></a>                               <span class="dt">num.bin =</span> <span class="dv">40</span>, </span>
<span id="cb78-4"><a href="data-wrangling-scrnaseq-1.html#cb78-4"></a>                               <span class="dt">mean.cutoff =</span> <span class="kw">c</span>(<span class="fl">0.0125</span>, <span class="dv">1</span>),</span>
<span id="cb78-5"><a href="data-wrangling-scrnaseq-1.html#cb78-5"></a>                               <span class="dt">dispersion.cutoff =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>))</span>
<span id="cb78-6"><a href="data-wrangling-scrnaseq-1.html#cb78-6"></a></span>
<span id="cb78-7"><a href="data-wrangling-scrnaseq-1.html#cb78-7"></a><span class="co">## Check number of variable genes</span></span>
<span id="cb78-8"><a href="data-wrangling-scrnaseq-1.html#cb78-8"></a><span class="kw">length</span>(seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>var.features)</span></code></pre></div>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="data-wrangling-scrnaseq-1.html#cb79-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> seurat, <span class="dt">selection.method =</span> <span class="st">&quot;vst&quot;</span>, <span class="dt">nfeatures =</span> <span class="dv">2000</span>, </span>
<span id="cb79-2"><a href="data-wrangling-scrnaseq-1.html#cb79-2"></a>                               <span class="dt">mean.function =</span> ExpMean, <span class="dt">dispersion.function =</span> LogVMR, </span>
<span id="cb79-3"><a href="data-wrangling-scrnaseq-1.html#cb79-3"></a>                               <span class="dt">num.bin =</span> <span class="dv">10</span>, </span>
<span id="cb79-4"><a href="data-wrangling-scrnaseq-1.html#cb79-4"></a>                               <span class="dt">mean.cutoff =</span> <span class="kw">c</span>(<span class="fl">0.0125</span>, <span class="dv">1</span>),</span>
<span id="cb79-5"><a href="data-wrangling-scrnaseq-1.html#cb79-5"></a>                               <span class="dt">dispersion.cutoff =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>))</span>
<span id="cb79-6"><a href="data-wrangling-scrnaseq-1.html#cb79-6"></a></span>
<span id="cb79-7"><a href="data-wrangling-scrnaseq-1.html#cb79-7"></a><span class="co">## Check number of variable genes</span></span>
<span id="cb79-8"><a href="data-wrangling-scrnaseq-1.html#cb79-8"></a><span class="kw">length</span>(seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>var.features)</span></code></pre></div>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="data-wrangling-scrnaseq-1.html#cb80-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> seurat, <span class="dt">selection.method =</span> ??, <span class="dt">nfeatures =</span> <span class="dv">2000</span>, </span>
<span id="cb80-2"><a href="data-wrangling-scrnaseq-1.html#cb80-2"></a>                               <span class="dt">mean.function =</span> ExpMean, <span class="dt">dispersion.function =</span> LogVMR, </span>
<span id="cb80-3"><a href="data-wrangling-scrnaseq-1.html#cb80-3"></a>                               <span class="dt">num.bin =</span> ??, </span>
<span id="cb80-4"><a href="data-wrangling-scrnaseq-1.html#cb80-4"></a>                               <span class="dt">mean.cutoff =</span> <span class="kw">c</span>(??, ??),</span>
<span id="cb80-5"><a href="data-wrangling-scrnaseq-1.html#cb80-5"></a>                               <span class="dt">dispersion.cutoff =</span> <span class="kw">c</span>(??, ??))</span>
<span id="cb80-6"><a href="data-wrangling-scrnaseq-1.html#cb80-6"></a></span>
<span id="cb80-7"><a href="data-wrangling-scrnaseq-1.html#cb80-7"></a><span class="co">## Check number of variable genes</span></span>
<span id="cb80-8"><a href="data-wrangling-scrnaseq-1.html#cb80-8"></a><span class="kw">length</span>(seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>var.features)</span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="data-wrangling-scrnaseq-1.html#cb81-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> seurat, <span class="dt">selection.method =</span> ??, <span class="dt">nfeatures =</span> <span class="dv">2000</span>, </span>
<span id="cb81-2"><a href="data-wrangling-scrnaseq-1.html#cb81-2"></a>                               <span class="dt">mean.function =</span> ExpMean, <span class="dt">dispersion.function =</span> LogVMR, </span>
<span id="cb81-3"><a href="data-wrangling-scrnaseq-1.html#cb81-3"></a>                               <span class="dt">num.bin =</span> ??, </span>
<span id="cb81-4"><a href="data-wrangling-scrnaseq-1.html#cb81-4"></a>                               <span class="dt">mean.cutoff =</span> <span class="kw">c</span>(??, ??),</span>
<span id="cb81-5"><a href="data-wrangling-scrnaseq-1.html#cb81-5"></a>                               <span class="dt">dispersion.cutoff =</span> <span class="kw">c</span>(??, ??))</span>
<span id="cb81-6"><a href="data-wrangling-scrnaseq-1.html#cb81-6"></a></span>
<span id="cb81-7"><a href="data-wrangling-scrnaseq-1.html#cb81-7"></a><span class="co">## Check number of variable genes</span></span>
<span id="cb81-8"><a href="data-wrangling-scrnaseq-1.html#cb81-8"></a><span class="kw">length</span>(seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>var.features)</span></code></pre></div>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="data-wrangling-scrnaseq-1.html#cb82-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>()</span>
<span id="cb82-2"><a href="data-wrangling-scrnaseq-1.html#cb82-2"></a></span>
<span id="cb82-3"><a href="data-wrangling-scrnaseq-1.html#cb82-3"></a><span class="co">## Check number of variable genes</span></span>
<span id="cb82-4"><a href="data-wrangling-scrnaseq-1.html#cb82-4"></a><span class="kw">length</span>(seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>var.features)</span></code></pre></div>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="data-wrangling-scrnaseq-1.html#cb83-1"></a><span class="co"># plot variable features with and without labels</span></span>
<span id="cb83-2"><a href="data-wrangling-scrnaseq-1.html#cb83-2"></a>plot1 &lt;-<span class="st"> </span><span class="kw">VariableFeaturePlot</span>(seurat)</span>
<span id="cb83-3"><a href="data-wrangling-scrnaseq-1.html#cb83-3"></a>plot2 &lt;-<span class="st"> </span><span class="kw">LabelPoints</span>(<span class="dt">plot =</span> plot1, <span class="dt">points =</span> top10, <span class="dt">repel =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## When using repel, set xnudge and ynudge to 0 for optimal results</code></pre>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="data-wrangling-scrnaseq-1.html#cb85-1"></a><span class="co">## CombinePlots(plots = list(plot1, plot2))</span></span>
<span id="cb85-2"><a href="data-wrangling-scrnaseq-1.html#cb85-2"></a>plot2</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_vargenes-1.png" width="672" /></p>
</div>
<div id="gene-set-expression-across-cells" class="section level2">
<h2><span class="header-section-number">4.8</span> Gene set expression across cells</h2>
<p>Sometimes we want to ask what is the expression of a set of a genes across cells. This set of genes may make up a gene expression program we are interested in. Another benefit at looking at gene sets is it reduces the effects of drop outs.</p>
<p>Below, we look at genes involved in: T cells, the cell cycle and the stress signature upon cell dissociation. We calculate these genes average expression levels on the single cell level, while controlling for technical effects.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="data-wrangling-scrnaseq-1.html#cb86-1"></a><span class="co"># Read in a list of cell cycle markers, from Tirosh et al, 2015.</span></span>
<span id="cb86-2"><a href="data-wrangling-scrnaseq-1.html#cb86-2"></a><span class="co"># We can segregate this list into markers of G2/M phase and markers of S phase.</span></span>
<span id="cb86-3"><a href="data-wrangling-scrnaseq-1.html#cb86-3"></a>s.genes &lt;-<span class="st"> </span>Seurat<span class="op">::</span>cc.genes<span class="op">$</span>s.genes</span>
<span id="cb86-4"><a href="data-wrangling-scrnaseq-1.html#cb86-4"></a>s.genes &lt;-<span class="st"> </span>s.genes[s.genes <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(seurat)] <span class="co"># genes in dataset</span></span>
<span id="cb86-5"><a href="data-wrangling-scrnaseq-1.html#cb86-5"></a>g2m.genes &lt;-<span class="st"> </span>Seurat<span class="op">::</span>cc.genes<span class="op">$</span>g2m.genes</span>
<span id="cb86-6"><a href="data-wrangling-scrnaseq-1.html#cb86-6"></a>g2m.genes &lt;-<span class="st"> </span>g2m.genes[g2m.genes <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(seurat)] <span class="co"># genes in dataset</span></span>
<span id="cb86-7"><a href="data-wrangling-scrnaseq-1.html#cb86-7"></a></span>
<span id="cb86-8"><a href="data-wrangling-scrnaseq-1.html#cb86-8"></a>seurat &lt;-<span class="st"> </span><span class="kw">CellCycleScoring</span>(<span class="dt">object =</span> seurat, <span class="dt">s.features =</span> s.genes, <span class="dt">g2m.features =</span> g2m.genes, <span class="dt">set.ident =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p><strong>Task: Use markers for dissociation to calculate dissociation score</strong></p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="data-wrangling-scrnaseq-1.html#cb87-1"></a><span class="co"># Genes upregulated during dissociation of tissue into single cells.</span></span>
<span id="cb87-2"><a href="data-wrangling-scrnaseq-1.html#cb87-2"></a>genes.dissoc &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ATF3&quot;</span>, <span class="st">&quot;BTG2&quot;</span>, <span class="st">&quot;CEBPB&quot;</span>, <span class="st">&quot;CEBPD&quot;</span>, <span class="st">&quot;CXCL3&quot;</span>, <span class="st">&quot;CXCL2&quot;</span>, <span class="st">&quot;CXCL1&quot;</span>, <span class="st">&quot;DNAJA1&quot;</span>, <span class="st">&quot;DNAJB1&quot;</span>, <span class="st">&quot;DUSP1&quot;</span>, </span>
<span id="cb87-3"><a href="data-wrangling-scrnaseq-1.html#cb87-3"></a>                  <span class="st">&quot;EGR1&quot;</span>, <span class="st">&quot;FOS&quot;</span>, <span class="st">&quot;FOSB&quot;</span>, <span class="st">&quot;HSP90AA1&quot;</span>, <span class="st">&quot;HSP90AB1&quot;</span>, <span class="st">&quot;HSPA1A&quot;</span>, <span class="st">&quot;HSPA1B&quot;</span>, <span class="st">&quot;HSPA1A&quot;</span>, <span class="st">&quot;HSPA1B&quot;</span>, </span>
<span id="cb87-4"><a href="data-wrangling-scrnaseq-1.html#cb87-4"></a>                  <span class="st">&quot;HSPA8&quot;</span>, <span class="st">&quot;HSPB1&quot;</span>, <span class="st">&quot;HSPE1&quot;</span>, <span class="st">&quot;HSPH1&quot;</span>, <span class="st">&quot;ID3&quot;</span>, <span class="st">&quot;IER2&quot;</span>, <span class="st">&quot;JUN&quot;</span>, <span class="st">&quot;JUNB&quot;</span>, <span class="st">&quot;JUND&quot;</span>, <span class="st">&quot;MT1X&quot;</span>, <span class="st">&quot;NFKBIA&quot;</span>, </span>
<span id="cb87-5"><a href="data-wrangling-scrnaseq-1.html#cb87-5"></a>                  <span class="st">&quot;NR4A1&quot;</span>, <span class="st">&quot;PPP1R15A&quot;</span>, <span class="st">&quot;SOCS3&quot;</span>, <span class="st">&quot;ZFP36&quot;</span>)</span>
<span id="cb87-6"><a href="data-wrangling-scrnaseq-1.html#cb87-6"></a><span class="co">#### seurat &lt;- ?(?, genes.list = list(?), ctrl.size = 20, enrich.name = &quot;genes_dissoc&quot;)</span></span>
<span id="cb87-7"><a href="data-wrangling-scrnaseq-1.html#cb87-7"></a>seurat &lt;-<span class="st"> </span><span class="kw">AddModuleScore</span>(seurat, <span class="dt">features =</span> <span class="kw">list</span>(genes.dissoc), <span class="dt">ctrl.size =</span> <span class="dv">20</span>, <span class="dt">enrich.name =</span> <span class="st">&quot;genes_dissoc&quot;</span>)</span></code></pre></div>
<p><strong>Task:</strong>
- Plot the correlation between number of genes and S score.
- How do we know the name of these scores in the seurat meta data?</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="data-wrangling-scrnaseq-1.html#cb88-1"></a><span class="co">### GenePlot(seurat, &quot;S.Score&quot;, &quot;nGene&quot;)</span></span>
<span id="cb88-2"><a href="data-wrangling-scrnaseq-1.html#cb88-2"></a><span class="kw">FeatureScatter</span>(seurat, <span class="st">&quot;S.Score&quot;</span>, <span class="st">&quot;nFeature_RNA&quot;</span>)</span></code></pre></div>
<p><strong>Bonus: Can you cluster the data based on the variable genes alone?</strong></p>
<p>Congratulations! You can identify and visualize cell subsets and the marker genes that describe these cell subsets. This is a very powerful analysis pattern often seen in publications. Well done!</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="expression-qc-and-normalization-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="identifying-cell-populations-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["KrumlovSingleCellWorkshop2020.pdf", "KrumlovSingleCellWorkshop2020.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
