<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta name="generator" content="bookdown 0.16 and GitBook 2.6.7" />

  <meta property="og:title" content="4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/" />
  <meta property="og:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />
  
  <meta name="github-repo" content="broadinstitute/KrumlovSingleCellWorkship2020" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4 Data Wrangling scRNAseq | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  
  <meta name="twitter:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />

<meta name="author" content="Kirk Gosik" />


<meta name="date" content="2020-01-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="expression-qc-and-normalization-1.html"/>
<link rel="next" href="the-following-examples-use-the-pbmc-3k-dataset.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#introduction-1"><i class="fa fa-check"></i><b>1.5.1</b> Introduction</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#transcriptome-quantification"><i class="fa fa-check"></i><b>1.5.2</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#expression-qc-and-normalization"><i class="fa fa-check"></i><b>1.5.3</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#data-wrangling-scrnaseq"><i class="fa fa-check"></i><b>1.5.4</b> Data Wrangling scRNAseq</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#identifying-cell-populations"><i class="fa fa-check"></i><b>1.5.5</b> Identifying Cell Populations</a></li>
<li class="chapter" data-level="1.5.6" data-path="index.html"><a href="index.html#feature-selection-and-cluster-analysis"><i class="fa fa-check"></i><b>1.5.6</b> Feature Selection and Cluster Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html"><i class="fa fa-check"></i><b>2</b> Transcriptome Quantification</a><ul>
<li class="chapter" data-level="2.1" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html#google-slides"><i class="fa fa-check"></i><b>2.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html"><i class="fa fa-check"></i><b>3</b> Expression QC and Normalization</a><ul>
<li class="chapter" data-level="3.1" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html#google-slides-1"><i class="fa fa-check"></i><b>3.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html"><i class="fa fa-check"></i><b>4</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#goal"><i class="fa fa-check"></i><b>4.1</b> Goal</a></li>
<li class="chapter" data-level="4.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#introduction-2"><i class="fa fa-check"></i><b>4.2</b> Introduction</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#load-necessary-packages"><i class="fa fa-check"></i><b>4.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>4.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>4.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>4.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>4.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>4.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>4.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>4.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#beginning-with-seurat"><i class="fa fa-check"></i><b>4.4</b> Beginning with Seurat:</a><ul>
<li class="chapter" data-level="4.4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>4.4.1</b> Creating a seurat object</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html"><i class="fa fa-check"></i><b>5</b> The following examples use the PBMC 3k dataset</a><ul>
<li class="chapter" data-level="5.1" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>5.1</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="5.2" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>5.2</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="5.2.1" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>5.2.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>5.3</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="5.4" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>5.4</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html"><i class="fa fa-check"></i><b>6</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="6.1" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html#google-slides-2"><i class="fa fa-check"></i><b>6.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html"><i class="fa fa-check"></i><b>7</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#abstract"><i class="fa fa-check"></i><b>7.1</b> Abstract</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#seurat-tutorial-redo"><i class="fa fa-check"></i><b>7.2</b> Seurat Tutorial Redo</a><ul>
<li class="chapter" data-level="7.2.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#preprocessing-steps"><i class="fa fa-check"></i><b>7.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="7.2.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>7.2.2</b> Start of Identifying Cell Types</a></li>
<li class="chapter" data-level="7.2.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#run-non-linear-dimensional-reduction-umaptsne"><i class="fa fa-check"></i><b>7.2.3</b> Run non-linear dimensional reduction (UMAP/tSNE)</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#feature-selection"><i class="fa fa-check"></i><b>7.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="7.3.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#differential-expression-analysis"><i class="fa fa-check"></i><b>7.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="7.3.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#check-clusters"><i class="fa fa-check"></i><b>7.3.2</b> Check Clusters</a></li>
<li class="chapter" data-level="7.3.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#view-entire-object-structure"><i class="fa fa-check"></i><b>7.3.3</b> View Entire Object Structure</a></li>
<li class="chapter" data-level="7.3.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>7.3.4</b> Practice Visualizing/Embedding</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#other-options-for-analysis"><i class="fa fa-check"></i><b>7.4</b> Other Options For Analysis</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-wrangling-scrnaseq-1" class="section level1">
<h1><span class="header-section-number">4</span> Data Wrangling scRNAseq</h1>
<div id="goal" class="section level2">
<h2><span class="header-section-number">4.1</span> Goal</h2>
<ul>
<li>To give you experience with the analysis of single cell RNA sequencing (scRNA-seq) including performing quality control and identifying cell type subsets.</li>
<li>To introduce you to scRNA-seq analysis using the Seurat package.</li>
</ul>
</div>
<div id="introduction-2" class="section level2">
<h2><span class="header-section-number">4.2</span> Introduction</h2>
<p>Data produced in a single cell RNA-seq experiment has several interesting characteristics that make it distinct from data produced in a bulk population RNA-seq experiment. Two characteristics that are important to keep in mind when working with scRNA-Seq are drop-out (the excessive amount of zeros due to limiting mRNA) and the potential for quality control (QC) metrics to be confounded with biology. This combined with the ability to measure heterogeniety from cells in samples has shifted the field away from the typical analysis in population-based RNA-Seq. Here we demonstrate some approaches to quality control, followed by identifying and analyzing cell subsets.</p>
<p>For this tutorial, we will be analyzing the a dataset of Non-Small Cell Lung Cancer Cells (NSCLC) freely available from 10X Genomics (<a href="https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex" class="uri">https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex</a>), using the Seurat R package (<a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a>), a popular and powerful set of tools to conduct scRNA-seq analysis in R. In this dataset, there are 7802 single cells that were sequenced on the Illumina NovaSeq 6000. Please note this tutorial borrows heavily from Seurat’s tutorials, so feel free to go through them in more detail.</p>
<div id="load-necessary-packages" class="section level3">
<h3><span class="header-section-number">4.2.1</span> Load necessary packages</h3>
<p>When loading libraries, we are asking R to load code for us written by someone else. It is a convenient way to leverage and reproduce methodology developed by others.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="data-wrangling-scrnaseq-1.html#cb1-1"></a>knitr<span class="op">::</span>opts_knit<span class="op">$</span><span class="kw">set</span>(<span class="dt">progress=</span><span class="ot">FALSE</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</span>
<span id="cb1-2"><a href="data-wrangling-scrnaseq-1.html#cb1-2"></a></span>
<span id="cb1-3"><a href="data-wrangling-scrnaseq-1.html#cb1-3"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb1-4"><a href="data-wrangling-scrnaseq-1.html#cb1-4"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb1-5"><a href="data-wrangling-scrnaseq-1.html#cb1-5"></a><span class="kw">library</span>(Matrix)</span></code></pre></div>
</div>
<div id="read-in-nsclc-counts-matrix." class="section level3">
<h3><span class="header-section-number">4.2.2</span> Read in NSCLC counts matrix.</h3>
<p>The data for Non-Small Cell Lung Cancer Cells (NSCLC) is freely available from 10X Genomics (<a href="https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex" class="uri">https://support.10xgenomics.com/single-cell-vdj/datasets/2.2.0/vdj_v1_hs_nsclc_5gex</a>). We start by reading in the counts matrix generated by the Cell Ranger count program.</p>
<p>Task: Change the directory name to mydir/ where you saved your data</p>
<p><strong>Task: Check the <code>dirname</code> to dir where you saved your data</strong></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="data-wrangling-scrnaseq-1.html#cb2-1"></a><span class="co">## dirname &lt;- &quot;/home/genomics/workshop_materials/scrna/&quot;</span></span>
<span id="cb2-2"><a href="data-wrangling-scrnaseq-1.html#cb2-2"></a>dirname &lt;-<span class="st"> &quot;/Users/kgosik/Documents/data/cellranger/vdj-nsclc/&quot;</span></span>
<span id="cb2-3"><a href="data-wrangling-scrnaseq-1.html#cb2-3"></a>counts_matrix_filename =<span class="st"> </span><span class="kw">paste0</span>(dirname, <span class="st">&quot;filtered_gene_bc_matrices/GRCh38/&quot;</span>)</span>
<span id="cb2-4"><a href="data-wrangling-scrnaseq-1.html#cb2-4"></a>counts &lt;-<span class="st"> </span><span class="kw">Read10X</span>(<span class="dt">data.dir =</span> counts_matrix_filename)  <span class="co"># Seurat function to read in 10x count data</span></span>
<span id="cb2-5"><a href="data-wrangling-scrnaseq-1.html#cb2-5"></a></span>
<span id="cb2-6"><a href="data-wrangling-scrnaseq-1.html#cb2-6"></a><span class="co"># To minimize memory use on the docker - choose only the first 1000 cells</span></span>
<span id="cb2-7"><a href="data-wrangling-scrnaseq-1.html#cb2-7"></a>counts &lt;-<span class="st"> </span>counts[, <span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>]</span></code></pre></div>
</div>
<div id="lets-examine-the-sparse-counts-matrix" class="section level3">
<h3><span class="header-section-number">4.2.3</span> Let’s examine the sparse counts matrix</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="data-wrangling-scrnaseq-1.html#cb3-1"></a>counts[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## 10 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
##               AAACCTGAGCTAGTCT AAACCTGAGGGCACTA AAACCTGAGTACGTTC
## RP11-34P13.3                 .                .                .
## FAM138A                      .                .                .
## OR4F5                        .                .                .
## RP11-34P13.7                 .                .                .
## RP11-34P13.8                 .                .                .
## RP11-34P13.14                .                .                .
## RP11-34P13.9                 .                .                .
## FO538757.3                   .                .                .
## FO538757.2                   .                .                .
## AP006222.2                   .                .                .</code></pre>
<p>Here we see the upper left corner of the sparse matrix. The columns are indexed by 10x cell barcodes (each 16 nt long), and the rows are the gene names. We mentioned these matrices are sparse, here we see only zeroes (indicated by the “.” symbol); this is the most common value in these sparse matrices. Next, let us look at the dimensions of this matrix.</p>
</div>
<div id="how-big-is-the-matrix" class="section level3">
<h3><span class="header-section-number">4.2.4</span> How big is the matrix?</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="data-wrangling-scrnaseq-1.html#cb5-1"></a><span class="kw">dim</span>(counts) <span class="co"># report number of genes (rows) and number of cells (columns)</span></span></code></pre></div>
<pre><code>## [1] 33694  1000</code></pre>
<p>Here we see the counts matrix has 33694 genes and 7802 cells.</p>
</div>
<div id="how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix" class="section level3">
<h3><span class="header-section-number">4.2.5</span> How much memory does a sparse matrix take up relative to a dense matrix?</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="data-wrangling-scrnaseq-1.html#cb7-1"></a><span class="kw">object.size</span>(counts) <span class="co"># size in bytes</span></span></code></pre></div>
<pre><code>## [1] 23923312 bytes</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="data-wrangling-scrnaseq-1.html#cb9-1"></a><span class="kw">object.size</span>(<span class="kw">as.matrix</span>(counts)) <span class="co"># size in bytes</span></span></code></pre></div>
<pre><code>## [1] 271919144 bytes</code></pre>
<p>We see here that the sparse matrix takes 225 Mb in memory while storing the matrix in a dense format (where all count values including zeros are stored) takes almost 10 times as much memory! This memory saving is very important, especially as data sets are now being created that are beyond a million cells. These matrices can become unmanageable without special computing resources.</p>
<p>In the sparse representation, we assume that the majority of count values in a matrix are zero. We only store the non-zero values. This is implemented in the Matrix package using a dgTMatrix object.</p>
</div>
</div>
<div id="filtering-low-quality-cells" class="section level2">
<h2><span class="header-section-number">4.3</span> Filtering low-quality cells</h2>
<p>You can learn a lot about your scRNA-seq data’s quality with simple plotting. Let’s do some plotting to look at the number of reads per cell, reads per genes, expressed genes per cell (often called complexity), and rarity of genes (cells expressing genes).</p>
<div id="look-at-the-summary-counts-for-genes-and-cells" class="section level3">
<h3><span class="header-section-number">4.3.1</span> Look at the summary counts for genes and cells</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="data-wrangling-scrnaseq-1.html#cb11-1"></a>counts_per_cell &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(counts)</span>
<span id="cb11-2"><a href="data-wrangling-scrnaseq-1.html#cb11-2"></a>counts_per_gene &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">rowSums</span>(counts)</span>
<span id="cb11-3"><a href="data-wrangling-scrnaseq-1.html#cb11-3"></a>genes_per_cell &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="co"># count gene only if it has non-zero reads mapped.</span></span></code></pre></div>
<p><strong>Task: In a similar way, can you calculate cells per genes? replace the ‘?’ in the command below</strong></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="data-wrangling-scrnaseq-1.html#cb12-1"></a><span class="co">#### cells_per_gene &lt;- Matrix::?(counts&gt;?) # only count cells where the gene is expressed</span></span>
<span id="cb12-2"><a href="data-wrangling-scrnaseq-1.html#cb12-2"></a>cells_per_gene &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">rowSums</span>(counts<span class="op">&gt;</span><span class="dv">0</span>) <span class="co"># only count cells where the gene is expressed</span></span></code></pre></div>
<p>colSums and rowSums are functions that work on each row or column in a matrix and return the column sums or row sums as a vector. If this is true counts_per_cell should have 1 entry per cell. Let’s make sure the length of the returned vector matches the matrix dimension for column. How would you do that? ( Hint:length() ).</p>
<p>Notes:
1. Matrix::colSums is a way to force functions from the Matrix library to be used. There are many libraries that implement colSums, we are forcing the one from the Matrix library to be used here to make sure it handles the dgTmatrix (sparse matrix) correctly. This is good practice.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="data-wrangling-scrnaseq-1.html#cb13-1"></a><span class="kw">hist</span>(<span class="kw">log10</span>(counts_per_cell<span class="op">+</span><span class="dv">1</span>),<span class="dt">main=</span><span class="st">&#39;counts per cell&#39;</span>,<span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_counts-1.png" width="672" /></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="data-wrangling-scrnaseq-1.html#cb14-1"></a><span class="kw">hist</span>(<span class="kw">log10</span>(genes_per_cell<span class="op">+</span><span class="dv">1</span>), <span class="dt">main=</span><span class="st">&#39;genes per cell&#39;</span>, <span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_counts-2.png" width="672" /></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="data-wrangling-scrnaseq-1.html#cb15-1"></a><span class="kw">plot</span>(counts_per_cell, genes_per_cell, <span class="dt">log=</span><span class="st">&#39;xy&#39;</span>, <span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span>
<span id="cb15-2"><a href="data-wrangling-scrnaseq-1.html#cb15-2"></a><span class="kw">title</span>(<span class="st">&#39;counts vs genes per cell&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_counts-3.png" width="672" /></p>
<p>Here we see examples of plotting a new plot, the histogram. R makes this really easy with the hist function. We are also transforming the values to log10 before plotting, this is done with the log10 method. When logging count data, the + 1 is used to avoid log10(0) which is not defined.</p>
<p>Can you make a histogram of counts per gene in log10 scale?</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="data-wrangling-scrnaseq-1.html#cb16-1"></a><span class="kw">hist</span>(<span class="kw">log10</span>(counts_per_gene<span class="op">+</span><span class="dv">1</span>), <span class="dt">main=</span><span class="st">&#39;counts per gene&#39;</span>, <span class="dt">col=</span><span class="st">&#39;wheat&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/hist_genes-1.png" width="672" /></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="data-wrangling-scrnaseq-1.html#cb17-1"></a><span class="co">### hist(?(?+1), main=&#39;counts per gene&#39;, col=&#39;wheat&#39;)</span></span></code></pre></div>
</div>
<div id="plot-cells-ranked-by-their-number-of-detected-genes." class="section level3">
<h3><span class="header-section-number">4.3.2</span> Plot cells ranked by their number of detected genes.</h3>
<p>Here we rank each cell by its library complexity, ie the number of genes detected per cell. This is a very useful plot as it shows the distribution of library complexity in the sequencing run. One can use this plot to investigate observations (potential cells) that are actually failed libraries (lower end outliers) or observations that are cell doublets (higher end outliers).</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="data-wrangling-scrnaseq-1.html#cb18-1"></a><span class="kw">plot</span>(<span class="kw">sort</span>(genes_per_cell), <span class="dt">xlab=</span><span class="st">&#39;cell&#39;</span>, <span class="dt">log=</span><span class="st">&#39;y&#39;</span>, <span class="dt">main=</span><span class="st">&#39;genes per cell (ordered)&#39;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/rank_cells-1.png" width="672" /></p>
</div>
</div>
<div id="beginning-with-seurat" class="section level2">
<h2><span class="header-section-number">4.4</span> Beginning with Seurat:</h2>
<p><a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a></p>
<div id="creating-a-seurat-object" class="section level3">
<h3><span class="header-section-number">4.4.1</span> Creating a seurat object</h3>
<p>To analyze our single cell data we will use a seurat object. Can you create an Seurat object with the 10x data and save it in an object called ‘seurat’? hint: CreateSeuratObject().
Can you include only genes that are are expressed in 3 or more cells and cells with complexity of 350 genes or more?
How many genes are you left with? How many cells?</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="data-wrangling-scrnaseq-1.html#cb19-1"></a><span class="co">### seurat&lt;-CreateSeuratObject(raw.data = counts, ? = 3, ? = 350, project = &quot;10X_NSCLC&quot;)</span></span>
<span id="cb19-2"><a href="data-wrangling-scrnaseq-1.html#cb19-2"></a>seurat &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(<span class="dt">counts =</span> counts, <span class="dt">min.cells =</span> <span class="dv">3</span>, <span class="dt">min.features =</span> <span class="dv">350</span>, <span class="dt">project =</span> <span class="st">&quot;10X_NSCLC&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)</code></pre>
<div id="object-slots" class="section level4">
<h4><span class="header-section-number">4.4.1.1</span> Object Slots</h4>
<p><a href="https://github.com/satijalab/seurat/wiki/Seurat#slots">reference</a></p>
<p>Almost all our analysis will be on the single object, of class Seurat. This object contains various “slots” (designated by <a href="mailto:seurat@slotname" class="email">seurat@slotname</a>) that will store not only the raw count data, but also the results from various computations below. This has the advantage that we do not need to keep track of inidividual variables of interest - they can all be collapsed into a single object as long as these slots are pre-defined as:</p>
<ul>
<li>assays - A list of assays within this object</li>
<li>meta.data - Cell-level meta data</li>
<li>active.assay - Name of active, or default, assay</li>
<li>active.ident - Identity classes for the current object</li>
<li>graphs - A list of nearest neighbor graphs</li>
<li>reductions - A list of DimReduc objects</li>
<li>project.name - User-defined project name (optional)</li>
<li>tools - Empty list. Tool developers can store any internal data from their methods here</li>
<li>misc - Empty slot. User can store additional information here
version Seurat version used when creating the object</li>
</ul>
<p><code>seurat@assays$RNA@counts</code> is a slot that stores the original gene count matrix. We can view the first 10 rows (genes) and the first 10 columns (cells).</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="data-wrangling-scrnaseq-1.html#cb21-1"></a>seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>counts[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;</code></pre>
<pre><code>##    [[ suppressing 10 column names &#39;AAACCTGAGCTAGTCT&#39;, &#39;AAACCTGAGGGCACTA&#39;, &#39;AAACCTGAGTACGTTC&#39; ... ]]</code></pre>
<pre><code>##                                  
## FO538757.2    . . . . 1 2 . . . .
## RP4-669L17.10 . . . . . . . . . .
## RP11-206L10.9 . . . . . . . . . .
## LINC00115     . . . . . . . . . .
## NOC2L         . . . . 2 4 . 1 . .
## KLHL17        . . . . . . . . . .
## PLEKHN1       . . . . . . . . . .
## HES4          . . . . . 9 . . . 1
## ISG15         . . . . 1 1 . . 1 .
## AGRN          . . . . . . . . . .</code></pre>
</div>
<div id="object-information" class="section level4">
<h4><span class="header-section-number">4.4.1.2</span> Object Information</h4>
<p>Summary information about Seurat objects can be had quickly and easily using standard R functions. Object shape/dimensions can be found using the dim, ncol, and nrow functions; cell and feature names can be found using the colnames and rownames functions, respectively, or the dimnames function. A vector of names of Assay, DimReduc, and Graph objects contained in a Seurat object can be had by using names.</p>
</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="expression-qc-and-normalization-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="the-following-examples-use-the-pbmc-3k-dataset.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["KrumlovSingleCellWorkshop2020.pdf", "KrumlovSingleCellWorkshop2020.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
