<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="6 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="6 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/" />
  <meta property="og:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />
  
  <meta name="github-repo" content="broadinstitute/KrumlovSingleCellWorkship2020" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  
  <meta name="twitter:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />

<meta name="author" content="Kirk Gosik" />


<meta name="date" content="2020-01-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="identifying-cell-populations-1.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#introduction-1"><i class="fa fa-check"></i><b>1.5.1</b> Introduction</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#transcriptome-quantification"><i class="fa fa-check"></i><b>1.5.2</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#expression-qc-and-normalization"><i class="fa fa-check"></i><b>1.5.3</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#data-wrangling-scrnaseq"><i class="fa fa-check"></i><b>1.5.4</b> Data Wrangling scRNAseq</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#identifying-cell-populations"><i class="fa fa-check"></i><b>1.5.5</b> Identifying Cell Populations</a></li>
<li class="chapter" data-level="1.5.6" data-path="index.html"><a href="index.html#feature-selection-and-cluster-analysis"><i class="fa fa-check"></i><b>1.5.6</b> Feature Selection and Cluster Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html"><i class="fa fa-check"></i><b>2</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="3" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html"><i class="fa fa-check"></i><b>3</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html"><i class="fa fa-check"></i><b>4</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#goal"><i class="fa fa-check"></i><b>4.1</b> Goal</a></li>
<li class="chapter" data-level="4.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#introduction-2"><i class="fa fa-check"></i><b>4.2</b> Introduction</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#load-necessary-packages"><i class="fa fa-check"></i><b>4.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>4.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>4.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>4.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>4.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>4.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>4.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>4.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#beginning-with-seurat-httpsatijalab.orgseurat"><i class="fa fa-check"></i><b>4.4</b> Beginning with Seurat: <span>http://satijalab.org/seurat/</span></a><ul>
<li class="chapter" data-level="4.4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>4.4.1</b> Creating a seurat object</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>4.5</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="4.6" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>4.6</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="4.6.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>4.6.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>4.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="4.8" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>4.8</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html"><i class="fa fa-check"></i><b>5</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="5.1" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html#google-slides"><i class="fa fa-check"></i><b>5.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html"><i class="fa fa-check"></i><b>6</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#abstract"><i class="fa fa-check"></i><b>6.1</b> Abstract</a></li>
<li class="chapter" data-level="6.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#seurat-tutorial"><i class="fa fa-check"></i><b>6.2</b> Seurat Tutorial</a><ul>
<li class="chapter" data-level="6.2.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#preprocessing-steps"><i class="fa fa-check"></i><b>6.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="6.2.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>6.2.2</b> Start of Identifying Cell Types</a></li>
<li class="chapter" data-level="6.2.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#run-non-linear-dimensional-reduction-umaptsne"><i class="fa fa-check"></i><b>6.2.3</b> Run non-linear dimensional reduction (UMAP/tSNE)</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#feature-selection"><i class="fa fa-check"></i><b>6.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="6.3.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="6.3.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.3.2</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="6.3.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#independent-components-analysis-ica"><i class="fa fa-check"></i><b>6.3.3</b> Independent Components Analysis (ICA)</a></li>
<li class="chapter" data-level="6.3.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#clustering"><i class="fa fa-check"></i><b>6.3.4</b> Clustering</a></li>
<li class="chapter" data-level="6.3.5" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#check-clusters"><i class="fa fa-check"></i><b>6.3.5</b> Check Clusters</a></li>
<li class="chapter" data-level="6.3.6" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#view-entire-object-structure"><i class="fa fa-check"></i><b>6.3.6</b> View Entire Object Structure</a></li>
<li class="chapter" data-level="6.3.7" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>6.3.7</b> Practice Visualizing/Embedding</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#other-options-for-analysis"><i class="fa fa-check"></i><b>6.4</b> Other Options For Analysis</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="feature-selection-and-cluster-analysis-1" class="section level1">
<h1><span class="header-section-number">6</span> Feature Selection and Cluster Analysis</h1>
<div id="abstract" class="section level2">
<h2><span class="header-section-number">6.1</span> Abstract</h2>
<p>Many methods have been used to determine differential gene expression from single-cell RNA (scRNA)-seq data. We evaluated 36 approaches using experimental and synthetic data and found considerable differences in the number and characteristics of the genes that are called differentially expressed. Prefiltering of lowly expressed genes has important effects, particularly for some of the methods developed for bulk RNA-seq data analysis. However, we found that bulk RNA-seq analysis methods do not generally perform worse than those developed specifically for scRNA-seq. We also present conquer, a repository of consistently processed, analysis-ready public scRNA-seq data sets that is aimed at simplifying method evaluation and reanalysis of published results. Each data set provides abundance estimates for both genes and transcripts, as well as quality control and exploratory analysis reports. <span class="citation">(<span class="citeproc-not-found" data-reference-id="soneson2018bias"><strong>???</strong></span>)</span></p>
<p>Cells are the basic building blocks of organisms and each cell is unique. Single-cell RNA sequencing has emerged as an indispensable tool to dissect the cellular heterogeneity and decompose tissues into cell types and/or cell states, which offers enormous potential for de novo discovery. Single-cell transcriptomic atlases provide unprecedented resolution to reveal complex cellular events and deepen our understanding of biological systems. In this review, we summarize and compare single-cell RNA sequencing technologies, that were developed since 2009, to facilitate a well-informed choice of method. The applications of these methods in different biological contexts are also discussed. We anticipate an ever-increasing role of single-cell RNA sequencing in biology with further improvement in providing spatial information and coupling to other cellular modalities. In the future, such biological findings will greatly benefit medical research. <span class="citation">(<span class="citeproc-not-found" data-reference-id="hedlund2018single"><strong>???</strong></span>)</span></p>
</div>
<div id="seurat-tutorial" class="section level2">
<h2><span class="header-section-number">6.2</span> Seurat Tutorial</h2>
<p>Task: Change the directory name to mydir/ where you saved your data</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" title="1"><span class="kw">library</span>(Seurat)</a>
<a class="sourceLine" id="cb56-2" title="2"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb56-3" title="3"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb56-4" title="4"><span class="kw">library</span>(CountClust)</a>
<a class="sourceLine" id="cb56-5" title="5"></a>
<a class="sourceLine" id="cb56-6" title="6">dirname &lt;-<span class="st"> &quot;/Users/kgosik/Documents/data/cellranger/pbmc3k/&quot;</span></a>
<a class="sourceLine" id="cb56-7" title="7">counts_matrix_filename =<span class="st"> </span><span class="kw">paste0</span>(dirname,<span class="st">&quot;/filtered_gene_bc_matrices/hg19/&quot;</span>)</a>
<a class="sourceLine" id="cb56-8" title="8">counts &lt;-<span class="st"> </span><span class="kw">Read10X</span>(<span class="dt">data.dir =</span> counts_matrix_filename)  <span class="co"># Seurat function to read in 10x count data</span></a>
<a class="sourceLine" id="cb56-9" title="9"></a>
<a class="sourceLine" id="cb56-10" title="10"><span class="co"># To minimize memory use on the docker - choose only the first 1000 cells</span></a>
<a class="sourceLine" id="cb56-11" title="11">counts &lt;-<span class="st"> </span>counts[,<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>]</a></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" title="1"><span class="co">### seurat&lt;-CreateSeuratObject(raw.data = counts, ? = 3, ? = 350, project = &quot;10X_NSCLC&quot;)</span></a>
<a class="sourceLine" id="cb57-2" title="2">seurat &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(<span class="dt">counts =</span> counts, <span class="dt">min.cells =</span> <span class="dv">3</span>, <span class="dt">min.features =</span> <span class="dv">350</span>, <span class="dt">project =</span> <span class="st">&quot;10X_NSCLC&quot;</span>)</a></code></pre></div>
<div id="preprocessing-steps" class="section level3">
<h3><span class="header-section-number">6.2.1</span> Preprocessing Steps</h3>
<p>This was all covered in the last Lab!</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" title="1"><span class="co"># The number of genes and UMIs (nGene and nUMI) are automatically calculated</span></a>
<a class="sourceLine" id="cb58-2" title="2"><span class="co"># for every object by Seurat.  For non-UMI data, nUMI represents the sum of</span></a>
<a class="sourceLine" id="cb58-3" title="3"><span class="co"># the non-normalized values within a cell We calculate the percentage of</span></a>
<a class="sourceLine" id="cb58-4" title="4"><span class="co"># mitochondrial genes here and store it in percent.mito using AddMetaData.</span></a>
<a class="sourceLine" id="cb58-5" title="5"><span class="co"># We use object@raw.data since this represents non-transformed and</span></a>
<a class="sourceLine" id="cb58-6" title="6"><span class="co"># non-log-normalized counts The % of UMI mapping to MT-genes is a common</span></a>
<a class="sourceLine" id="cb58-7" title="7"><span class="co"># scRNA-seq QC metric.</span></a>
<a class="sourceLine" id="cb58-8" title="8">seurat[[<span class="st">&quot;percent.mt&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">PercentageFeatureSet</span>(<span class="dt">object =</span> seurat, <span class="dt">pattern =</span> <span class="st">&quot;^MT-&quot;</span>)</a>
<a class="sourceLine" id="cb58-9" title="9"><span class="co">##VlnPlot(object = seurat, features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;), ncol = 3)</span></a></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" title="1"><span class="co"># GenePlot is typically used to visualize gene-gene relationships, but can</span></a>
<a class="sourceLine" id="cb59-2" title="2"><span class="co"># be used for anything calculated by the object, i.e. columns in</span></a>
<a class="sourceLine" id="cb59-3" title="3"><span class="co"># object@meta.data, PC scores etc.  Since there is a rare subset of cells</span></a>
<a class="sourceLine" id="cb59-4" title="4"><span class="co"># with an outlier level of high mitochondrial percentage and also low UMI</span></a>
<a class="sourceLine" id="cb59-5" title="5"><span class="co"># content, we filter these as well</span></a>
<a class="sourceLine" id="cb59-6" title="6"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb59-7" title="7"><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> seurat, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;percent.mt&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/scatter_plot-1.png" width="672" /></p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb60-1" title="1"><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> seurat, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;nFeature_RNA&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/scatter_plot-2.png" width="672" /></p>
<p><strong>Already A Subset of the Data</strong></p>
<p>This is not run becuase we are already working with a subset. This is here for reference.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" title="1"><span class="co"># We filter out cells that have unique gene counts over 2,500 or less than</span></a>
<a class="sourceLine" id="cb61-2" title="2"><span class="co"># 200 Note that low.thresholds and high.thresholds are used to define a</span></a>
<a class="sourceLine" id="cb61-3" title="3"><span class="co"># &#39;gate&#39;.  -Inf and Inf should be used if you don&#39;t want a lower or upper</span></a>
<a class="sourceLine" id="cb61-4" title="4"><span class="co"># threshold.</span></a>
<a class="sourceLine" id="cb61-5" title="5"></a>
<a class="sourceLine" id="cb61-6" title="6">seurat &lt;-<span class="st"> </span><span class="kw">SubsetData</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb61-7" title="7">                    <span class="dt">subset.names =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;percent.mt&quot;</span>), </a>
<a class="sourceLine" id="cb61-8" title="8">                    <span class="dt">low.thresholds =</span> <span class="kw">c</span>(<span class="dv">200</span>, <span class="op">-</span><span class="ot">Inf</span>), </a>
<a class="sourceLine" id="cb61-9" title="9">                    <span class="dt">high.thresholds =</span> <span class="kw">c</span>(<span class="dv">2500</span>, <span class="fl">0.1</span>))</a></code></pre></div>
<pre><code>## Warning: &#39;SubsetData&#39; is deprecated.
## Use &#39;subset&#39; instead.
## See help(&quot;Deprecated&quot;)</code></pre>
<pre><code>## Warning: &#39;OldWhichCells&#39; is deprecated.
## Use &#39;WhichCells&#39; instead.
## See help(&quot;Deprecated&quot;)</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" title="1">seurat &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> seurat, <span class="dt">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="dt">scale.factor =</span> <span class="dv">10000</span>)</a></code></pre></div>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" title="1">seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb65-2" title="2">                               <span class="dt">mean.function =</span> ExpMean, </a>
<a class="sourceLine" id="cb65-3" title="3">                               <span class="dt">dispersion.function =</span> LogVMR, </a>
<a class="sourceLine" id="cb65-4" title="4">                               <span class="dt">do.plot =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## Warning: The following arguments are not used: do.plot</code></pre>
</div>
<div id="start-of-identifying-cell-types" class="section level3">
<h3><span class="header-section-number">6.2.2</span> Start of Identifying Cell Types</h3>
<div id="scaling" class="section level4">
<h4><span class="header-section-number">6.2.2.1</span> Scaling</h4>
<p>This part is where you mean center the data, substract the mean. You also divide by the standard deviation to make everything to a ‘standard normal’, where the mean is zero and the standard deviation is 1.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" title="1">seurat &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(<span class="dt">object =</span> seurat, <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">&quot;batch&quot;</span>, <span class="st">&quot;percent.mt&quot;</span>))</a></code></pre></div>
<pre><code>## Regressing out batch, percent.mt</code></pre>
<pre><code>## Centering and scaling data matrix</code></pre>
<p>Try Regressing Other Variables</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb70-1" title="1"><span class="co">## randomly making a batch id data.frame</span></a>
<a class="sourceLine" id="cb70-2" title="2">batch_ids &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">barcode =</span> <span class="kw">rownames</span>(seurat<span class="op">@</span>meta.data), </a>
<a class="sourceLine" id="cb70-3" title="3">                        <span class="dt">batch_id =</span> <span class="kw">sample</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">2</span>, <span class="kw">NROW</span>(seurat<span class="op">@</span>meta.data), <span class="dt">replace =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb70-4" title="4">                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb70-5" title="5"></a>
<a class="sourceLine" id="cb70-6" title="6"></a>
<a class="sourceLine" id="cb70-7" title="7"><span class="kw">row.names</span>(batch_ids) &lt;-<span class="st"> </span><span class="kw">row.names</span>(seurat<span class="op">@</span>meta.data)</a>
<a class="sourceLine" id="cb70-8" title="8">seurat &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> seurat, <span class="dt">metadata =</span> batch_ids, <span class="dt">col.name =</span> <span class="ot">NULL</span>)</a>
<a class="sourceLine" id="cb70-9" title="9">seurat &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(<span class="dt">object =</span> seurat, <span class="dt">vars.to.regress =</span> <span class="st">&#39;batch_id&#39;</span>)</a></code></pre></div>
<pre><code>## Regressing out batch_id</code></pre>
<pre><code>## Centering and scaling data matrix</code></pre>
</div>
<div id="perform-linear-dimensional-reduction-pca" class="section level4">
<h4><span class="header-section-number">6.2.2.2</span> Perform linear dimensional reduction (PCA)</h4>
<p>This will run pca on the</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb73-1" title="1">seurat &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb73-2" title="2">               <span class="dt">pc.genes =</span> seurat<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>var.genes, </a>
<a class="sourceLine" id="cb73-3" title="3">               <span class="dt">do.print =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb73-4" title="4">               <span class="dt">pcs.print =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, </a>
<a class="sourceLine" id="cb73-5" title="5">               <span class="dt">genes.print =</span> <span class="dv">5</span>)</a></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  MALAT1, RPS27A, LTB, IL32, RPL10A, AES, GLTSCR2, RPL19, CD69, ACAP1 
##     B2M, CD27, STK17A, CTSW, CD247, JUN, GIMAP5, RARRES3, SELL, CCL5 
##     GZMA, CST7, CD8B, MYC, MAL, C12orf57, TRAF3IP3, CMPK1, TNFAIP8, GZMK 
## Negative:  CST3, AIF1, TYROBP, LST1, FTL, FTH1, FCER1G, S100A9, LYZ, CFD 
##     FCN1, CD68, TYMP, LGALS1, IFITM3, SPI1, SERPINA1, CTSS, PSAP, S100A8 
##     LGALS2, S100A11, SAT1, IFI30, CFP, COTL1, LGALS3, PYCARD, TKT, S100A6 
## PC_ 2 
## Positive:  NKG7, GZMB, CST7, PRF1, GZMA, FGFBP2, CTSW, B2M, GNLY, GZMH 
##     CCL5, SPON2, CD247, CCL4, XCL2, AKR1C3, FCGR3A, CLIC3, HOPX, SRGN 
##     IL32, TTC38, S100A4, APMAP, ANXA1, RARRES3, CD99, PRSS23, TPST2, HAVCR2 
## Negative:  CD79A, MS4A1, HLA-DQA1, HLA-DRA, HLA-DQB1, TCL1A, LINC00926, CD79B, VPREB3, CD74 
##     HLA-DMA, HLA-DRB1, HLA-DPB1, HLA-DMB, BANK1, HLA-DPA1, HLA-DQA2, HLA-DRB5, CD37, EAF2 
##     KIAA0125, LTB, SPIB, IRF8, BLNK, IGLL5, PPAPDC1B, RPL10A, RPL19, SWAP70 
## PC_ 3 
## Positive:  HIST1H2AC, PF4, TSC22D1, ACRBP, PPBP, TMEM40, PTCRA, MAP3K7CL, SDPR, MMD 
##     TUBB1, HIST1H2BJ, NRGN, GNG11, ITGA2B, ESAM, CLDN5, CMTM5, SPARC, SLC40A1 
##     RGS18, NEXN, RUFY1, PCP2, GP9, F13A1, CD9, PNMA1, CA2, ARG2 
## Negative:  RPL19, MT-CO1, S100A8, S100A4, LGALS2, S100A9, S100A6, TYROBP, CYBA, TSPO 
##     NACA, CCL3, FCN1, MS4A6A, CD14, GSTP1, NKG7, PRF1, IFI6, LGALS1 
##     CST7, LYZ, GZMA, NEAT1, FGFBP2, NCF2, MALAT1, RPL10A, CDA, VIM 
## PC_ 4 
## Positive:  VIM, FYB, TMSB4X, RGS10, S100A8, IL32, GIMAP7, S100A6, S100A4, MAL 
##     NDFIP1, NGFRAP1, CD27, GIMAP5, S100A9, CD14, S100A12, NELL2, CD44, LGALS2 
##     S100A11, MS4A6A, TRABD2A, LYZ, C12orf57, FCN1, AIF1, CD5, TSPO, ANXA1 
## Negative:  HLA-DQA1, CD79A, CD79B, MS4A1, HLA-DQB1, CD74, HLA-DPB1, HLA-DRB1, HLA-DPA1, HLA-DRA 
##     LINC00926, HLA-DQA2, TCL1A, HLA-DRB5, VPREB3, GZMB, FGFBP2, HLA-DMB, BANK1, GNLY 
##     PRF1, NKG7, HLA-DMA, CST7, PDLIM1, KIAA0125, SPIB, SPON2, EAF2, FCGR3A 
## PC_ 5 
## Positive:  LTB, NAP1L1, CKB, VMO1, MS4A4A, EMR2, CDKN1C, NACA, HES4, LYPD2 
##     PILRA, MS4A7, TRAF3IP3, PTGES3, LILRA3, GDI2, VIM, RP11-290F20.3, RPS27A, RPL10A 
##     TUBA1B, CXCL16, IFITM2, CYTIP, ANXA5, CD27, CTD-2006K23.1, ICAM4, ABI3, TPPP3 
## Negative:  S100A8, LGALS2, GZMB, MS4A6A, S100A9, FGFBP2, GNLY, CD14, NKG7, S100A12 
##     SPON2, CST7, CCL3, PRF1, GSTP1, GPX1, CCL4, VCAN, GZMA, TYROBP 
##     PRSS23, CCL5, AKR1C3, XCL2, GZMH, LYZ, IGFBP7, FCN1, CYBA, HAVCR2</code></pre>
</div>
<div id="perform-linear-dimensional-reduction-ica" class="section level4">
<h4><span class="header-section-number">6.2.2.3</span> Perform linear dimensional reduction (ICA)</h4>
<p>Try running Independent Component Analysis. If you need help with the inputs try using the ?RunICA menu.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb75-1" title="1">seurat &lt;-<span class="st"> </span><span class="kw">RunICA</span>(seurat, <span class="dt">nics =</span> <span class="dv">5</span>)</a></code></pre></div>
<pre><code>## IC_ 1 
## Positive:  IL32, MALAT1, RPS27A, LTB, AES, B2M, CD27, GIMAP7, ACAP1, GIMAP5 
##     RPL10A, RARRES3, TRAF3IP3, GLTSCR2, MAL, CD247, JUN, C12orf57, CD69, RPL19 
##     ITM2A, NGFRAP1, CD8B, NDFIP1, TNFAIP8, NELL2, TRABD2A, TUBA4A, CMPK1, STK17A 
## Negative:  HLA-DRA, HLA-DRB1, TYROBP, CST3, S100A9, FTL, LYZ, LGALS2, S100A8, HLA-DPA1 
##     FCN1, HLA-DRB5, CD74, HLA-DPB1, FCER1G, FTH1, LST1, AIF1, CTSS, TYMP 
##     LGALS1, CYBA, CFD, HLA-DMA, SPI1, CD68, GSTP1, PSAP, PYCARD, MS4A6A 
## IC_ 2 
## Positive:  S100A4, S100A6, S100A8, S100A9, TMSB4X, S100A11, TYROBP, LGALS1, FCN1, LYZ 
##     AIF1, CST3, LGALS2, FCER1G, LST1, CFD, TYMP, CD14, FTL, LGALS3 
##     MS4A6A, FTH1, SRGN, CD68, NEAT1, PSAP, GAPDH, TIMP1, GSTP1, VIM 
## Negative:  CD79A, MS4A1, CD79B, HLA-DQA1, LINC00926, TCL1A, HLA-DQB1, VPREB3, BANK1, HLA-DQA2 
##     KIAA0125, SPIB, CD74, LTB, PDLIM1, HLA-DPB1, PPAPDC1B, HLA-DMB, BLNK, EAF2 
##     IGLL5, HLA-DRA, SWAP70, BLK, HLA-DPA1, CD37, RPS27A, HLA-DRB1, AC079767.4, MZB1 
## IC_ 3 
## Positive:  MALAT1, TCL1A, S100A12, RPS27A, LINC00926, S100A8, CCL5, RPL19, RPL10A, MS4A1 
##     VCAN, CD79A, CTSW, KCNG1, VPREB3, AC003075.4, STK17A, CST7, JUND, MYL9 
##     SELL, MS4A6A, SEC14L2, KLRC1, MGST1, CD69, RP5-821D11.7, PRSS23, IER3, CD8B 
## Negative:  FCGR3A, PILRA, RP11-290F20.3, RHOC, MS4A7, HES4, LRRC25, CKB, SERPINA1, ABI3 
##     COTL1, IFITM3, MS4A4A, EMR2, CDKN1C, VMO1, SPI1, LILRA3, CEBPB, WARS 
##     HCK, LST1, BID, IFITM2, ANXA5, TIMP1, CXCL16, ACTB, TPPP3, FCER1G 
## IC_ 4 
## Positive:  GZMB, NKG7, FGFBP2, CST7, PRF1, GNLY, GZMA, SPON2, CTSW, GZMH 
##     CCL4, XCL2, AKR1C3, CCL5, CLIC3, B2M, HOPX, PRSS23, TTC38, CD247 
##     HAVCR2, APMAP, KLRF1, XCL1, GPR56, FCRL6, IGFBP7, FCGR3A, TPST2, CD160 
## Negative:  LTB, VIM, AIF1, FTH1, COTL1, RPL10A, LST1, RPL19, EEF2, FTL 
##     LYZ, NAP1L1, S100A8, HLA-DRA, PLP2, CFP, CST3, CD68, CFD, FCN1 
##     LGALS2, C6orf48, S100A9, GPX1, SERPINA1, CTSS, LINC00936, SAT1, SPI1, FAM26F 
## IC_ 5 
## Positive:  PF4, HIST1H2AC, PPBP, TSC22D1, ACRBP, TMEM40, PTCRA, NRGN, MAP3K7CL, SDPR 
##     GNG11, MMD, ITGA2B, HIST1H2BJ, TUBB1, ESAM, CLDN5, SPARC, CMTM5, RGS18 
##     SLC40A1, F13A1, RUFY1, PCP2, GP9, NEXN, CD9, CA2, MPP1, PNMA1 
## Negative:  RPL19, RPL10A, MALAT1, NACA, MT-CO1, RPS27A, GLTSCR2, VIM, SRSF3, B2M 
##     RBM3, TRAF3IP3, AES, CSNK2B, RAC2, HSP90AB1, NHP2L1, KLF6, VAMP8, EEF2 
##     TSPO, NDUFA11, C1QBP, CDC37, IL32, PSMC5, ZFP36, MED28, COPS6, ISG15</code></pre>
</div>
<div id="visualizing-pca-in-different-ways" class="section level4">
<h4><span class="header-section-number">6.2.2.4</span> Visualizing PCA in Different Ways</h4>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb77-1" title="1"><span class="kw">DimPlot</span>(seurat, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_pca-1.png" width="672" /></p>
</div>
<div id="visualizing-ica-in-different-ways" class="section level4">
<h4><span class="header-section-number">6.2.2.5</span> Visualizing ICA in Different Ways</h4>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" title="1"><span class="kw">DimPlot</span>(seurat, <span class="dt">reduction =</span> <span class="st">&quot;ica&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_ica-1.png" width="672" /></p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb79-1" title="1"><span class="co"># ProjectDim scores each gene in the dataset (including genes not included</span></a>
<a class="sourceLine" id="cb79-2" title="2"><span class="co"># in the PCA) based on their correlation with the calculated components.</span></a>
<a class="sourceLine" id="cb79-3" title="3"><span class="co"># Though we don&#39;t use this further here, it can be used to identify markers</span></a>
<a class="sourceLine" id="cb79-4" title="4"><span class="co"># that are strongly correlated with cellular heterogeneity, but may not have</span></a>
<a class="sourceLine" id="cb79-5" title="5"><span class="co"># passed through variable gene selection.  The results of the projected PCA</span></a>
<a class="sourceLine" id="cb79-6" title="6"><span class="co"># can be explored by setting use.full=T in the functions above</span></a>
<a class="sourceLine" id="cb79-7" title="7">seurat &lt;-<span class="st"> </span><span class="kw">ProjectDim</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  MALAT1, RPS27A, LTB, IL32, RPL10A, AES, GLTSCR2, RPL19, CD69, ACAP1 
##     B2M, CD27, STK17A, CTSW, CD247, JUN, GIMAP5, RARRES3, SELL, CCL5 
## Negative:  CST3, AIF1, TYROBP, LST1, FTL, FTH1, FCER1G, S100A9, LYZ, CFD 
##     FCN1, CD68, TYMP, LGALS1, IFITM3, SPI1, SERPINA1, CTSS, PSAP, S100A8 
## PC_ 2 
## Positive:  NKG7, GZMB, CST7, PRF1, GZMA, FGFBP2, CTSW, B2M, GNLY, GZMH 
##     CCL5, SPON2, CD247, CCL4, XCL2, AKR1C3, FCGR3A, CLIC3, HOPX, SRGN 
## Negative:  CD79A, MS4A1, HLA-DQA1, HLA-DRA, HLA-DQB1, TCL1A, LINC00926, CD79B, VPREB3, CD74 
##     HLA-DMA, HLA-DRB1, HLA-DPB1, HLA-DMB, BANK1, HLA-DPA1, HLA-DQA2, HLA-DRB5, CD37, EAF2 
## PC_ 3 
## Positive:  HIST1H2AC, PF4, TSC22D1, ACRBP, PPBP, TMEM40, PTCRA, MAP3K7CL, SDPR, MMD 
##     TUBB1, HIST1H2BJ, NRGN, GNG11, ITGA2B, ESAM, CLDN5, CMTM5, SPARC, SLC40A1 
## Negative:  RPL19, MT-CO1, S100A8, S100A4, LGALS2, S100A9, S100A6, TYROBP, CYBA, TSPO 
##     NACA, CCL3, FCN1, MS4A6A, CD14, GSTP1, NKG7, PRF1, IFI6, LGALS1 
## PC_ 4 
## Positive:  VIM, FYB, TMSB4X, RGS10, S100A8, IL32, GIMAP7, S100A6, S100A4, MAL 
##     NDFIP1, NGFRAP1, CD27, GIMAP5, S100A9, CD14, S100A12, NELL2, CD44, LGALS2 
## Negative:  HLA-DQA1, CD79A, CD79B, MS4A1, HLA-DQB1, CD74, HLA-DPB1, HLA-DRB1, HLA-DPA1, HLA-DRA 
##     LINC00926, HLA-DQA2, TCL1A, HLA-DRB5, VPREB3, GZMB, FGFBP2, HLA-DMB, BANK1, GNLY 
## PC_ 5 
## Positive:  LTB, NAP1L1, CKB, VMO1, MS4A4A, EMR2, CDKN1C, NACA, HES4, LYPD2 
##     PILRA, MS4A7, TRAF3IP3, PTGES3, LILRA3, GDI2, VIM, RP11-290F20.3, RPS27A, RPL10A 
## Negative:  S100A8, LGALS2, GZMB, MS4A6A, S100A9, FGFBP2, GNLY, CD14, NKG7, S100A12 
##     SPON2, CST7, CCL3, PRF1, GSTP1, GPX1, CCL4, VCAN, GZMA, TYROBP</code></pre>
</div>
<div id="genes-by-pcs" class="section level4">
<h4><span class="header-section-number">6.2.2.6</span> Genes by PCs</h4>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb81-1" title="1"><span class="kw">DimHeatmap</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb81-2" title="2">          <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, </a>
<a class="sourceLine" id="cb81-3" title="3">          <span class="dt">cells =</span> <span class="dv">50</span>, </a>
<a class="sourceLine" id="cb81-4" title="4">          <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</a>
<a class="sourceLine" id="cb81-5" title="5">          <span class="dt">balanced =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/pca_heatmap-1.png" width="672" /></p>
<p>Check other PCs to plot</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb82-1" title="1"><span class="kw">PCHeatmap</span>()</a></code></pre></div>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb83-1" title="1">seurat &lt;-<span class="st"> </span><span class="kw">JackStraw</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a>
<a class="sourceLine" id="cb83-2" title="2">seurat &lt;-<span class="st"> </span><span class="kw">ScoreJackStraw</span>(seurat, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>)</a></code></pre></div>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb84-1" title="1"><span class="kw">ElbowPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">ndims =</span> <span class="dv">10</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/pca_elbow-1.png" width="672" /></p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" title="1"><span class="co"># save.SNN = T saves the SNN so that the clustering algorithm can be rerun</span></a>
<a class="sourceLine" id="cb85-2" title="2"><span class="co"># using the same graph but with a different resolution value (see docs for</span></a>
<a class="sourceLine" id="cb85-3" title="3"><span class="co"># full details)</span></a>
<a class="sourceLine" id="cb85-4" title="4"><span class="kw">set.seed</span>(<span class="dv">2019</span>)</a>
<a class="sourceLine" id="cb85-5" title="5">seurat &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(<span class="dt">object =</span> seurat, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</a></code></pre></div>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" title="1">seurat &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb88-2" title="2">                       <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, </a>
<a class="sourceLine" id="cb88-3" title="3">                       <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, </a>
<a class="sourceLine" id="cb88-4" title="4">                       <span class="dt">resolution =</span> <span class="dv">1</span>)</a></code></pre></div>
<pre><code>## Warning: The following arguments are not used: reduction, dims

## Warning: The following arguments are not used: reduction, dims</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 987
## Number of edges: 35554
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.7739
## Number of communities: 8
## Elapsed time: 0 seconds</code></pre>
</div>
</div>
<div id="run-non-linear-dimensional-reduction-umaptsne" class="section level3">
<h3><span class="header-section-number">6.2.3</span> Run non-linear dimensional reduction (UMAP/tSNE)</h3>
<p>Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb91-1" title="1"><span class="kw">set.seed</span>(<span class="kw">runif</span>(<span class="dv">100</span>))</a>
<a class="sourceLine" id="cb91-2" title="2">seurat &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(seurat, <span class="dt">reduction.use =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">perplexity=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb91-3" title="3"></a>
<a class="sourceLine" id="cb91-4" title="4"><span class="co"># note that you can set do.label=T to help label individual clusters</span></a>
<a class="sourceLine" id="cb91-5" title="5"><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/run_tsne-1.png" width="672" /></p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" title="1"><span class="kw">set.seed</span>(<span class="kw">runif</span>(<span class="dv">100</span>))</a>
<a class="sourceLine" id="cb92-2" title="2">seurat &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(seurat, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">perplexity=</span><span class="dv">10</span>)</a></code></pre></div>
<pre><code>## Warning: The following arguments are not used: perplexity</code></pre>
<pre><code>## 16:57:38 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>## 16:57:38 Read 987 rows and found 10 numeric columns</code></pre>
<pre><code>## 16:57:38 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>## 16:57:38 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>## 0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>## [----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>## **************************************************|
## 16:57:38 Writing NN index file to temp file /var/folders/_1/7d41wl1d0nv1vxmgmcbn5jfj84xmh7/T//RtmpHKteoJ/file1357d476ddc5b
## 16:57:38 Searching Annoy index using 1 thread, search_k = 3000
## 16:57:38 Annoy recall = 100%
## 16:57:39 Commencing smooth kNN distance calibration using 1 thread
## 16:57:40 Initializing from normalized Laplacian + noise
## 16:57:40 Commencing optimization for 500 epochs, with 38670 positive edges
## 16:57:42 Optimization finished</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" title="1"><span class="co"># note that you can set do.label=T to help label individual clusters</span></a>
<a class="sourceLine" id="cb101-2" title="2"><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/run_umap-1.png" width="672" /></p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb102-1" title="1"><span class="co"># find all markers of cluster 1</span></a>
<a class="sourceLine" id="cb102-2" title="2">cluster1.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">1</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>)</a>
<a class="sourceLine" id="cb102-3" title="3"><span class="kw">print</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> cluster1.markers, <span class="dt">n =</span> <span class="dv">5</span>))</a></code></pre></div>
<pre><code>##                p_val avg_logFC pct.1 pct.2     p_val_adj
## S100A8 1.586796e-147  3.698205 0.951 0.095 1.867024e-143
## LGALS2 4.746642e-147  2.612902 0.886 0.056 5.584899e-143
## S100A9 2.397414e-123  3.745692 0.978 0.217 2.820797e-119
## MS4A6A 1.419272e-115  1.881602 0.674 0.024 1.669916e-111
## FCN1   8.514604e-113  2.360585 0.924 0.154 1.001828e-108</code></pre>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb104-1" title="1"><span class="co"># find all markers distinguishing cluster 2 from clusters 0 and 1</span></a>
<a class="sourceLine" id="cb104-2" title="2">cluster5.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">5</span>, <span class="dt">ident.2 =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">min.pct =</span> <span class="fl">0.25</span>)</a>
<a class="sourceLine" id="cb104-3" title="3"><span class="kw">print</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> cluster5.markers, <span class="dt">n =</span> <span class="dv">5</span>))</a></code></pre></div>
<pre><code>##               p_val avg_logFC pct.1 pct.2    p_val_adj
## FCGR3A 3.353326e-65 2.6292920 0.968 0.089 3.945523e-61
## RHOC   2.286079e-49 1.7673220 0.921 0.119 2.689801e-45
## CDKN1C 1.882880e-39 0.9945063 0.460 0.007 2.215397e-35
## HES4   5.069482e-37 1.1358796 0.619 0.047 5.964753e-33
## CKB    1.398361e-35 0.9644735 0.365 0.000 1.645312e-31</code></pre>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb106-1" title="1"><span class="co"># find markers for every cluster compared to all remaining cells, report</span></a>
<a class="sourceLine" id="cb106-2" title="2"><span class="co"># only the positive ones</span></a>
<a class="sourceLine" id="cb106-3" title="3">pbmc.markers &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">only.pos =</span> <span class="ot">TRUE</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, <span class="dt">thresh.use =</span> <span class="fl">0.25</span>)</a></code></pre></div>
<pre><code>## Calculating cluster 0</code></pre>
<pre><code>## Calculating cluster 1</code></pre>
<pre><code>## Calculating cluster 2</code></pre>
<pre><code>## Calculating cluster 3</code></pre>
<pre><code>## Calculating cluster 4</code></pre>
<pre><code>## Calculating cluster 5</code></pre>
<pre><code>## Calculating cluster 6</code></pre>
<pre><code>## Calculating cluster 7</code></pre>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb115-1" title="1">pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">2</span>, avg_logFC)</a></code></pre></div>
<pre><code>## # A tibble: 16 x 7
## # Groups:   cluster [8]
##        p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene  
##        &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt; 
##  1 1.20e- 43     1.06  0.529 0.115 1.41e- 39 0       CCR7  
##  2 1.71e- 29     0.864 0.407 0.094 2.01e- 25 0       LEF1  
##  3 1.59e-147     3.70  0.951 0.095 1.87e-143 1       S100A8
##  4 2.40e-123     3.75  0.978 0.217 2.82e-119 1       S100A9
##  5 6.05e- 34     0.866 0.947 0.465 7.12e- 30 2       IL32  
##  6 2.94e- 33     0.961 0.988 0.665 3.45e- 29 2       LTB   
##  7 1.47e-157     2.94  0.94  0.046 1.73e-153 3       CD79A 
##  8 8.48e- 99     2.46  0.632 0.029 9.98e- 95 3       TCL1A 
##  9 2.62e- 70     2.18  0.633 0.053 3.09e- 66 4       GZMK  
## 10 3.31e- 63     2.00  0.959 0.247 3.90e- 59 4       CCL5  
## 11 1.27e- 69     2.24  0.968 0.137 1.49e- 65 5       FCGR3A
## 12 1.02e- 45     2.01  0.968 0.314 1.20e- 41 5       LST1  
## 13 2.28e-115     3.45  1     0.066 2.69e-111 6       GZMB  
## 14 9.18e- 71     3.45  0.968 0.15  1.08e- 66 6       GNLY  
## 15 6.56e-  7     0.911 0.571 0.303 7.72e-  3 7       CCL5  
## 16 2.92e-  5     0.853 0.375 0.184 3.43e-  1 7       LYAR</code></pre>
<div id="finding-differentially-expressed-features-cluster-biomarkers" class="section level4">
<h4><span class="header-section-number">6.2.3.1</span> Finding differentially expressed features (cluster biomarkers)</h4>
<p>Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in ident.1), compared to all other cells. FindAllMarkers automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.</p>
<p>The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb117-1" title="1">cluster1.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb117-2" title="2">                                <span class="dt">ident.1 =</span> <span class="dv">0</span>, </a>
<a class="sourceLine" id="cb117-3" title="3">                                <span class="dt">thresh.use =</span> <span class="fl">0.25</span>, </a>
<a class="sourceLine" id="cb117-4" title="4">                                <span class="dt">test.use =</span> <span class="st">&quot;roc&quot;</span>, </a>
<a class="sourceLine" id="cb117-5" title="5">                                <span class="dt">only.pos =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## Warning: The following arguments are not used: thresh.use</code></pre>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb119-1" title="1"><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;CD79A&quot;</span>))</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_cluster1_markers-1.png" width="672" /></p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb120-1" title="1"><span class="co"># you can plot raw UMI counts as well</span></a>
<a class="sourceLine" id="cb120-2" title="2"><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb120-3" title="3">        <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;NKG7&quot;</span>, <span class="st">&quot;PF4&quot;</span>), </a>
<a class="sourceLine" id="cb120-4" title="4">        <span class="dt">use.raw =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb120-5" title="5">        <span class="dt">y.log =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## Warning in as_grob.default(plot): Cannot convert object of class logical
## into a grob.

## Warning in as_grob.default(plot): Cannot convert object of class logical
## into a grob.</code></pre>
<pre><code>## Warning: Graphs cannot be vertically aligned unless the axis parameter is
## set. Placing graphs unaligned.</code></pre>
<pre><code>## Warning: Graphs cannot be horizontally aligned unless the axis parameter is
## set. Placing graphs unaligned.</code></pre>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_by_umi-1.png" width="672" /></p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb124-1" title="1"><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb124-2" title="2">            <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;GNLY&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;FCER1A&quot;</span>, <span class="st">&quot;FCGR3A&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;PPBP&quot;</span>, <span class="st">&quot;CD8A&quot;</span>),</a>
<a class="sourceLine" id="cb124-3" title="3">            <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;blue&quot;</span>), </a>
<a class="sourceLine" id="cb124-4" title="4">            <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>)</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/gene_tsne-1.png" width="672" /></p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb125-1" title="1">top10 &lt;-<span class="st"> </span>pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">10</span>, avg_logFC)</a>
<a class="sourceLine" id="cb125-2" title="2"><span class="co"># setting slim.col.label to TRUE will print just the cluster IDS instead of</span></a>
<a class="sourceLine" id="cb125-3" title="3"><span class="co"># every cell name</span></a>
<a class="sourceLine" id="cb125-4" title="4"><span class="kw">DoHeatmap</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> top10<span class="op">$</span>gene, <span class="dt">label =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## Warning in DoHeatmap(object = seurat, features = top10$gene, label =
## TRUE): The following features were omitted as they were not found in the
## scale.data slot for the RNA assay: RPL35, LYAR, CD8A, CD40LG, AQP3, CD2,
## IL7R, PIK3IP1, NOSIP, PRKCQ-AS1, CD3D, CD3E, LEF1, CCR7, LDHB</code></pre>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/do_heatmap-1.png" width="672" /></p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" title="1">new.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Memory CD4 T&quot;</span>, <span class="st">&quot;CD14+ Mono&quot;</span>, <span class="st">&quot;Naive CD4 T&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;CD8 T&quot;</span>, <span class="st">&quot;FCGR3A+ Mono&quot;</span>, <span class="st">&quot;NK&quot;</span>, <span class="st">&quot;DC&quot;</span>, <span class="st">&quot;Mk&quot;</span>)</a>
<a class="sourceLine" id="cb127-2" title="2"><span class="kw">names</span>(<span class="dt">x =</span> new.cluster.ids) &lt;-<span class="st"> </span><span class="kw">levels</span>(<span class="dt">x =</span> seurat)</a>
<a class="sourceLine" id="cb127-3" title="3">seurat &lt;-<span class="st"> </span><span class="kw">RenameIdents</span>(<span class="dt">object =</span> seurat, new.cluster.ids)</a></code></pre></div>
<pre><code>## Warning: Cannot find identity NA</code></pre>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" title="1"><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&#39;umap&#39;</span>, <span class="dt">label =</span> <span class="ot">TRUE</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span><span class="kw">NoLegend</span>()</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/new_names-1.png" width="672" /></p>
</div>
<div id="further-subdivisions-within-cell-types" class="section level4">
<h4><span class="header-section-number">6.2.3.2</span> Further subdivisions within cell types</h4>
<p>If you perturb some of our parameter choices above (for example, setting resolution=0.8 or changing the number of PCs), you might see the CD4 T cells subdivide into two groups. You can explore this subdivision to find markers separating the two T cell subsets. However, before reclustering (which will overwrite <a href="mailto:object@ident" class="email">object@ident</a>), we can stash our renamed identities to be easily recovered later.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb130-1" title="1"><span class="co"># First lets stash our identities for later</span></a>
<a class="sourceLine" id="cb130-2" title="2">seurat &lt;-<span class="st"> </span><span class="kw">StashIdent</span>(<span class="dt">object =</span> seurat, <span class="dt">save.name =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>)</a></code></pre></div>
<pre><code>## With Seurat 3.X, stashing identity classes can be accomplished with the following:
## seurat[[&quot;ClusterNames_0.6&quot;]] &lt;- Idents(object = seurat)</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb132-1" title="1"><span class="co"># Note that if you set save.snn=T above, you don&#39;t need to recalculate the</span></a>
<a class="sourceLine" id="cb132-2" title="2"><span class="co"># SNN, and can simply put: pbmc &lt;- FindClusters(pbmc,resolution = 0.8)</span></a>
<a class="sourceLine" id="cb132-3" title="3">seurat &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb132-4" title="4">                       <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, </a>
<a class="sourceLine" id="cb132-5" title="5">                       <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, </a>
<a class="sourceLine" id="cb132-6" title="6">                       <span class="dt">resolution =</span> <span class="fl">0.8</span>)</a></code></pre></div>
<pre><code>## Warning: The following arguments are not used: reduction, dims

## Warning: The following arguments are not used: reduction, dims</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 987
## Number of edges: 35554
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8065
## Number of communities: 7
## Elapsed time: 0 seconds</code></pre>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" title="1"><span class="co">## Warning in BuildSNN(object = object, genes.use = genes.use, reduction.type</span></a>
<a class="sourceLine" id="cb135-2" title="2"><span class="co">## = reduction.type, : Build parameters exactly match those of already</span></a>
<a class="sourceLine" id="cb135-3" title="3"><span class="co">## computed and stored SNN. To force recalculation, set force.recalc to TRUE.</span></a>
<a class="sourceLine" id="cb135-4" title="4"><span class="co"># Demonstration of how to plot two tSNE plots side by side, and how to color</span></a>
<a class="sourceLine" id="cb135-5" title="5"><span class="co"># points based on different criteria</span></a>
<a class="sourceLine" id="cb135-6" title="6">plot1 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb135-7" title="7">                 <span class="dt">reduction=</span> <span class="st">&quot;tsne&quot;</span>,</a>
<a class="sourceLine" id="cb135-8" title="8">                 <span class="dt">label =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">NoLegend</span>()</a>
<a class="sourceLine" id="cb135-9" title="9"></a>
<a class="sourceLine" id="cb135-10" title="10">plot2 &lt;-<span class="st"> </span><span class="kw">TSNEPlot</span>(<span class="dt">object =</span> seurat, </a>
<a class="sourceLine" id="cb135-11" title="11">                  <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>, </a>
<a class="sourceLine" id="cb135-12" title="12">                  <span class="dt">group.by =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>,</a>
<a class="sourceLine" id="cb135-13" title="13">                  <span class="dt">label =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">NoLegend</span>()</a>
<a class="sourceLine" id="cb135-14" title="14"></a>
<a class="sourceLine" id="cb135-15" title="15"><span class="kw">CombinePlots</span>(<span class="kw">list</span>(plot1, plot2))</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/compare_params-1.png" width="672" /></p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb136-1" title="1"><span class="co"># Find discriminating markers</span></a>
<a class="sourceLine" id="cb136-2" title="2">tcell.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb136-3" title="3"></a>
<a class="sourceLine" id="cb136-4" title="4"><span class="co"># Most of the markers tend to be expressed in C1 (i.e. S100A4). However, we</span></a>
<a class="sourceLine" id="cb136-5" title="5"><span class="co"># can see that CCR7 is upregulated in C0, strongly indicating that we can</span></a>
<a class="sourceLine" id="cb136-6" title="6"><span class="co"># differentiate memory from naive CD4 cells.  cols.use demarcates the color</span></a>
<a class="sourceLine" id="cb136-7" title="7"><span class="co"># palette from low to high expression</span></a>
<a class="sourceLine" id="cb136-8" title="8"><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;S100A4&quot;</span>, <span class="st">&quot;CCR7&quot;</span>), <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;blue&quot;</span>))</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/tcell_markers-1.png" width="672" /></p>
</div>
</div>
</div>
<div id="feature-selection" class="section level2">
<h2><span class="header-section-number">6.3</span> Feature Selection</h2>
<div id="differential-expression-analysis" class="section level3">
<h3><span class="header-section-number">6.3.1</span> Differential Expression Analysis</h3>
<div id="differential-expression-tests" class="section level4">
<h4><span class="header-section-number">6.3.1.1</span> Differential Expression Tests</h4>
<p>One of the most commonly performed tasks for RNA-seq data is differential gene expression (DE) analysis. Although well-established tools exist for such analysis in bulk RNA-seq data, methods for scRNA-seq data are just emerging. Given the special characteristics of scRNA-seq data, including generally low library sizes, high noise levels and a large fraction of so-called ‘dropout’ events, it is unclear whether DE methods that have been developed for bulk RNA-seq are suitable also for scRNA-seq. Check the help page out for the FindMarkers function by using <code> ?FindMarkers </code></p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb137-1" title="1"><span class="co">## Differential expression using DESeq2</span></a>
<a class="sourceLine" id="cb137-2" title="2"><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>, <span class="dt">test.use =</span> <span class="st">&quot;DESeq2&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb138-1" title="1"><span class="co">## Use the help function and run other tests.  Do they find similar markers?</span></a>
<a class="sourceLine" id="cb138-2" title="2"><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>, <span class="dt">test.use =</span> ??)</a></code></pre></div>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb139-1" title="1"><span class="co">## t-test</span></a>
<a class="sourceLine" id="cb139-2" title="2"><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>, <span class="dt">test.use =</span> <span class="st">&quot;t&quot;</span>)</a></code></pre></div>
</div>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3><span class="header-section-number">6.3.2</span> Dimensionality Reduction</h3>
<div id="principal-components-analysis-pca" class="section level4">
<h4><span class="header-section-number">6.3.2.1</span> Principal Components Analysis (PCA)</h4>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb140-1" title="1"><span class="kw">RunPCA</span>()</a></code></pre></div>
</div>
</div>
<div id="independent-components-analysis-ica" class="section level3">
<h3><span class="header-section-number">6.3.3</span> Independent Components Analysis (ICA)</h3>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb141-1" title="1"><span class="kw">RunICA</span>()</a></code></pre></div>
</div>
<div id="clustering" class="section level3">
<h3><span class="header-section-number">6.3.4</span> Clustering</h3>
<div id="kmeans" class="section level4">
<h4><span class="header-section-number">6.3.4.1</span> Kmeans</h4>
</div>
<div id="louvain" class="section level4">
<h4><span class="header-section-number">6.3.4.2</span> Louvain</h4>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb142-1" title="1"><span class="co">##  Neighborhood graph</span></a>
<a class="sourceLine" id="cb142-2" title="2"><span class="kw">FindNeighbors</span>(object, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb142-3" title="3">  <span class="dt">assay =</span> <span class="ot">NULL</span>, <span class="dt">features =</span> <span class="ot">NULL</span>, <span class="dt">k.param =</span> <span class="dv">20</span>, <span class="dt">compute.SNN =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb142-4" title="4">  <span class="dt">prune.SNN =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">15</span>, <span class="dt">nn.method =</span> <span class="st">&quot;rann&quot;</span>, <span class="dt">annoy.metric =</span> <span class="st">&quot;euclidean&quot;</span>,</a>
<a class="sourceLine" id="cb142-5" title="5">  <span class="dt">nn.eps =</span> <span class="dv">0</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>, <span class="dt">force.recalc =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb142-6" title="6">  <span class="dt">do.plot =</span> <span class="ot">FALSE</span>, <span class="dt">graph.name =</span> <span class="ot">NULL</span>, ...)</a>
<a class="sourceLine" id="cb142-7" title="7"></a>
<a class="sourceLine" id="cb142-8" title="8"><span class="kw">FindClusters</span>(object, <span class="dt">genes.use =</span> <span class="ot">NULL</span>, <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>,</a>
<a class="sourceLine" id="cb142-9" title="9"><span class="dt">dims.use =</span> <span class="ot">NULL</span>, <span class="dt">k.param =</span> <span class="dv">30</span>, <span class="dt">plot.SNN =</span> <span class="ot">FALSE</span>, <span class="dt">prune.SNN =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">15</span>,</a>
<a class="sourceLine" id="cb142-10" title="10"><span class="dt">print.output =</span> <span class="ot">TRUE</span>, <span class="dt">distance.matrix =</span> <span class="ot">NULL</span>, <span class="dt">save.SNN =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb142-11" title="11"><span class="dt">reuse.SNN =</span> <span class="ot">FALSE</span>, <span class="dt">force.recalc =</span> <span class="ot">FALSE</span>, <span class="dt">nn.eps =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb142-12" title="12"><span class="dt">modularity.fxn =</span> <span class="dv">1</span>, <span class="dt">resolution =</span> <span class="fl">0.8</span>, <span class="dt">algorithm =</span> <span class="dv">1</span>, <span class="dt">n.start =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb142-13" title="13"><span class="dt">n.iter =</span> <span class="dv">10</span>, <span class="dt">random.seed =</span> <span class="dv">0</span>, <span class="dt">temp.file.location =</span> <span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb142-14" title="14"><span class="dt">edge.file.name =</span> <span class="ot">NULL</span>)</a></code></pre></div>
</div>
</div>
<div id="check-clusters" class="section level3">
<h3><span class="header-section-number">6.3.5</span> Check Clusters</h3>
<p>Use Classifier to predict cell cluster. See how it predicts using hold out data.
<a href="https://satijalab.org/seurat/v3.0/integration.html">reference</a></p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb143-1" title="1"><span class="co"># Assign the test object a three level attribute</span></a>
<a class="sourceLine" id="cb143-2" title="2">groups &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">&quot;train&quot;</span>, <span class="st">&quot;test&quot;</span>), <span class="dt">size =</span> <span class="kw">NROW</span>(seurat<span class="op">@</span>meta.data), <span class="dt">replace =</span> <span class="ot">TRUE</span>, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="fl">0.8</span>, <span class="fl">0.2</span>))</a>
<a class="sourceLine" id="cb143-3" title="3"><span class="kw">names</span>(groups) &lt;-<span class="st"> </span><span class="kw">colnames</span>(seurat)</a>
<a class="sourceLine" id="cb143-4" title="4">seurat &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> seurat, <span class="dt">metadata =</span> groups, <span class="dt">col.name =</span> <span class="st">&quot;group&quot;</span>)</a>
<a class="sourceLine" id="cb143-5" title="5"></a>
<a class="sourceLine" id="cb143-6" title="6"><span class="co"># Find Anchors</span></a>
<a class="sourceLine" id="cb143-7" title="7">seurat.list &lt;-<span class="st"> </span><span class="kw">SplitObject</span>(seurat, <span class="dt">split.by =</span> <span class="st">&quot;group&quot;</span>)</a>
<a class="sourceLine" id="cb143-8" title="8">seurat.anchors &lt;-<span class="st"> </span><span class="kw">FindIntegrationAnchors</span>(<span class="dt">object.list =</span> seurat.list, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</a></code></pre></div>
<pre><code>## Computing 2000 integration features</code></pre>
<pre><code>## Scaling features for provided objects</code></pre>
<pre><code>## Finding all pairwise anchors</code></pre>
<pre><code>## Running CCA</code></pre>
<pre><code>## Merging objects</code></pre>
<pre><code>## Finding neighborhoods</code></pre>
<pre><code>## Finding anchors</code></pre>
<pre><code>##  Found 978 anchors</code></pre>
<pre><code>## Filtering anchors</code></pre>
<pre><code>##  Retained 763 anchors</code></pre>
<pre><code>## Extracting within-dataset neighbors</code></pre>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb155-1" title="1">seurat.integrated &lt;-<span class="st"> </span><span class="kw">IntegrateData</span>(<span class="dt">anchorset =</span> seurat.anchors, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</a></code></pre></div>
<pre><code>## Merging dataset 1 into 2</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb161-1" title="1">seurat.query &lt;-<span class="st"> </span>seurat.list[[<span class="st">&quot;train&quot;</span>]]</a>
<a class="sourceLine" id="cb161-2" title="2">seurat.anchors &lt;-<span class="st"> </span><span class="kw">FindTransferAnchors</span>(<span class="dt">reference =</span> seurat.integrated, </a>
<a class="sourceLine" id="cb161-3" title="3">                                      <span class="dt">query =</span> seurat.query, </a>
<a class="sourceLine" id="cb161-4" title="4">                                      <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</a></code></pre></div>
<pre><code>## Performing PCA on the provided reference using 2000 features as input.</code></pre>
<pre><code>## Projecting PCA</code></pre>
<pre><code>## Finding neighborhoods</code></pre>
<pre><code>## Finding anchors</code></pre>
<pre><code>##  Found 2423 anchors</code></pre>
<pre><code>## Filtering anchors</code></pre>
<pre><code>##  Retained 2197 anchors</code></pre>
<pre><code>## Extracting within-dataset neighbors</code></pre>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb170-1" title="1">predictions &lt;-<span class="st"> </span><span class="kw">TransferData</span>(<span class="dt">anchorset =</span> seurat.anchors, </a>
<a class="sourceLine" id="cb170-2" title="2">                            <span class="dt">refdata =</span> seurat.integrated<span class="op">$</span>ClusterNames_<span class="fl">0.6</span>, </a>
<a class="sourceLine" id="cb170-3" title="3">                            <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</a></code></pre></div>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Predicting cell labels</code></pre>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb174-1" title="1">seurat.query &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(seurat.query, <span class="dt">metadata =</span> predictions)</a>
<a class="sourceLine" id="cb174-2" title="2"><span class="kw">table</span>(seurat.query<span class="op">@</span>meta.data<span class="op">$</span>ClusterNames_<span class="fl">0.6</span>, seurat.query<span class="op">@</span>meta.data<span class="op">$</span>predicted.id)</a></code></pre></div>
<pre><code>##               
##                  B CD14+ Mono CD8 T  DC FCGR3A+ Mono Memory CD4 T
##   Memory CD4 T   0          0     0   0            0          182
##   CD14+ Mono     0        150     0   0            1            0
##   Naive CD4 T    0          0     1   1            0           14
##   B            106          0     0   0            0            0
##   CD8 T          0          0    70   1            0            0
##   FCGR3A+ Mono   0          1     0   0           51            0
##   NK             0          0     2   0            0            0
##   DC             0          0     1  30            0           17
##               
##                Naive CD4 T  NK
##   Memory CD4 T           0   0
##   CD14+ Mono             0   0
##   Naive CD4 T          113   0
##   B                      0   0
##   CD8 T                  2   0
##   FCGR3A+ Mono           0   0
##   NK                     0  47
##   DC                     0   0</code></pre>
<p>Visually check by comparing centroids of clusters in gene space and embedding space.</p>
</div>
<div id="view-entire-object-structure" class="section level3">
<h3><span class="header-section-number">6.3.6</span> View Entire Object Structure</h3>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb176-1" title="1"><span class="kw">str</span>(seurat)</a></code></pre></div>
<pre><code>## Formal class &#39;Seurat&#39; [package &quot;Seurat&quot;] with 12 slots
##   ..@ assays      :List of 1
##   .. ..$ RNA:Formal class &#39;Assay&#39; [package &quot;Seurat&quot;] with 7 slots
##   .. .. .. ..@ counts       :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. .. .. .. .. ..@ i       : int [1:838799] 20 59 64 119 132 150 152 182 184 185 ...
##   .. .. .. .. .. ..@ p       : int [1:988] 0 779 2127 3253 4211 4732 5512 6292 7081 7612 ...
##   .. .. .. .. .. ..@ Dim     : int [1:2] 11766 987
##   .. .. .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. .. .. ..$ : chr [1:11766] &quot;LINC00115&quot; &quot;NOC2L&quot; &quot;KLHL17&quot; &quot;PLEKHN1&quot; ...
##   .. .. .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..@ x       : num [1:838799] 1 1 2 1 1 1 1 41 1 1 ...
##   .. .. .. .. .. ..@ factors : list()
##   .. .. .. ..@ data         :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. .. .. .. .. ..@ i       : int [1:838799] 20 59 64 119 132 150 152 182 184 185 ...
##   .. .. .. .. .. ..@ p       : int [1:988] 0 779 2127 3253 4211 4732 5512 6292 7081 7612 ...
##   .. .. .. .. .. ..@ Dim     : int [1:2] 11766 987
##   .. .. .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. .. .. ..$ : chr [1:11766] &quot;LINC00115&quot; &quot;NOC2L&quot; &quot;KLHL17&quot; &quot;PLEKHN1&quot; ...
##   .. .. .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..@ x       : num [1:838799] 1.64 1.64 2.23 1.64 1.64 ...
##   .. .. .. .. .. ..@ factors : list()
##   .. .. .. ..@ scale.data   : num [1:2000, 1:987] -0.319 -0.317 -0.939 -0.154 -0.275 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2000] &quot;NOC2L&quot; &quot;HES4&quot; &quot;ISG15&quot; &quot;TNFRSF4&quot; ...
##   .. .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. ..@ key          : chr &quot;rna_&quot;
##   .. .. .. ..@ var.features : chr [1:2000] &quot;LYZ&quot; &quot;S100A9&quot; &quot;FTL&quot; &quot;GNLY&quot; ...
##   .. .. .. ..@ meta.features:&#39;data.frame&#39;:   11766 obs. of  5 variables:
##   .. .. .. .. ..$ vst.mean                 : num [1:11766] 0.00507 0.11854 0.00405 0.00405 0.09321 ...
##   .. .. .. .. ..$ vst.variance             : num [1:11766] 0.00505 0.24658 0.00404 0.00404 0.1698 ...
##   .. .. .. .. ..$ vst.variance.expected    : num [1:11766] 0.0053 0.2029 0.00417 0.00417 0.14985 ...
##   .. .. .. .. ..$ vst.variance.standardized: num [1:11766] 0.953 1.215 0.97 0.97 1.133 ...
##   .. .. .. .. ..$ vst.variable             : logi [1:11766] FALSE TRUE FALSE FALSE TRUE TRUE ...
##   .. .. .. ..@ misc         : NULL
##   ..@ meta.data   :&#39;data.frame&#39;: 987 obs. of  11 variables:
##   .. ..$ orig.ident      : Factor w/ 1 level &quot;10X_NSCLC&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ nCount_RNA      : num [1:987] 2419 4898 3144 2637 980 ...
##   .. ..$ nFeature_RNA    : int [1:987] 779 1348 1126 958 521 780 780 789 531 550 ...
##   .. ..$ percent.mt      : num [1:987] 3.018 3.797 0.891 1.744 1.224 ...
##   .. ..$ barcode         : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. ..$ batch_id        : int [1:987] 0 1 1 2 0 2 2 1 1 0 ...
##   .. ..$ RNA_snn_res.1   : Factor w/ 8 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 8 4 3 2 7 3 5 5 8 2 ...
##   .. ..$ seurat_clusters : Factor w/ 7 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 4 5 3 2 7 3 4 4 4 2 ...
##   .. ..$ ClusterNames_0.6: Factor w/ 8 levels &quot;Memory CD4 T&quot;,..: 8 4 3 2 7 3 5 5 8 2 ...
##   .. ..$ RNA_snn_res.0.8 : Factor w/ 7 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 4 5 3 2 7 3 4 4 4 2 ...
##   .. ..$ group           : chr [1:987] &quot;test&quot; &quot;test&quot; &quot;train&quot; &quot;train&quot; ...
##   ..@ active.assay: chr &quot;RNA&quot;
##   ..@ active.ident: Factor w/ 7 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 4 5 3 2 7 3 4 4 4 2 ...
##   .. ..- attr(*, &quot;names&quot;)= chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   ..@ graphs      :List of 2
##   .. ..$ RNA_nn :Formal class &#39;Graph&#39; [package &quot;Seurat&quot;] with 6 slots
##   .. .. .. ..@ i       : int [1:19740] 0 41 77 119 171 212 215 231 232 265 ...
##   .. .. .. ..@ p       : int [1:988] 0 35 45 53 61 68 80 124 128 138 ...
##   .. .. .. ..@ Dim     : int [1:2] 987 987
##   .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. ..@ x       : num [1:19740] 1 1 1 1 1 1 1 1 1 1 ...
##   .. .. .. ..@ factors : list()
##   .. ..$ RNA_snn:Formal class &#39;Graph&#39; [package &quot;Seurat&quot;] with 6 slots
##   .. .. .. ..@ i       : int [1:72095] 0 5 8 12 41 67 77 112 119 123 ...
##   .. .. .. ..@ p       : int [1:988] 0 81 123 174 255 309 403 475 557 615 ...
##   .. .. .. ..@ Dim     : int [1:2] 987 987
##   .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. ..@ x       : num [1:72095] 1 0.0811 0.0811 0.0811 0.3333 ...
##   .. .. .. ..@ factors : list()
##   ..@ neighbors   : list()
##   ..@ reductions  :List of 4
##   .. ..$ pca :Formal class &#39;DimReduc&#39; [package &quot;Seurat&quot;] with 8 slots
##   .. .. .. ..@ cell.embeddings           : num [1:987, 1:50] 5.102 0.331 2.578 -13.581 3.815 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..$ : chr [1:50] &quot;PC_1&quot; &quot;PC_2&quot; &quot;PC_3&quot; &quot;PC_4&quot; ...
##   .. .. .. ..@ feature.loadings          : num [1:2000, 1:50] -0.1091 -0.1094 -0.1125 0.0151 -0.1107 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2000] &quot;LYZ&quot; &quot;S100A9&quot; &quot;FTL&quot; &quot;GNLY&quot; ...
##   .. .. .. .. .. ..$ : chr [1:50] &quot;PC_1&quot; &quot;PC_2&quot; &quot;PC_3&quot; &quot;PC_4&quot; ...
##   .. .. .. ..@ feature.loadings.projected: num [1:2000, 1:50] 661 -3169 -2757 789 205 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2000] &quot;NOC2L&quot; &quot;HES4&quot; &quot;ISG15&quot; &quot;TNFRSF4&quot; ...
##   .. .. .. .. .. ..$ : chr [1:50] &quot;PC_1&quot; &quot;PC_2&quot; &quot;PC_3&quot; &quot;PC_4&quot; ...
##   .. .. .. ..@ assay.used                : chr &quot;RNA&quot;
##   .. .. .. ..@ stdev                     : num [1:50] 7.39 4.8 4.17 4.05 3.46 ...
##   .. .. .. ..@ key                       : chr &quot;PC_&quot;
##   .. .. .. ..@ jackstraw                 :Formal class &#39;JackStrawData&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. .. .. ..@ empirical.p.values     : num [1:2000, 1:20] 0 0 0 0.0005 0 0 0 0.0565 0 0 ...
##   .. .. .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. .. .. ..$ : chr [1:2000] &quot;LYZ&quot; &quot;S100A9&quot; &quot;FTL&quot; &quot;GNLY&quot; ...
##   .. .. .. .. .. .. .. ..$ : chr [1:20] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot; &quot;PC4&quot; ...
##   .. .. .. .. .. ..@ fake.reduction.scores  : num [1:2000, 1:20] 0.005743 0.006021 0.001467 -0.006214 -0.000962 ...
##   .. .. .. .. .. ..@ empirical.p.values.full: logi [1, 1] NA
##   .. .. .. .. .. ..@ overall.p.values       : num [1:20, 1:2] 1 2 3 4 5 6 7 8 9 10 ...
##   .. .. .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. .. .. ..$ : NULL
##   .. .. .. .. .. .. .. ..$ : chr [1:2] &quot;PC&quot; &quot;Score&quot;
##   .. .. .. ..@ misc                      :List of 1
##   .. .. .. .. ..$ total.variance: num 1761
##   .. ..$ ica :Formal class &#39;DimReduc&#39; [package &quot;Seurat&quot;] with 8 slots
##   .. .. .. ..@ cell.embeddings           : num [1:987, 1:5] 0.849 -0.887 1.308 -1.59 -0.706 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..$ : chr [1:5] &quot;IC_1&quot; &quot;IC_2&quot; &quot;IC_3&quot; &quot;IC_4&quot; ...
##   .. .. .. ..@ feature.loadings          : num [1:2000, 1:5] -653.8 -691.6 -688.2 -23.3 -611.5 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2000] &quot;LYZ&quot; &quot;S100A9&quot; &quot;FTL&quot; &quot;GNLY&quot; ...
##   .. .. .. .. .. ..$ : chr [1:5] &quot;IC_1&quot; &quot;IC_2&quot; &quot;IC_3&quot; &quot;IC_4&quot; ...
##   .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] 
##   .. .. .. ..@ assay.used                : chr &quot;RNA&quot;
##   .. .. .. ..@ stdev                     : num(0) 
##   .. .. .. ..@ key                       : chr &quot;IC_&quot;
##   .. .. .. ..@ jackstraw                 :Formal class &#39;JackStrawData&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] 
##   .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] 
##   .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] 
##   .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] 
##   .. .. .. ..@ misc                      : list()
##   .. ..$ tsne:Formal class &#39;DimReduc&#39; [package &quot;Seurat&quot;] with 8 slots
##   .. .. .. ..@ cell.embeddings           : num [1:987, 1:2] -8.12 8.47 -24.9 26.51 -3.99 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..$ : chr [1:2] &quot;tSNE_1&quot; &quot;tSNE_2&quot;
##   .. .. .. ..@ feature.loadings          : num[0 , 0 ] 
##   .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] 
##   .. .. .. ..@ assay.used                : chr &quot;RNA&quot;
##   .. .. .. ..@ stdev                     : num(0) 
##   .. .. .. ..@ key                       : chr &quot;tSNE_&quot;
##   .. .. .. ..@ jackstraw                 :Formal class &#39;JackStrawData&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] 
##   .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] 
##   .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] 
##   .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] 
##   .. .. .. ..@ misc                      : list()
##   .. ..$ umap:Formal class &#39;DimReduc&#39; [package &quot;Seurat&quot;] with 8 slots
##   .. .. .. ..@ cell.embeddings           : num [1:987, 1:2] -3.53 -4.396 -1.722 9.476 -0.423 ...
##   .. .. .. .. ..- attr(*, &quot;scaled:center&quot;)= num [1:2] -1.302 0.637
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..$ : chr [1:2] &quot;UMAP_1&quot; &quot;UMAP_2&quot;
##   .. .. .. ..@ feature.loadings          : num[0 , 0 ] 
##   .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] 
##   .. .. .. ..@ assay.used                : chr &quot;RNA&quot;
##   .. .. .. ..@ stdev                     : num(0) 
##   .. .. .. ..@ key                       : chr &quot;UMAP_&quot;
##   .. .. .. ..@ jackstraw                 :Formal class &#39;JackStrawData&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] 
##   .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] 
##   .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] 
##   .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] 
##   .. .. .. ..@ misc                      : list()
##   ..@ project.name: chr &quot;10X_NSCLC&quot;
##   ..@ misc        : list()
##   ..@ version     :Classes &#39;package_version&#39;, &#39;numeric_version&#39;  hidden list of 1
##   .. ..$ : int [1:3] 3 1 1
##   ..@ commands    :List of 12
##   .. ..$ NormalizeData.RNA       :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;NormalizeData.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:55:48&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;NormalizeData(object = seurat, normalization.method = \&quot;LogNormalize\&quot;, &quot; &quot;    scale.factor = 10000)&quot;
##   .. .. .. ..@ params     :List of 5
##   .. .. .. .. ..$ assay               : chr &quot;RNA&quot;
##   .. .. .. .. ..$ normalization.method: chr &quot;LogNormalize&quot;
##   .. .. .. .. ..$ scale.factor        : num 10000
##   .. .. .. .. ..$ margin              : num 1
##   .. .. .. .. ..$ verbose             : logi TRUE
##   .. ..$ FindVariableFeatures.RNA:Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;FindVariableFeatures.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:55:48&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;FindVariableFeatures(object = seurat, mean.function = ExpMean, &quot; &quot;    dispersion.function = LogVMR, do.plot = FALSE)&quot;
##   .. .. .. ..@ params     :List of 12
##   .. .. .. .. ..$ assay              : chr &quot;RNA&quot;
##   .. .. .. .. ..$ selection.method   : chr &quot;vst&quot;
##   .. .. .. .. ..$ loess.span         : num 0.3
##   .. .. .. .. ..$ clip.max           : chr &quot;auto&quot;
##   .. .. .. .. ..$ mean.function      :function (x)  
##   .. .. .. .. ..$ dispersion.function:function (x)  
##   .. .. .. .. ..$ num.bin            : num 20
##   .. .. .. .. ..$ binning.method     : chr &quot;equal_width&quot;
##   .. .. .. .. ..$ nfeatures          : num 2000
##   .. .. .. .. ..$ mean.cutoff        : num [1:2] 0.1 8
##   .. .. .. .. ..$ dispersion.cutoff  : num [1:2] 1 Inf
##   .. .. .. .. ..$ verbose            : logi TRUE
##   .. ..$ ScaleData.RNA           :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;ScaleData.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:55:58&quot;
##   .. .. .. ..@ call.string: chr &quot;ScaleData(object = seurat, vars.to.regress = \&quot;batch_id\&quot;)&quot;
##   .. .. .. ..@ params     :List of 11
##   .. .. .. .. ..$ features          : chr [1:2000] &quot;LYZ&quot; &quot;S100A9&quot; &quot;FTL&quot; &quot;GNLY&quot; ...
##   .. .. .. .. ..$ assay             : chr &quot;RNA&quot;
##   .. .. .. .. ..$ vars.to.regress   : chr &quot;batch_id&quot;
##   .. .. .. .. ..$ model.use         : chr &quot;linear&quot;
##   .. .. .. .. ..$ use.umi           : logi FALSE
##   .. .. .. .. ..$ do.scale          : logi TRUE
##   .. .. .. .. ..$ do.center         : logi TRUE
##   .. .. .. .. ..$ scale.max         : num 10
##   .. .. .. .. ..$ block.size        : num 1000
##   .. .. .. .. ..$ min.cells.to.block: num 987
##   .. .. .. .. ..$ verbose           : logi TRUE
##   .. ..$ RunPCA.RNA              :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;RunPCA.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:56:00&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;RunPCA(object = seurat, pc.genes = seurat@assays$RNA@var.genes, &quot; &quot;    do.print = TRUE, pcs.print = 1:5, genes.print = 5)&quot;
##   .. .. .. ..@ params     :List of 10
##   .. .. .. .. ..$ assay          : chr &quot;RNA&quot;
##   .. .. .. .. ..$ npcs           : num 50
##   .. .. .. .. ..$ rev.pca        : logi FALSE
##   .. .. .. .. ..$ weight.by.var  : logi TRUE
##   .. .. .. .. ..$ verbose        : logi TRUE
##   .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5
##   .. .. .. .. ..$ nfeatures.print: num 30
##   .. .. .. .. ..$ reduction.name : chr &quot;pca&quot;
##   .. .. .. .. ..$ reduction.key  : chr &quot;PC_&quot;
##   .. .. .. .. ..$ seed.use       : num 42
##   .. ..$ RunICA.RNA              :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;RunICA.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:56:15&quot;
##   .. .. .. ..@ call.string: chr &quot;RunICA(seurat, nics = 5)&quot;
##   .. .. .. ..@ params     :List of 10
##   .. .. .. .. ..$ assay          : chr &quot;RNA&quot;
##   .. .. .. .. ..$ nics           : num 5
##   .. .. .. .. ..$ rev.ica        : logi FALSE
##   .. .. .. .. ..$ ica.function   : chr &quot;icafast&quot;
##   .. .. .. .. ..$ verbose        : logi TRUE
##   .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5
##   .. .. .. .. ..$ nfeatures.print: num 30
##   .. .. .. .. ..$ reduction.name : chr &quot;ica&quot;
##   .. .. .. .. ..$ reduction.key  : chr &quot;IC_&quot;
##   .. .. .. .. ..$ seed.use       : num 42
##   .. ..$ ProjectDim.RNA.pca      :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;ProjectDim.RNA.pca&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:56:16&quot;
##   .. .. .. ..@ call.string: chr &quot;ProjectDim(object = seurat, reduction = \&quot;pca\&quot;)&quot;
##   .. .. .. ..@ params     :List of 7
##   .. .. .. .. ..$ reduction      : chr &quot;pca&quot;
##   .. .. .. .. ..$ assay          : chr &quot;RNA&quot;
##   .. .. .. .. ..$ dims.print     : int [1:5] 1 2 3 4 5
##   .. .. .. .. ..$ nfeatures.print: num 20
##   .. .. .. .. ..$ overwrite      : logi FALSE
##   .. .. .. .. ..$ do.center      : logi FALSE
##   .. .. .. .. ..$ verbose        : logi TRUE
##   .. ..$ JackStraw.RNA.pca       :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;JackStraw.RNA.pca&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:57:34&quot;
##   .. .. .. ..@ call.string: chr &quot;JackStraw(object = seurat, reduction = \&quot;pca\&quot;)&quot;
##   .. .. .. ..@ params     :List of 7
##   .. .. .. .. ..$ reduction    : chr &quot;pca&quot;
##   .. .. .. .. ..$ assay        : chr &quot;RNA&quot;
##   .. .. .. .. ..$ dims         : num 20
##   .. .. .. .. ..$ num.replicate: num 100
##   .. .. .. .. ..$ prop.freq    : num 0.01
##   .. .. .. .. ..$ verbose      : logi TRUE
##   .. .. .. .. ..$ maxit        : num 1000
##   .. ..$ ScoreJackStraw          :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;ScoreJackStraw&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:57:34&quot;
##   .. .. .. ..@ call.string: chr &quot;ScoreJackStraw(seurat, dims = 1:20)&quot;
##   .. .. .. ..@ params     :List of 4
##   .. .. .. .. ..$ reduction   : chr &quot;pca&quot;
##   .. .. .. .. ..$ dims        : int [1:20] 1 2 3 4 5 6 7 8 9 10 ...
##   .. .. .. .. ..$ score.thresh: num 1e-05
##   .. .. .. .. ..$ do.plot     : logi FALSE
##   .. ..$ FindNeighbors.RNA.pca   :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;FindNeighbors.RNA.pca&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:57:35&quot;
##   .. .. .. ..@ call.string: chr &quot;FindNeighbors(object = seurat, dims = 1:10)&quot;
##   .. .. .. ..@ params     :List of 13
##   .. .. .. .. ..$ reduction   : chr &quot;pca&quot;
##   .. .. .. .. ..$ dims        : int [1:10] 1 2 3 4 5 6 7 8 9 10
##   .. .. .. .. ..$ assay       : chr &quot;RNA&quot;
##   .. .. .. .. ..$ k.param     : num 20
##   .. .. .. .. ..$ compute.SNN : logi TRUE
##   .. .. .. .. ..$ prune.SNN   : num 0.0667
##   .. .. .. .. ..$ nn.method   : chr &quot;rann&quot;
##   .. .. .. .. ..$ annoy.metric: chr &quot;euclidean&quot;
##   .. .. .. .. ..$ nn.eps      : num 0
##   .. .. .. .. ..$ verbose     : logi TRUE
##   .. .. .. .. ..$ force.recalc: logi FALSE
##   .. .. .. .. ..$ do.plot     : logi FALSE
##   .. .. .. .. ..$ graph.name  : chr [1:2] &quot;RNA_nn&quot; &quot;RNA_snn&quot;
##   .. ..$ RunTSNE                 :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;RunTSNE&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:57:37&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;RunTSNE(seurat, reduction.use = \&quot;pca\&quot;, dims.use = 1:10, &quot; &quot;    perplexity = 10)&quot;
##   .. .. .. ..@ params     :List of 9
##   .. .. .. .. ..$ reduction     : chr &quot;pca&quot;
##   .. .. .. .. ..$ cells         : chr [1:987] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. ..$ dims          : int [1:5] 1 2 3 4 5
##   .. .. .. .. ..$ seed.use      : num 1
##   .. .. .. .. ..$ tsne.method   : chr &quot;Rtsne&quot;
##   .. .. .. .. ..$ add.iter      : num 0
##   .. .. .. .. ..$ dim.embed     : num 2
##   .. .. .. .. ..$ reduction.name: chr &quot;tsne&quot;
##   .. .. .. .. ..$ reduction.key : chr &quot;tSNE_&quot;
##   .. ..$ RunUMAP.RNA.pca         :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;RunUMAP.RNA.pca&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:57:42&quot;
##   .. .. .. ..@ call.string: chr &quot;RunUMAP(seurat, dims = 1:10, perplexity = 10)&quot;
##   .. .. .. ..@ params     :List of 20
##   .. .. .. .. ..$ dims                : int [1:10] 1 2 3 4 5 6 7 8 9 10
##   .. .. .. .. ..$ reduction           : chr &quot;pca&quot;
##   .. .. .. .. ..$ assay               : chr &quot;RNA&quot;
##   .. .. .. .. ..$ umap.method         : chr &quot;uwot&quot;
##   .. .. .. .. ..$ n.neighbors         : int 30
##   .. .. .. .. ..$ n.components        : int 2
##   .. .. .. .. ..$ metric              : chr &quot;cosine&quot;
##   .. .. .. .. ..$ learning.rate       : num 1
##   .. .. .. .. ..$ min.dist            : num 0.3
##   .. .. .. .. ..$ spread              : num 1
##   .. .. .. .. ..$ set.op.mix.ratio    : num 1
##   .. .. .. .. ..$ local.connectivity  : int 1
##   .. .. .. .. ..$ repulsion.strength  : num 1
##   .. .. .. .. ..$ negative.sample.rate: int 5
##   .. .. .. .. ..$ uwot.sgd            : logi FALSE
##   .. .. .. .. ..$ seed.use            : int 42
##   .. .. .. .. ..$ angular.rp.forest   : logi FALSE
##   .. .. .. .. ..$ verbose             : logi TRUE
##   .. .. .. .. ..$ reduction.name      : chr &quot;umap&quot;
##   .. .. .. .. ..$ reduction.key       : chr &quot;UMAP_&quot;
##   .. ..$ FindClusters            :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. ..@ name       : chr &quot;FindClusters&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-09 16:58:34&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;FindClusters(object = seurat, reduction = \&quot;pca\&quot;, dims = 1:10, &quot; &quot;    resolution = 0.8)&quot;
##   .. .. .. ..@ params     :List of 9
##   .. .. .. .. ..$ graph.name      : chr &quot;RNA_snn&quot;
##   .. .. .. .. ..$ modularity.fxn  : num 1
##   .. .. .. .. ..$ resolution      : num 0.8
##   .. .. .. .. ..$ algorithm       : num 1
##   .. .. .. .. ..$ n.start         : num 10
##   .. .. .. .. ..$ n.iter          : num 10
##   .. .. .. .. ..$ random.seed     : num 0
##   .. .. .. .. ..$ group.singletons: logi TRUE
##   .. .. .. .. ..$ verbose         : logi TRUE
##   ..@ tools       : list()</code></pre>
<div id="probabilistic-lda" class="section level4">
<h4><span class="header-section-number">6.3.6.1</span> Probabilistic (LDA)</h4>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb178-1" title="1"><span class="kw">data</span>(<span class="st">&quot;MouseDeng2014.FitGoM&quot;</span>)</a>
<a class="sourceLine" id="cb178-2" title="2"><span class="kw">names</span>(MouseDeng2014.FitGoM)</a></code></pre></div>
<pre><code>## [1] &quot;clust_3&quot; &quot;clust_6&quot;</code></pre>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb180-1" title="1">omega &lt;-<span class="st"> </span>MouseDeng2014.FitGoM<span class="op">$</span>clust_<span class="dv">6</span><span class="op">$</span>omega</a>
<a class="sourceLine" id="cb180-2" title="2"></a>
<a class="sourceLine" id="cb180-3" title="3"></a>
<a class="sourceLine" id="cb180-4" title="4">annotation &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb180-5" title="5">  <span class="dt">sample_id =</span> <span class="kw">paste0</span>(<span class="st">&quot;X&quot;</span>, <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">NROW</span>(omega))),</a>
<a class="sourceLine" id="cb180-6" title="6">  <span class="dt">tissue_label =</span> <span class="kw">factor</span>(<span class="kw">rownames</span>(omega),</a>
<a class="sourceLine" id="cb180-7" title="7">                        <span class="dt">levels =</span> <span class="kw">rev</span>(<span class="kw">c</span>(<span class="st">&quot;zy&quot;</span>, <span class="st">&quot;early2cell&quot;</span>, <span class="st">&quot;mid2cell&quot;</span>, <span class="st">&quot;late2cell&quot;</span>,</a>
<a class="sourceLine" id="cb180-8" title="8">                                        <span class="st">&quot;4cell&quot;</span>, <span class="st">&quot;8cell&quot;</span>, <span class="st">&quot;16cell&quot;</span>, <span class="st">&quot;earlyblast&quot;</span>, <span class="st">&quot;midblast&quot;</span>,</a>
<a class="sourceLine" id="cb180-9" title="9">                                       <span class="st">&quot;lateblast&quot;</span>)))</a>
<a class="sourceLine" id="cb180-10" title="10">  )</a>
<a class="sourceLine" id="cb180-11" title="11"></a>
<a class="sourceLine" id="cb180-12" title="12"><span class="kw">rownames</span>(omega) &lt;-<span class="st"> </span>annotation<span class="op">$</span>sample_id;</a>
<a class="sourceLine" id="cb180-13" title="13"></a>
<a class="sourceLine" id="cb180-14" title="14"><span class="kw">StructureGGplot</span>(<span class="dt">omega =</span> omega,</a>
<a class="sourceLine" id="cb180-15" title="15">                <span class="dt">annotation =</span> annotation,</a>
<a class="sourceLine" id="cb180-16" title="16">                <span class="dt">palette =</span> RColorBrewer<span class="op">::</span><span class="kw">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Accent&quot;</span>),</a>
<a class="sourceLine" id="cb180-17" title="17">                <span class="dt">yaxis_label =</span> <span class="st">&quot;Amplification batch&quot;</span>,</a>
<a class="sourceLine" id="cb180-18" title="18">                <span class="dt">order_sample =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb180-19" title="19">                <span class="dt">axis_tick =</span> <span class="kw">list</span>(<span class="dt">axis_ticks_length =</span> <span class="fl">.1</span>,</a>
<a class="sourceLine" id="cb180-20" title="20">                                 <span class="dt">axis_ticks_lwd_y =</span> <span class="fl">.1</span>,</a>
<a class="sourceLine" id="cb180-21" title="21">                                 <span class="dt">axis_ticks_lwd_x =</span> <span class="fl">.1</span>,</a>
<a class="sourceLine" id="cb180-22" title="22">                                 <span class="dt">axis_label_size =</span> <span class="dv">7</span>,</a>
<a class="sourceLine" id="cb180-23" title="23">                                 <span class="dt">axis_label_face =</span> <span class="st">&quot;bold&quot;</span>))</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/example_lda-1.png" width="672" /></p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb181-1" title="1"><span class="kw">set.seed</span>(<span class="dv">2019</span>)</a>
<a class="sourceLine" id="cb181-2" title="2"></a>
<a class="sourceLine" id="cb181-3" title="3"><span class="co">## Preprocessing Steps</span></a>
<a class="sourceLine" id="cb181-4" title="4">pbmc_small &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> pbmc_small, <span class="dt">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="dt">scale.factor =</span> <span class="dv">10000</span>)</a>
<a class="sourceLine" id="cb181-5" title="5">pbmc_small &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(<span class="dt">object =</span> pbmc_small)</a></code></pre></div>
<pre><code>## Warning in irlba(A = t(x = object), nv = npcs, ...): You&#39;re computing too
## large a percentage of total singular values, use a standard svd instead.</code></pre>
<pre><code>## Warning in irlba(A = t(x = object), nv = npcs, ...): did not converge--
## results might be invalid!; try increasing work or maxit</code></pre>
<pre><code>## PC_ 1 
## Positive:  SDPR, PF4, PPBP, TUBB1, CA2, TREML1, MYL9, PGRMC1, RUFY1, PARVB 
##     VDAC3, IGLL5, AKR1C3, CD1C, RP11-290F20.3, GNLY, S100A8, S100A9, HLA-DQA1, HLA-DPB1 
## Negative:  HLA-DPB1, HLA-DQA1, S100A9, S100A8, GNLY, RP11-290F20.3, CD1C, AKR1C3, IGLL5, VDAC3 
##     PARVB, RUFY1, PGRMC1, MYL9, TREML1, CA2, TUBB1, PPBP, PF4, SDPR 
## PC_ 2 
## Positive:  HLA-DPB1, HLA-DQA1, S100A8, S100A9, CD1C, RP11-290F20.3, PARVB, IGLL5, MYL9, SDPR 
##     PPBP, CA2, RUFY1, TREML1, PF4, TUBB1, PGRMC1, VDAC3, AKR1C3, GNLY 
## Negative:  GNLY, AKR1C3, VDAC3, PGRMC1, TUBB1, PF4, TREML1, RUFY1, CA2, PPBP 
##     SDPR, MYL9, IGLL5, PARVB, RP11-290F20.3, CD1C, S100A9, S100A8, HLA-DQA1, HLA-DPB1 
## PC_ 3 
## Positive:  S100A9, S100A8, RP11-290F20.3, AKR1C3, PARVB, GNLY, PPBP, PGRMC1, MYL9, TUBB1 
##     CA2, TREML1, SDPR, VDAC3, PF4, RUFY1, HLA-DPB1, IGLL5, CD1C, HLA-DQA1 
## Negative:  HLA-DQA1, CD1C, IGLL5, HLA-DPB1, RUFY1, PF4, VDAC3, SDPR, TREML1, CA2 
##     TUBB1, MYL9, PGRMC1, PPBP, GNLY, PARVB, AKR1C3, RP11-290F20.3, S100A8, S100A9 
## PC_ 4 
## Positive:  IGLL5, RP11-290F20.3, VDAC3, PPBP, TUBB1, TREML1, PF4, CA2, PARVB, MYL9 
##     SDPR, RUFY1, PGRMC1, S100A9, HLA-DQA1, HLA-DPB1, GNLY, S100A8, AKR1C3, CD1C 
## Negative:  CD1C, AKR1C3, S100A8, GNLY, HLA-DPB1, HLA-DQA1, S100A9, PGRMC1, RUFY1, SDPR 
##     MYL9, PARVB, CA2, PF4, TREML1, TUBB1, PPBP, VDAC3, RP11-290F20.3, IGLL5 
## PC_ 5 
## Positive:  MYL9, PARVB, IGLL5, TREML1, AKR1C3, PGRMC1, HLA-DPB1, S100A9, TUBB1, PF4 
##     SDPR, GNLY, PPBP, S100A8, CA2, HLA-DQA1, CD1C, RUFY1, RP11-290F20.3, VDAC3 
## Negative:  VDAC3, RP11-290F20.3, RUFY1, CD1C, HLA-DQA1, CA2, S100A8, PPBP, GNLY, SDPR 
##     PF4, TUBB1, S100A9, HLA-DPB1, PGRMC1, AKR1C3, TREML1, IGLL5, PARVB, MYL9</code></pre>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb185-1" title="1">pbmc_small &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> pbmc_small, </a>
<a class="sourceLine" id="cb185-2" title="2">                           <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, </a>
<a class="sourceLine" id="cb185-3" title="3">                           <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, </a>
<a class="sourceLine" id="cb185-4" title="4">                           <span class="dt">resolution =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb185-5" title="5">                           <span class="dt">print.output =</span> <span class="dv">0</span>)</a></code></pre></div>
<pre><code>## Warning: The following arguments are not used: reduction, dims.use,
## print.output</code></pre>
<pre><code>## Suggested parameter: dims instead of dims.use; verbose instead of print.output</code></pre>
<pre><code>## Warning: The following arguments are not used: reduction, dims.use,
## print.output</code></pre>
<pre><code>## Suggested parameter: dims instead of dims.use; verbose instead of print.output</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 80
## Number of edges: 3124
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.1433
## Number of communities: 3
## Elapsed time: 0 seconds</code></pre>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb191-1" title="1">pbmc_counts &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(pbmc_small<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>data)</a>
<a class="sourceLine" id="cb191-2" title="2">pbmc_meta &lt;-<span class="st"> </span>pbmc_small<span class="op">@</span>meta.data</a>
<a class="sourceLine" id="cb191-3" title="3">gene_names &lt;-<span class="st"> </span><span class="kw">rownames</span>(pbmc_counts)</a>
<a class="sourceLine" id="cb191-4" title="4"></a>
<a class="sourceLine" id="cb191-5" title="5"></a>
<a class="sourceLine" id="cb191-6" title="6">pbmc_FitGoM &lt;-<span class="st"> </span><span class="kw">FitGoM</span>(<span class="kw">t</span>(pbmc_counts), <span class="dt">K=</span><span class="dv">4</span>)</a></code></pre></div>
<pre><code>## options not specified: switching to default BIC, other choice is BF for Bayes factor</code></pre>
<pre><code>## Fitting a Grade of Membership model
##             (Taddy M., AISTATS 2012, JMLR 22,
##             http://proceedings.mlr.press/v22/taddy12/taddy12.pdf)</code></pre>
<pre><code>## 
## Estimating on a 80 document collection.
## Fit and Bayes Factor Estimation for K = 4
## log posterior increase: 1006, 69.7, 71, done.
## log BF( 4 ) = 5090.41</code></pre>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb195-1" title="1">omega &lt;-<span class="st"> </span><span class="kw">data.frame</span>(pbmc_FitGoM<span class="op">$</span>fit<span class="op">$</span>omega)</a>
<a class="sourceLine" id="cb195-2" title="2"></a>
<a class="sourceLine" id="cb195-3" title="3">annotation &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample_id =</span> <span class="kw">rownames</span>(omega),</a>
<a class="sourceLine" id="cb195-4" title="4">                         <span class="dt">tissue_label =</span> <span class="kw">paste0</span>(<span class="st">&quot;topic&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>))</a>
<a class="sourceLine" id="cb195-5" title="5"></a>
<a class="sourceLine" id="cb195-6" title="6"><span class="kw">colnames</span>(omega) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;topic&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)</a>
<a class="sourceLine" id="cb195-7" title="7"><span class="kw">rownames</span>(omega) &lt;-<span class="st"> </span>annotation<span class="op">$</span>sample_id;</a>
<a class="sourceLine" id="cb195-8" title="8"></a>
<a class="sourceLine" id="cb195-9" title="9"><span class="kw">StructureGGplot</span>(<span class="dt">omega =</span> omega,</a>
<a class="sourceLine" id="cb195-10" title="10">                <span class="dt">annotation =</span> annotation,</a>
<a class="sourceLine" id="cb195-11" title="11">                <span class="dt">palette =</span> RColorBrewer<span class="op">::</span><span class="kw">brewer.pal</span>(<span class="dv">4</span>, <span class="st">&quot;Dark2&quot;</span>),</a>
<a class="sourceLine" id="cb195-12" title="12">                <span class="dt">yaxis_label =</span> <span class="st">&quot;Cells&quot;</span>,</a>
<a class="sourceLine" id="cb195-13" title="13">                <span class="dt">order_sample =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb195-14" title="14">                <span class="dt">axis_tick =</span> <span class="kw">list</span>(<span class="dt">axis_ticks_length =</span> <span class="fl">.1</span>,</a>
<a class="sourceLine" id="cb195-15" title="15">                                 <span class="dt">axis_ticks_lwd_y =</span> <span class="fl">.1</span>,</a>
<a class="sourceLine" id="cb195-16" title="16">                                 <span class="dt">axis_ticks_lwd_x =</span> <span class="fl">.1</span>,</a>
<a class="sourceLine" id="cb195-17" title="17">                                 <span class="dt">axis_label_size =</span> <span class="dv">7</span>,</a>
<a class="sourceLine" id="cb195-18" title="18">                                 <span class="dt">axis_label_face =</span> <span class="st">&quot;bold&quot;</span>))</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/pbmc_lda-1.png" width="672" /></p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb196-1" title="1"><span class="co"># ## Add Topic Scores to Meta Data Part of the Seurat Object</span></a>
<a class="sourceLine" id="cb196-2" title="2">pbmc_small &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(pbmc_small, omega)</a></code></pre></div>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb197-1" title="1">pbmc_small<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb197-2" title="2"><span class="st">  </span><span class="kw">group_by</span>(RNA_snn_res<span class="fl">.1</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb197-3" title="3"><span class="st">  </span><span class="kw">summarise</span>(<span class="dt">topic1 =</span> <span class="kw">mean</span>(topic1),</a>
<a class="sourceLine" id="cb197-4" title="4">            <span class="dt">topic2 =</span> <span class="kw">mean</span>(topic2),</a>
<a class="sourceLine" id="cb197-5" title="5">            <span class="dt">topic3 =</span> <span class="kw">mean</span>(topic3),</a>
<a class="sourceLine" id="cb197-6" title="6">            <span class="dt">topic4 =</span> <span class="kw">mean</span>(topic4))</a></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   RNA_snn_res.1 topic1 topic2 topic3 topic4
##   &lt;fct&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
## 1 0              0.281 0.290  0.221  0.207 
## 2 1              0.913 0.0253 0.0170 0.0450
## 3 2              0.595 0.0663 0.194  0.145</code></pre>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb199-1" title="1"><span class="co">## ggplot object, you can add layers</span></a>
<a class="sourceLine" id="cb199-2" title="2">p1 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(pbmc_small, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Resolution 1&quot;</span>) <span class="co">## return ggplot object</span></a>
<a class="sourceLine" id="cb199-3" title="3">p1</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/tsne_plot_pbmc-1.png" width="672" /></p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb200-1" title="1">p2 &lt;-<span class="st"> </span><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc_small, </a>
<a class="sourceLine" id="cb200-2" title="2">                  <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;topic1&quot;</span>, <span class="st">&quot;topic2&quot;</span>, <span class="st">&quot;topic3&quot;</span>, <span class="st">&quot;topic4&quot;</span>), </a>
<a class="sourceLine" id="cb200-3" title="3">                  <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;blue&quot;</span>), </a>
<a class="sourceLine" id="cb200-4" title="4">                  <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>) <span class="co">## return ggplot object</span></a>
<a class="sourceLine" id="cb200-5" title="5"></a>
<a class="sourceLine" id="cb200-6" title="6">p2</a></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/topic_plot-1.png" width="672" /></p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb201-1" title="1"><span class="kw">CombinePlots</span>(<span class="kw">list</span>(p1, p2))</a></code></pre></div>
<pre><code>## Warning: Graphs cannot be vertically aligned unless the axis parameter is
## set. Placing graphs unaligned.</code></pre>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/combine_topic_plot-1.png" width="672" /></p>
</div>
<div id="extract-top-feature" class="section level4">
<h4><span class="header-section-number">6.3.6.2</span> Extract Top Feature</h4>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb203-1" title="1">theta_mat &lt;-<span class="st"> </span>pbmc_FitGoM<span class="op">$</span>fit<span class="op">$</span>theta</a>
<a class="sourceLine" id="cb203-2" title="2"></a>
<a class="sourceLine" id="cb203-3" title="3">top_features &lt;-<span class="st"> </span><span class="kw">ExtractTopFeatures</span>(theta_mat, </a>
<a class="sourceLine" id="cb203-4" title="4">                                   <span class="dt">top_features=</span><span class="dv">100</span>,</a>
<a class="sourceLine" id="cb203-5" title="5">                                   <span class="dt">method=</span><span class="st">&quot;poisson&quot;</span>, </a>
<a class="sourceLine" id="cb203-6" title="6">                                   <span class="dt">options=</span><span class="st">&quot;min&quot;</span>)</a>
<a class="sourceLine" id="cb203-7" title="7"></a>
<a class="sourceLine" id="cb203-8" title="8">gene_list &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, </a>
<a class="sourceLine" id="cb203-9" title="9">                     <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">dim</span>(top_features<span class="op">$</span>indices)[<span class="dv">1</span>],</a>
<a class="sourceLine" id="cb203-10" title="10">                            <span class="cf">function</span>(x) gene_names[top_features<span class="op">$</span>indices[x,]]))</a></code></pre></div>
<p>We tabulate the top <span class="math inline">\(5\)</span> genes for these <span class="math inline">\(6\)</span> clusters.</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb204-1" title="1">out_table &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="cf">function</span>(i) <span class="kw">toString</span>(gene_list[,i])))</a>
<a class="sourceLine" id="cb204-2" title="2"><span class="kw">rownames</span>(out_table) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Cluster&quot;</span>, <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>))</a>
<a class="sourceLine" id="cb204-3" title="3">out_table</a></code></pre></div>
<pre><code>##           [,1]                             
## Cluster 1 &quot;GRN, GNLY, GZMK, PIK3IP1&quot;       
## Cluster 2 &quot;HLA-DMB, GZMB, HNRNPH1, TNFAIP8&quot;
## Cluster 3 &quot;IFI30, PRF1, SATB1, THYN1&quot;      
## Cluster 4 &quot;LY86, GZMH, SIT1, SP100&quot;        
## Cluster 5 &quot;PF4, CCL4, CRBN, CCR7&quot;</code></pre>
</div>
</div>
<div id="practice-visualizingembedding" class="section level3">
<h3><span class="header-section-number">6.3.7</span> Practice Visualizing/Embedding</h3>
<div id="tsne" class="section level4">
<h4><span class="header-section-number">6.3.7.1</span> tSNE</h4>
<p>Change the parameter settings for tSNE</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb206-1" title="1"><span class="kw">RunTSNE</span>()</a></code></pre></div>
</div>
<div id="umap" class="section level4">
<h4><span class="header-section-number">6.3.7.2</span> UMAP</h4>
<p>Change the parameter settings for UMAP</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb207-1" title="1"><span class="kw">RunUMAP</span>()</a></code></pre></div>
</div>
</div>
</div>
<div id="other-options-for-analysis" class="section level2">
<h2><span class="header-section-number">6.4</span> Other Options For Analysis</h2>
<ul>
<li><a href="https://satijalab.org/seurat/vignettes.html">More Seurat Vignettes</a></li>
<li><a href="https://broadinstitute.github.io/2019_scWorkshop/">Single Cell Analysis Workshop</a></li>
<li><a href="https://scrnaseq-course.cog.sanger.ac.uk/website/index.html">Hemberg Lab Course</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment</a></li>
<li><a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">Scanpy</a></li>
<li><a href="https://github.com/klarman-cell-observatory/pegasus/blob/master/notebooks/Pegasus%20Tutorial%20in%20Workshop.ipynb">Pegasus</a></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="identifying-cell-populations-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["KrumlovSingleCellWorkshop2020.pdf", "KrumlovSingleCellWorkshop2020.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
