<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="7 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta name="generator" content="bookdown 0.16 and GitBook 2.6.7" />

  <meta property="og:title" content="7 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/" />
  <meta property="og:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />
  
  <meta name="github-repo" content="broadinstitute/KrumlovSingleCellWorkship2020" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  
  <meta name="twitter:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />

<meta name="author" content="Kirk Gosik" />


<meta name="date" content="2020-01-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="identifying-cell-populations-1.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#introduction-1"><i class="fa fa-check"></i><b>1.5.1</b> Introduction</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#transcriptome-quantification"><i class="fa fa-check"></i><b>1.5.2</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#expression-qc-and-normalization"><i class="fa fa-check"></i><b>1.5.3</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#data-wrangling-scrnaseq"><i class="fa fa-check"></i><b>1.5.4</b> Data Wrangling scRNAseq</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#identifying-cell-populations"><i class="fa fa-check"></i><b>1.5.5</b> Identifying Cell Populations</a></li>
<li class="chapter" data-level="1.5.6" data-path="index.html"><a href="index.html#feature-selection-and-cluster-analysis"><i class="fa fa-check"></i><b>1.5.6</b> Feature Selection and Cluster Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html"><i class="fa fa-check"></i><b>2</b> Transcriptome Quantification</a><ul>
<li class="chapter" data-level="2.1" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html#google-slides"><i class="fa fa-check"></i><b>2.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html"><i class="fa fa-check"></i><b>3</b> Expression QC and Normalization</a><ul>
<li class="chapter" data-level="3.1" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html#google-slides-1"><i class="fa fa-check"></i><b>3.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html"><i class="fa fa-check"></i><b>4</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#goal"><i class="fa fa-check"></i><b>4.1</b> Goal</a></li>
<li class="chapter" data-level="4.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#introduction-2"><i class="fa fa-check"></i><b>4.2</b> Introduction</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#load-necessary-packages"><i class="fa fa-check"></i><b>4.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>4.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>4.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>4.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>4.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>4.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>4.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>4.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#beginning-with-seurat"><i class="fa fa-check"></i><b>4.4</b> Beginning with Seurat:</a><ul>
<li class="chapter" data-level="4.4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>4.4.1</b> Creating a seurat object</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html"><i class="fa fa-check"></i><b>5</b> The following examples use the PBMC 3k dataset</a><ul>
<li class="chapter" data-level="5.1" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>5.1</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="5.2" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>5.2</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="5.2.1" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>5.2.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>5.3</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="5.4" data-path="the-following-examples-use-the-pbmc-3k-dataset.html"><a href="the-following-examples-use-the-pbmc-3k-dataset.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>5.4</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html"><i class="fa fa-check"></i><b>6</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="6.1" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html#google-slides-2"><i class="fa fa-check"></i><b>6.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html"><i class="fa fa-check"></i><b>7</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#abstract"><i class="fa fa-check"></i><b>7.1</b> Abstract</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#seurat-tutorial-redo"><i class="fa fa-check"></i><b>7.2</b> Seurat Tutorial Redo</a><ul>
<li class="chapter" data-level="7.2.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#preprocessing-steps"><i class="fa fa-check"></i><b>7.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="7.2.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>7.2.2</b> Start of Identifying Cell Types</a></li>
<li class="chapter" data-level="7.2.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#run-non-linear-dimensional-reduction-umaptsne"><i class="fa fa-check"></i><b>7.2.3</b> Run non-linear dimensional reduction (UMAP/tSNE)</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#feature-selection"><i class="fa fa-check"></i><b>7.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="7.3.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#differential-expression-analysis"><i class="fa fa-check"></i><b>7.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="7.3.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#check-clusters"><i class="fa fa-check"></i><b>7.3.2</b> Check Clusters</a></li>
<li class="chapter" data-level="7.3.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#view-entire-object-structure"><i class="fa fa-check"></i><b>7.3.3</b> View Entire Object Structure</a></li>
<li class="chapter" data-level="7.3.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>7.3.4</b> Practice Visualizing/Embedding</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#other-options-for-analysis"><i class="fa fa-check"></i><b>7.4</b> Other Options For Analysis</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="feature-selection-and-cluster-analysis-1" class="section level1">
<h1><span class="header-section-number">7</span> Feature Selection and Cluster Analysis</h1>
<div id="abstract" class="section level2">
<h2><span class="header-section-number">7.1</span> Abstract</h2>
<p>Many methods have been used to determine differential gene expression from single-cell RNA (scRNA)-seq data. We evaluated 36 approaches using experimental and synthetic data and found considerable differences in the number and characteristics of the genes that are called differentially expressed. Prefiltering of lowly expressed genes has important effects, particularly for some of the methods developed for bulk RNA-seq data analysis. However, we found that bulk RNA-seq analysis methods do not generally perform worse than those developed specifically for scRNA-seq. We also present conquer, a repository of consistently processed, analysis-ready public scRNA-seq data sets that is aimed at simplifying method evaluation and reanalysis of published results. Each data set provides abundance estimates for both genes and transcripts, as well as quality control and exploratory analysis reports. <span class="citation">(<span class="citeproc-not-found" data-reference-id="soneson2018bias"><strong>???</strong></span>)</span></p>
<p>Cells are the basic building blocks of organisms and each cell is unique. Single-cell RNA sequencing has emerged as an indispensable tool to dissect the cellular heterogeneity and decompose tissues into cell types and/or cell states, which offers enormous potential for de novo discovery. Single-cell transcriptomic atlases provide unprecedented resolution to reveal complex cellular events and deepen our understanding of biological systems. In this review, we summarize and compare single-cell RNA sequencing technologies, that were developed since 2009, to facilitate a well-informed choice of method. The applications of these methods in different biological contexts are also discussed. We anticipate an ever-increasing role of single-cell RNA sequencing in biology with further improvement in providing spatial information and coupling to other cellular modalities. In the future, such biological findings will greatly benefit medical research. <span class="citation">(<span class="citeproc-not-found" data-reference-id="hedlund2018single"><strong>???</strong></span>)</span></p>
</div>
<div id="seurat-tutorial-redo" class="section level2">
<h2><span class="header-section-number">7.2</span> Seurat Tutorial Redo</h2>
<p>For this tutorial, we will be analyzing the a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500. The raw data can be found <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">here</a>.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb90-1"><a href="feature-selection-and-cluster-analysis-1.html#cb90-1"></a><span class="fu">mkdir</span> /home/genomics/workshop_materials/scrna/pbmc3k</span>
<span id="cb90-2"><a href="feature-selection-and-cluster-analysis-1.html#cb90-2"></a><span class="fu">wget</span> https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz -O /home/genomics/workshop_materials/scrna/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz <span class="kw">&amp;&amp;</span> <span class="kw">\</span></span>
<span id="cb90-3"><a href="feature-selection-and-cluster-analysis-1.html#cb90-3"></a><span class="bu">cd</span> /home/genomics/workshop_materials/scrna/pbmc3k/ <span class="kw">&amp;&amp;</span> <span class="kw">\</span></span>
<span id="cb90-4"><a href="feature-selection-and-cluster-analysis-1.html#cb90-4"></a><span class="fu">tar</span> -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz</span></code></pre></div>
<p><strong>Task: Check the dirname to directory where you saved your data</strong></p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="feature-selection-and-cluster-analysis-1.html#cb91-1"></a>knitr<span class="op">::</span>opts_knit<span class="op">$</span><span class="kw">set</span>(<span class="dt">progress=</span><span class="ot">FALSE</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</span>
<span id="cb91-2"><a href="feature-selection-and-cluster-analysis-1.html#cb91-2"></a></span>
<span id="cb91-3"><a href="feature-selection-and-cluster-analysis-1.html#cb91-3"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb91-4"><a href="feature-selection-and-cluster-analysis-1.html#cb91-4"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb91-5"><a href="feature-selection-and-cluster-analysis-1.html#cb91-5"></a><span class="kw">library</span>(ggplot2)</span></code></pre></div>
<pre><code>## Warning: package &#39;ggplot2&#39; was built under R version 3.5.2</code></pre>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="feature-selection-and-cluster-analysis-1.html#cb93-1"></a><span class="kw">library</span>(CountClust)</span></code></pre></div>
<pre><code>## Warning: package &#39;CountClust&#39; was built under R version 3.5.2</code></pre>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="feature-selection-and-cluster-analysis-1.html#cb95-1"></a><span class="co">## dirname &lt;- &quot;/home/genomics/workshop_materials/scrna/&quot;</span></span>
<span id="cb95-2"><a href="feature-selection-and-cluster-analysis-1.html#cb95-2"></a>dirname &lt;-<span class="st"> &quot;/Users/kgosik/Documents/data/cellranger/&quot;</span></span>
<span id="cb95-3"><a href="feature-selection-and-cluster-analysis-1.html#cb95-3"></a>counts_matrix_filename =<span class="st"> </span><span class="kw">paste0</span>(dirname, <span class="st">&quot;pbmc3k/filtered_gene_bc_matrices/hg19/&quot;</span>)</span>
<span id="cb95-4"><a href="feature-selection-and-cluster-analysis-1.html#cb95-4"></a>counts &lt;-<span class="st"> </span><span class="kw">Read10X</span>(<span class="dt">data.dir =</span> counts_matrix_filename)  <span class="co"># Seurat function to read in 10x count data</span></span></code></pre></div>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="feature-selection-and-cluster-analysis-1.html#cb96-1"></a><span class="co">### seurat&lt;-CreateSeuratObject(raw.data = counts, ? = 3, ? = 350, project = &quot;10X_NSCLC&quot;)</span></span>
<span id="cb96-2"><a href="feature-selection-and-cluster-analysis-1.html#cb96-2"></a>seurat &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(<span class="dt">counts =</span> counts, <span class="dt">min.cells =</span> <span class="dv">3</span>, <span class="dt">min.features =</span> <span class="dv">350</span>, <span class="dt">project =</span> <span class="st">&quot;10X_PBMC&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)</code></pre>
<div id="preprocessing-steps" class="section level3">
<h3><span class="header-section-number">7.2.1</span> Preprocessing Steps</h3>
<p>This was all covered in the last Lab!</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="feature-selection-and-cluster-analysis-1.html#cb98-1"></a><span class="co"># The number of genes and UMIs (nFeature_RNA nCount_RNA) are automatically calculated</span></span>
<span id="cb98-2"><a href="feature-selection-and-cluster-analysis-1.html#cb98-2"></a><span class="co"># for every object by Seurat.  For non-UMI data, nCount_RNA represents the sum of</span></span>
<span id="cb98-3"><a href="feature-selection-and-cluster-analysis-1.html#cb98-3"></a><span class="co"># the non-normalized values within a cell We calculate the percentage of</span></span>
<span id="cb98-4"><a href="feature-selection-and-cluster-analysis-1.html#cb98-4"></a><span class="co"># mitochondrial genes here and store it in percent.mito using AddMetaData.</span></span>
<span id="cb98-5"><a href="feature-selection-and-cluster-analysis-1.html#cb98-5"></a><span class="co"># We use object@raw.data since this represents non-transformed and</span></span>
<span id="cb98-6"><a href="feature-selection-and-cluster-analysis-1.html#cb98-6"></a><span class="co"># non-log-normalized counts The % of UMI mapping to MT-genes is a common</span></span>
<span id="cb98-7"><a href="feature-selection-and-cluster-analysis-1.html#cb98-7"></a><span class="co"># scRNA-seq QC metric.</span></span>
<span id="cb98-8"><a href="feature-selection-and-cluster-analysis-1.html#cb98-8"></a>seurat[[<span class="st">&quot;percent.mt&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">PercentageFeatureSet</span>(<span class="dt">object =</span> seurat, <span class="dt">pattern =</span> <span class="st">&quot;^MT-&quot;</span>)</span>
<span id="cb98-9"><a href="feature-selection-and-cluster-analysis-1.html#cb98-9"></a><span class="co">##VlnPlot(object = seurat, features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;), ncol = 3)</span></span></code></pre></div>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="feature-selection-and-cluster-analysis-1.html#cb99-1"></a><span class="co"># FeatureScatter is typically used to visualize gene-gene relationships, but can</span></span>
<span id="cb99-2"><a href="feature-selection-and-cluster-analysis-1.html#cb99-2"></a><span class="co"># be used for anything calculated by the object, i.e. columns in</span></span>
<span id="cb99-3"><a href="feature-selection-and-cluster-analysis-1.html#cb99-3"></a><span class="co"># object@meta.data, PC scores etc.  Since there is a rare subset of cells</span></span>
<span id="cb99-4"><a href="feature-selection-and-cluster-analysis-1.html#cb99-4"></a><span class="co"># with an outlier level of high mitochondrial percentage and also low UMI</span></span>
<span id="cb99-5"><a href="feature-selection-and-cluster-analysis-1.html#cb99-5"></a><span class="co"># content, we filter these as well</span></span>
<span id="cb99-6"><a href="feature-selection-and-cluster-analysis-1.html#cb99-6"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb99-7"><a href="feature-selection-and-cluster-analysis-1.html#cb99-7"></a><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> seurat, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;percent.mt&quot;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/scatter_plot-1.png" width="672" /></p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="feature-selection-and-cluster-analysis-1.html#cb100-1"></a><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> seurat, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;nFeature_RNA&quot;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/scatter_plot-2.png" width="672" /></p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="feature-selection-and-cluster-analysis-1.html#cb101-1"></a><span class="co"># We filter out cells that have unique gene counts over 2,500 or less than</span></span>
<span id="cb101-2"><a href="feature-selection-and-cluster-analysis-1.html#cb101-2"></a><span class="co"># 200 Note that low.thresholds and high.thresholds are used to define a</span></span>
<span id="cb101-3"><a href="feature-selection-and-cluster-analysis-1.html#cb101-3"></a><span class="co"># &#39;gate&#39;.  -Inf and Inf should be used if you don&#39;t want a lower or upper</span></span>
<span id="cb101-4"><a href="feature-selection-and-cluster-analysis-1.html#cb101-4"></a><span class="co"># threshold.</span></span>
<span id="cb101-5"><a href="feature-selection-and-cluster-analysis-1.html#cb101-5"></a></span>
<span id="cb101-6"><a href="feature-selection-and-cluster-analysis-1.html#cb101-6"></a>seurat &lt;-<span class="st"> </span><span class="kw">SubsetData</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb101-7"><a href="feature-selection-and-cluster-analysis-1.html#cb101-7"></a>                    <span class="dt">subset.names =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;percent.mt&quot;</span>), </span>
<span id="cb101-8"><a href="feature-selection-and-cluster-analysis-1.html#cb101-8"></a>                    <span class="dt">low.thresholds =</span> <span class="kw">c</span>(<span class="dv">200</span>, <span class="op">-</span><span class="ot">Inf</span>), </span>
<span id="cb101-9"><a href="feature-selection-and-cluster-analysis-1.html#cb101-9"></a>                    <span class="dt">high.thresholds =</span> <span class="kw">c</span>(<span class="dv">2500</span>, <span class="fl">0.1</span>))</span></code></pre></div>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="feature-selection-and-cluster-analysis-1.html#cb102-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> seurat, <span class="dt">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="dt">scale.factor =</span> <span class="dv">10000</span>)</span></code></pre></div>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="feature-selection-and-cluster-analysis-1.html#cb103-1"></a><span class="co"># Read in a list of cell cycle markers, from Tirosh et al, 2015.</span></span>
<span id="cb103-2"><a href="feature-selection-and-cluster-analysis-1.html#cb103-2"></a><span class="co"># We can segregate this list into markers of G2/M phase and markers of S phase.</span></span>
<span id="cb103-3"><a href="feature-selection-and-cluster-analysis-1.html#cb103-3"></a>s.genes &lt;-<span class="st"> </span>Seurat<span class="op">::</span>cc.genes<span class="op">$</span>s.genes</span>
<span id="cb103-4"><a href="feature-selection-and-cluster-analysis-1.html#cb103-4"></a>s.genes &lt;-<span class="st"> </span>s.genes[s.genes <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(seurat)] <span class="co"># genes in dataset</span></span>
<span id="cb103-5"><a href="feature-selection-and-cluster-analysis-1.html#cb103-5"></a>g2m.genes &lt;-<span class="st"> </span>Seurat<span class="op">::</span>cc.genes<span class="op">$</span>g2m.genes</span>
<span id="cb103-6"><a href="feature-selection-and-cluster-analysis-1.html#cb103-6"></a>g2m.genes &lt;-<span class="st"> </span>g2m.genes[g2m.genes <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(seurat)] <span class="co"># genes in dataset</span></span>
<span id="cb103-7"><a href="feature-selection-and-cluster-analysis-1.html#cb103-7"></a></span>
<span id="cb103-8"><a href="feature-selection-and-cluster-analysis-1.html#cb103-8"></a>seurat &lt;-<span class="st"> </span><span class="kw">CellCycleScoring</span>(<span class="dt">object =</span> seurat, <span class="dt">s.features =</span> s.genes, <span class="dt">g2m.features =</span> g2m.genes, <span class="dt">set.ident =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="feature-selection-and-cluster-analysis-1.html#cb104-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb104-2"><a href="feature-selection-and-cluster-analysis-1.html#cb104-2"></a>                               <span class="dt">mean.function =</span> ExpMean, </span>
<span id="cb104-3"><a href="feature-selection-and-cluster-analysis-1.html#cb104-3"></a>                               <span class="dt">dispersion.function =</span> LogVMR)</span></code></pre></div>
</div>
<div id="start-of-identifying-cell-types" class="section level3">
<h3><span class="header-section-number">7.2.2</span> Start of Identifying Cell Types</h3>
<div id="scaling" class="section level4">
<h4><span class="header-section-number">7.2.2.1</span> Scaling</h4>
<p>This part is where you mean center the data, substract the mean. You also divide by the standard deviation to make everything to a ‘standard normal’, where the mean is zero and the standard deviation is 1.</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="feature-selection-and-cluster-analysis-1.html#cb105-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(<span class="dt">object =</span> seurat, <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">&quot;batch&quot;</span>, <span class="st">&quot;percent.mt&quot;</span>))</span></code></pre></div>
<pre><code>## Regressing out batch, percent.mt</code></pre>
<pre><code>## Centering and scaling data matrix</code></pre>
<p><strong>Task: Try Regressing Other Variables</strong></p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="feature-selection-and-cluster-analysis-1.html#cb108-1"></a><span class="co">## randomly making a batch id data.frame</span></span>
<span id="cb108-2"><a href="feature-selection-and-cluster-analysis-1.html#cb108-2"></a>batch_ids &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">barcode =</span> <span class="kw">rownames</span>(seurat<span class="op">@</span>meta.data), </span>
<span id="cb108-3"><a href="feature-selection-and-cluster-analysis-1.html#cb108-3"></a>                        <span class="dt">batch_id =</span> <span class="kw">sample</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">2</span>, <span class="kw">NROW</span>(seurat<span class="op">@</span>meta.data), <span class="dt">replace =</span> <span class="ot">TRUE</span>),</span>
<span id="cb108-4"><a href="feature-selection-and-cluster-analysis-1.html#cb108-4"></a>                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb108-5"><a href="feature-selection-and-cluster-analysis-1.html#cb108-5"></a></span>
<span id="cb108-6"><a href="feature-selection-and-cluster-analysis-1.html#cb108-6"></a></span>
<span id="cb108-7"><a href="feature-selection-and-cluster-analysis-1.html#cb108-7"></a><span class="kw">row.names</span>(batch_ids) &lt;-<span class="st"> </span><span class="kw">row.names</span>(seurat<span class="op">@</span>meta.data)</span>
<span id="cb108-8"><a href="feature-selection-and-cluster-analysis-1.html#cb108-8"></a>seurat &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> seurat, <span class="dt">metadata =</span> batch_ids, <span class="dt">col.name =</span> <span class="ot">NULL</span>)</span>
<span id="cb108-9"><a href="feature-selection-and-cluster-analysis-1.html#cb108-9"></a>seurat &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(<span class="dt">object =</span> seurat, <span class="dt">vars.to.regress =</span> <span class="st">&#39;batch_id&#39;</span>)</span></code></pre></div>
</div>
<div id="perform-linear-dimensional-reduction-pca" class="section level4">
<h4><span class="header-section-number">7.2.2.2</span> Perform linear dimensional reduction (PCA)</h4>
<p>This will run pca on the</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="feature-selection-and-cluster-analysis-1.html#cb109-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb109-2"><a href="feature-selection-and-cluster-analysis-1.html#cb109-2"></a>               <span class="co"># features = seurat@assays$RNA@var.features, </span></span>
<span id="cb109-3"><a href="feature-selection-and-cluster-analysis-1.html#cb109-3"></a>               <span class="dt">do.print =</span> <span class="ot">TRUE</span>, </span>
<span id="cb109-4"><a href="feature-selection-and-cluster-analysis-1.html#cb109-4"></a>               <span class="dt">pcs.print =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, </span>
<span id="cb109-5"><a href="feature-selection-and-cluster-analysis-1.html#cb109-5"></a>               <span class="dt">genes.print =</span> <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  MALAT1, LTB, IL32, IL7R, CD2, ACAP1, STK17A, CTSW, CD247, CCL5 
##     GIMAP5, AQP3, GZMA, CST7, TRAF3IP3, GZMK, MAL, HOPX, MYC, ETS1 
##     ITM2A, LYAR, BEX2, LDLRAP1, NKG7, KLRG1, ZAP70, RORA, SAMD3, TNFAIP8 
## Negative:  CST3, TYROBP, LST1, AIF1, FTL, FTH1, FCN1, LYZ, S100A9, TYMP 
##     FCER1G, CFD, LGALS1, LGALS2, S100A8, CTSS, SERPINA1, SPI1, IFITM3, PSAP 
##     CFP, IFI30, SAT1, COTL1, S100A11, NPC2, LGALS3, GSTP1, PYCARD, NCF2 
## PC_ 2 
## Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DRA, HLA-DQB1, LINC00926, CD79B, HLA-DRB1, CD74 
##     HLA-DPB1, HLA-DMA, HLA-DQA2, HLA-DRB5, HLA-DPA1, HLA-DMB, HVCN1, FCRLA, LTB, BLNK 
##     KIAA0125, P2RX5, IRF8, IGLL5, SWAP70, ARHGAP24, SMIM14, C16orf74, PPP1R14A, RP11-428G5.5 
## Negative:  NKG7, PRF1, GZMA, CST7, GZMB, FGFBP2, CTSW, GNLY, SPON2, GZMH 
##     CCL4, CCL5, FCGR3A, CD247, CLIC3, XCL2, AKR1C3, SRGN, HOPX, TTC38 
##     CTSC, S100A4, IGFBP7, ANXA1, IL32, ID2, ACTB, XCL1, APOBEC3G, SAMD3 
## PC_ 3 
## Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, CD74, MS4A1, HLA-DPB1, HLA-DPA1, HLA-DRB1, HLA-DQA2 
##     HLA-DRA, HLA-DRB5, TCL1A, LINC00926, HLA-DMA, HLA-DMB, HVCN1, FCRLA, KIAA0125, IRF8 
##     BLNK, GZMB, IGLL5, P2RX5, SWAP70, PLD4, FCGR3A, FGFBP2, SMIM14, LAT2 
## Negative:  IL7R, VIM, TMSB4X, S100A8, IL32, S100A6, S100A4, FYB, S100A9, MAL 
##     AQP3, RBP7, CD2, CD14, S100A10, LGALS2, S100A12, GIMAP4, NGFRAP1, ANXA1 
##     LDLRAP1, FOLR3, MS4A6A, FCN1, GIMAP5, LYZ, S100A11, TRABD2A, NELL2, AIF1 
## PC_ 4 
## Positive:  TMSB10, MALAT1, LGALS2, S100A8, S100A9, CD14, FCN1, LYZ, MS4A6A, S100A4 
##     CCL3, TYROBP, S100A6, ALDH2, GSTP1, FOLR3, RBP7, S100A10, S100A12, IFI6 
##     VIM, FCGR1A, NCF2, LGALS1, IL8, ANXA1, C1orf162, ID1, CTSS, FGFBP2 
## Negative:  PF4, PPBP, SDPR, HIST1H2AC, SPARC, NRGN, GNG11, TSC22D1, RGS18, ACRBP 
##     TUBB1, MMD, ITGA2B, PTCRA, GP9, TMEM40, CMTM5, CA2, CLU, MAP3K7CL 
##     F13A1, RP11-367G6.3, RUFY1, AP001189.4, GP1BA, MPP1, CD9, MYL9, TREML1, CLDN5 
## PC_ 5 
## Positive:  GZMB, FGFBP2, NKG7, GNLY, CCL4, PRF1, CST7, SPON2, GZMA, XCL2 
##     GZMH, CLIC3, CTSW, AKR1C3, TTC38, XCL1, CCL5, IGFBP7, S100A8, CCL3 
##     TYROBP, S100A9, HOPX, LGALS2, CD160, PTGDR, RBP7, CD14, HAVCR2, S100A12 
## Negative:  LTB, VIM, IL7R, AQP3, KIAA0101, PPA1, TYMS, LDHA, ZWINT, HN1 
##     MAL, ANXA5, CYTIP, CORO1B, IL32, CD2, TUBA1B, RRM2, PTGES3, TRADD 
##     COTL1, GPR183, PRDX1, BIRC5, ABRACL, GINS2, ATP5C1, GDI2, TRAF3IP3, DUT</code></pre>
</div>
<div id="visualizing-pca-in-different-ways" class="section level4">
<h4><span class="header-section-number">7.2.2.3</span> Visualizing PCA in Different Ways</h4>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="feature-selection-and-cluster-analysis-1.html#cb111-1"></a><span class="kw">DimPlot</span>(seurat, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_pca-1.png" width="672" /></p>
</div>
<div id="perform-linear-dimensional-reduction-ica" class="section level4">
<h4><span class="header-section-number">7.2.2.4</span> Perform linear dimensional reduction (ICA)</h4>
<p><strong>Task:</strong> Try running Independent Component Analysis. If you need help with the inputs try using the ?RunICA menu.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="feature-selection-and-cluster-analysis-1.html#cb112-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">RunICA</span>()</span></code></pre></div>
</div>
<div id="visualizing-ica-in-different-ways" class="section level4">
<h4><span class="header-section-number">7.2.2.5</span> Visualizing ICA in Different Ways</h4>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="feature-selection-and-cluster-analysis-1.html#cb113-1"></a><span class="kw">DimPlot</span>()</span></code></pre></div>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="feature-selection-and-cluster-analysis-1.html#cb114-1"></a><span class="co"># ProjectDim scores each gene in the dataset (including genes not included</span></span>
<span id="cb114-2"><a href="feature-selection-and-cluster-analysis-1.html#cb114-2"></a><span class="co"># in the PCA) based on their correlation with the calculated components.</span></span>
<span id="cb114-3"><a href="feature-selection-and-cluster-analysis-1.html#cb114-3"></a><span class="co"># Though we don&#39;t use this further here, it can be used to identify markers</span></span>
<span id="cb114-4"><a href="feature-selection-and-cluster-analysis-1.html#cb114-4"></a><span class="co"># that are strongly correlated with cellular heterogeneity, but may not have</span></span>
<span id="cb114-5"><a href="feature-selection-and-cluster-analysis-1.html#cb114-5"></a><span class="co"># passed through variable gene selection.  The results of the projected PCA</span></span>
<span id="cb114-6"><a href="feature-selection-and-cluster-analysis-1.html#cb114-6"></a><span class="co"># can be explored by setting use.full=T in the functions above</span></span>
<span id="cb114-7"><a href="feature-selection-and-cluster-analysis-1.html#cb114-7"></a>seurat &lt;-<span class="st"> </span><span class="kw">ProjectDim</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</span></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  MALAT1, LTB, IL32, IL7R, CD2, ACAP1, STK17A, CTSW, CD247, CCL5 
##     GIMAP5, AQP3, GZMA, CST7, TRAF3IP3, GZMK, MAL, HOPX, MYC, ETS1 
## Negative:  CST3, TYROBP, LST1, AIF1, FTL, FTH1, FCN1, LYZ, S100A9, TYMP 
##     FCER1G, CFD, LGALS1, LGALS2, S100A8, CTSS, SERPINA1, SPI1, IFITM3, PSAP 
## PC_ 2 
## Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DRA, HLA-DQB1, LINC00926, CD79B, HLA-DRB1, CD74 
##     HLA-DPB1, HLA-DMA, HLA-DQA2, HLA-DRB5, HLA-DPA1, HLA-DMB, HVCN1, FCRLA, LTB, BLNK 
## Negative:  NKG7, PRF1, GZMA, CST7, GZMB, FGFBP2, CTSW, GNLY, SPON2, GZMH 
##     CCL4, CCL5, FCGR3A, CD247, CLIC3, XCL2, AKR1C3, SRGN, HOPX, TTC38 
## PC_ 3 
## Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, CD74, MS4A1, HLA-DPB1, HLA-DPA1, HLA-DRB1, HLA-DQA2 
##     HLA-DRA, HLA-DRB5, TCL1A, LINC00926, HLA-DMA, HLA-DMB, HVCN1, FCRLA, KIAA0125, IRF8 
## Negative:  IL7R, VIM, TMSB4X, S100A8, IL32, S100A6, S100A4, FYB, S100A9, MAL 
##     AQP3, RBP7, CD2, CD14, S100A10, LGALS2, S100A12, GIMAP4, NGFRAP1, ANXA1 
## PC_ 4 
## Positive:  TMSB10, MALAT1, LGALS2, S100A8, S100A9, CD14, FCN1, LYZ, MS4A6A, S100A4 
##     CCL3, TYROBP, S100A6, ALDH2, GSTP1, FOLR3, RBP7, S100A10, S100A12, IFI6 
## Negative:  PF4, PPBP, SDPR, HIST1H2AC, SPARC, NRGN, GNG11, TSC22D1, RGS18, ACRBP 
##     TUBB1, MMD, ITGA2B, PTCRA, GP9, TMEM40, CMTM5, CA2, CLU, MAP3K7CL 
## PC_ 5 
## Positive:  GZMB, FGFBP2, NKG7, GNLY, CCL4, PRF1, CST7, SPON2, GZMA, XCL2 
##     GZMH, CLIC3, CTSW, AKR1C3, TTC38, XCL1, CCL5, IGFBP7, S100A8, CCL3 
## Negative:  LTB, VIM, IL7R, AQP3, KIAA0101, PPA1, TYMS, LDHA, ZWINT, HN1 
##     MAL, ANXA5, CYTIP, CORO1B, IL32, CD2, TUBA1B, RRM2, PTGES3, TRADD</code></pre>
</div>
<div id="genes-by-pcs" class="section level4">
<h4><span class="header-section-number">7.2.2.6</span> Genes by PCs</h4>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="feature-selection-and-cluster-analysis-1.html#cb116-1"></a><span class="kw">DimHeatmap</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb116-2"><a href="feature-selection-and-cluster-analysis-1.html#cb116-2"></a>          <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>, </span>
<span id="cb116-3"><a href="feature-selection-and-cluster-analysis-1.html#cb116-3"></a>          <span class="dt">cells =</span> <span class="dv">50</span>, </span>
<span id="cb116-4"><a href="feature-selection-and-cluster-analysis-1.html#cb116-4"></a>          <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</span>
<span id="cb116-5"><a href="feature-selection-and-cluster-analysis-1.html#cb116-5"></a>          <span class="dt">balanced =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/pca_heatmap-1.png" width="672" /></p>
<p>Check other PCs to plot</p>
<p><strong>Task: Check other PCs</strong></p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="feature-selection-and-cluster-analysis-1.html#cb117-1"></a><span class="kw">DimHeatmap</span>()</span></code></pre></div>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="feature-selection-and-cluster-analysis-1.html#cb118-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">JackStraw</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</span>
<span id="cb118-2"><a href="feature-selection-and-cluster-analysis-1.html#cb118-2"></a>seurat &lt;-<span class="st"> </span><span class="kw">ScoreJackStraw</span>(seurat, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>)</span></code></pre></div>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="feature-selection-and-cluster-analysis-1.html#cb119-1"></a><span class="kw">JackStrawPlot</span>(seurat, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>)</span></code></pre></div>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="feature-selection-and-cluster-analysis-1.html#cb120-1"></a><span class="kw">ElbowPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">ndims =</span> <span class="dv">10</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/pca_elbow-1.png" width="672" /></p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="feature-selection-and-cluster-analysis-1.html#cb121-1"></a><span class="co"># save.SNN = T saves the SNN so that the clustering algorithm can be rerun</span></span>
<span id="cb121-2"><a href="feature-selection-and-cluster-analysis-1.html#cb121-2"></a><span class="co"># using the same graph but with a different resolution value (see docs for</span></span>
<span id="cb121-3"><a href="feature-selection-and-cluster-analysis-1.html#cb121-3"></a><span class="co"># full details)</span></span>
<span id="cb121-4"><a href="feature-selection-and-cluster-analysis-1.html#cb121-4"></a><span class="kw">set.seed</span>(<span class="dv">2020</span>)</span>
<span id="cb121-5"><a href="feature-selection-and-cluster-analysis-1.html#cb121-5"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(<span class="dt">object =</span> seurat, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</span></code></pre></div>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="feature-selection-and-cluster-analysis-1.html#cb124-1"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb124-2"><a href="feature-selection-and-cluster-analysis-1.html#cb124-2"></a>                       <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, </span>
<span id="cb124-3"><a href="feature-selection-and-cluster-analysis-1.html#cb124-3"></a>                       <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, </span>
<span id="cb124-4"><a href="feature-selection-and-cluster-analysis-1.html#cb124-4"></a>                       <span class="dt">resolution =</span> <span class="fl">0.5</span>,</span>
<span id="cb124-5"><a href="feature-selection-and-cluster-analysis-1.html#cb124-5"></a>                       <span class="dt">random.seed =</span> <span class="dv">2020</span>)</span></code></pre></div>
<pre><code>## Warning: The following arguments are not used: reduction, dims

## Warning: The following arguments are not used: reduction, dims</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 2655
## Number of edges: 97412
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8712
## Number of communities: 8
## Elapsed time: 0 seconds</code></pre>
</div>
</div>
<div id="run-non-linear-dimensional-reduction-umaptsne" class="section level3">
<h3><span class="header-section-number">7.2.3</span> Run non-linear dimensional reduction (UMAP/tSNE)</h3>
<p>Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="feature-selection-and-cluster-analysis-1.html#cb127-1"></a><span class="kw">set.seed</span>(<span class="dv">2020</span>)</span>
<span id="cb127-2"><a href="feature-selection-and-cluster-analysis-1.html#cb127-2"></a>seurat &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(seurat, <span class="dt">reduction.use =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">perplexity=</span><span class="dv">10</span>)</span>
<span id="cb127-3"><a href="feature-selection-and-cluster-analysis-1.html#cb127-3"></a></span>
<span id="cb127-4"><a href="feature-selection-and-cluster-analysis-1.html#cb127-4"></a><span class="co"># note that you can set do.label=T to help label individual clusters</span></span>
<span id="cb127-5"><a href="feature-selection-and-cluster-analysis-1.html#cb127-5"></a><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/run_tsne-1.png" width="672" /></p>
<p><strong>Task: Try using UMAP for the non-linear dimension reduction technique (hint: <code>?RunUMAP</code>)</strong></p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="feature-selection-and-cluster-analysis-1.html#cb128-1"></a><span class="kw">set.seed</span>(<span class="dv">2020</span>)</span>
<span id="cb128-2"><a href="feature-selection-and-cluster-analysis-1.html#cb128-2"></a>seurat &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>()</span>
<span id="cb128-3"><a href="feature-selection-and-cluster-analysis-1.html#cb128-3"></a></span>
<span id="cb128-4"><a href="feature-selection-and-cluster-analysis-1.html#cb128-4"></a><span class="co"># note that you can set label=TRUE to help label individual clusters</span></span>
<span id="cb128-5"><a href="feature-selection-and-cluster-analysis-1.html#cb128-5"></a><span class="kw">DimPlot</span>()</span></code></pre></div>
<div id="finding-differentially-expressed-features-cluster-biomarkers" class="section level4">
<h4><span class="header-section-number">7.2.3.1</span> Finding differentially expressed features (cluster biomarkers)</h4>
<p>Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in ident.1), compared to all other cells. FindAllMarkers automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.</p>
<p>The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="feature-selection-and-cluster-analysis-1.html#cb129-1"></a><span class="co"># find all markers of cluster 1 using default parameters</span></span>
<span id="cb129-2"><a href="feature-selection-and-cluster-analysis-1.html#cb129-2"></a>cluster1.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">1</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>)</span>
<span id="cb129-3"><a href="feature-selection-and-cluster-analysis-1.html#cb129-3"></a></span>
<span id="cb129-4"><a href="feature-selection-and-cluster-analysis-1.html#cb129-4"></a><span class="kw">head</span>(cluster1.markers)</span></code></pre></div>
<pre><code>##             p_val avg_logFC pct.1 pct.2    p_val_adj
## LTB  2.398388e-95 0.8846283 0.982 0.643 3.285072e-91
## IL32 1.664226e-89 0.7966387 0.944 0.470 2.279490e-85
## LDHB 3.561648e-80 0.7142016 0.966 0.610 4.878389e-76
## IL7R 1.407429e-75 0.8335062 0.766 0.324 1.927756e-71
## CD3D 1.112564e-71 0.6199745 0.915 0.435 1.523880e-67
## AQP3 4.862610e-66 0.8776478 0.431 0.107 6.660317e-62</code></pre>
<p><strong>Task: Try tuning different parameters. How does that affect results?</strong></p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="feature-selection-and-cluster-analysis-1.html#cb131-1"></a><span class="co"># find all markers distinguishing cluster 5 from clusters 0 and 1</span></span>
<span id="cb131-2"><a href="feature-selection-and-cluster-analysis-1.html#cb131-2"></a>cluster5.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb131-3"><a href="feature-selection-and-cluster-analysis-1.html#cb131-3"></a>                                <span class="dt">ident.1 =</span> <span class="dv">5</span>, <span class="dt">ident.2 =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), </span>
<span id="cb131-4"><a href="feature-selection-and-cluster-analysis-1.html#cb131-4"></a>                                <span class="dt">min.pct =</span> ?? </span>
<span id="cb131-5"><a href="feature-selection-and-cluster-analysis-1.html#cb131-5"></a>                                <span class="dt">test.use =</span> ??, </span>
<span id="cb131-6"><a href="feature-selection-and-cluster-analysis-1.html#cb131-6"></a>                                <span class="dt">only.pos =</span> ??)</span>
<span id="cb131-7"><a href="feature-selection-and-cluster-analysis-1.html#cb131-7"></a></span>
<span id="cb131-8"><a href="feature-selection-and-cluster-analysis-1.html#cb131-8"></a><span class="kw">head</span>(cluster5.markers)</span></code></pre></div>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="feature-selection-and-cluster-analysis-1.html#cb132-1"></a>cluster3.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb132-2"><a href="feature-selection-and-cluster-analysis-1.html#cb132-2"></a>                                <span class="dt">ident.1 =</span> <span class="dv">3</span>, </span>
<span id="cb132-3"><a href="feature-selection-and-cluster-analysis-1.html#cb132-3"></a>                                <span class="dt">thresh.use =</span> <span class="fl">0.25</span>, </span>
<span id="cb132-4"><a href="feature-selection-and-cluster-analysis-1.html#cb132-4"></a>                                <span class="dt">test.use =</span> <span class="st">&quot;roc&quot;</span>, </span>
<span id="cb132-5"><a href="feature-selection-and-cluster-analysis-1.html#cb132-5"></a>                                <span class="dt">only.pos =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## Warning: The following arguments are not used: thresh.use</code></pre>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="feature-selection-and-cluster-analysis-1.html#cb134-1"></a><span class="kw">head</span>(cluster3.markers)</span></code></pre></div>
<pre><code>##          myAUC avg_diff power pct.1 pct.2
## CD74     0.984 2.029721 0.968 1.000 0.824
## CD79A    0.966 2.968324 0.932 0.938 0.043
## HLA-DRA  0.961 1.917675 0.922 1.000 0.492
## CD79B    0.949 2.394172 0.898 0.927 0.145
## HLA-DPB1 0.934 1.635536 0.868 0.988 0.447
## MS4A1    0.925 2.344449 0.850 0.865 0.053</code></pre>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="feature-selection-and-cluster-analysis-1.html#cb136-1"></a><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;CD79A&quot;</span>))</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_cluster3_markers-1.png" width="672" /></p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="feature-selection-and-cluster-analysis-1.html#cb137-1"></a><span class="co"># you can plot raw UMI counts as well</span></span>
<span id="cb137-2"><a href="feature-selection-and-cluster-analysis-1.html#cb137-2"></a><span class="kw">VlnPlot</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb137-3"><a href="feature-selection-and-cluster-analysis-1.html#cb137-3"></a>        <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;NKG7&quot;</span>, <span class="st">&quot;PF4&quot;</span>), </span>
<span id="cb137-4"><a href="feature-selection-and-cluster-analysis-1.html#cb137-4"></a>        <span class="dt">log =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/plot_by_umi-1.png" width="672" /></p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="feature-selection-and-cluster-analysis-1.html#cb138-1"></a><span class="co"># find markers for every cluster compared to all remaining cells, report</span></span>
<span id="cb138-2"><a href="feature-selection-and-cluster-analysis-1.html#cb138-2"></a><span class="co"># only the positive ones</span></span>
<span id="cb138-3"><a href="feature-selection-and-cluster-analysis-1.html#cb138-3"></a>pbmc.markers &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">only.pos =</span> <span class="ot">TRUE</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, <span class="dt">thresh.use =</span> <span class="fl">0.25</span>)</span></code></pre></div>
<pre><code>## Calculating cluster 0</code></pre>
<pre><code>## Calculating cluster 1</code></pre>
<pre><code>## Calculating cluster 2</code></pre>
<pre><code>## Calculating cluster 3</code></pre>
<pre><code>## Calculating cluster 4</code></pre>
<pre><code>## Calculating cluster 5</code></pre>
<pre><code>## Calculating cluster 6</code></pre>
<pre><code>## Calculating cluster 7</code></pre>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="feature-selection-and-cluster-analysis-1.html#cb147-1"></a>pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">2</span>, avg_logFC)</span></code></pre></div>
<pre><code>## # A tibble: 16 x 7
## # Groups:   cluster [8]
##        p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene     
##        &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;    
##  1 1.31e- 92     0.984 0.458 0.108 1.79e- 88 0       CCR7     
##  2 1.08e- 49     0.747 0.345 0.108 1.48e- 45 0       PRKCQ-AS1
##  3 2.40e- 95     0.885 0.982 0.643 3.29e- 91 1       LTB      
##  4 4.86e- 66     0.878 0.431 0.107 6.66e- 62 1       AQP3     
##  5 0.            3.84  0.996 0.217 0.        2       S100A9   
##  6 0.            3.77  0.971 0.123 0.        2       S100A8   
##  7 0.            2.97  0.938 0.043 0.        3       CD79A    
##  8 4.00e-276     2.50  0.63  0.022 5.48e-272 3       TCL1A    
##  9 1.16e-196     2.12  0.964 0.237 1.59e-192 4       CCL5     
## 10 2.16e-187     2.13  0.617 0.056 2.96e-183 4       GZMK     
## 11 6.15e-171     2.26  0.929 0.137 8.43e-167 5       FCGR3A   
## 12 4.31e-118     2.06  0.964 0.314 5.90e-114 5       LST1     
## 13 6.72e-267     3.35  0.961 0.07  9.20e-263 6       GZMB     
## 14 1.53e-175     3.42  0.935 0.135 2.10e-171 6       GNLY     
## 15 9.11e-253     2.68  0.829 0.01  1.25e-248 7       FCER1A   
## 16 6.56e- 23     1.96  1     0.51  8.98e- 19 7       HLA-DPB1</code></pre>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="feature-selection-and-cluster-analysis-1.html#cb149-1"></a><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb149-2"><a href="feature-selection-and-cluster-analysis-1.html#cb149-2"></a>            <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;GNLY&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;FCER1A&quot;</span>, <span class="st">&quot;FCGR3A&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;PPBP&quot;</span>, <span class="st">&quot;CD8A&quot;</span>),</span>
<span id="cb149-3"><a href="feature-selection-and-cluster-analysis-1.html#cb149-3"></a>            <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;blue&quot;</span>), </span>
<span id="cb149-4"><a href="feature-selection-and-cluster-analysis-1.html#cb149-4"></a>            <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>)</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/gene_tsne-1.png" width="672" /></p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="feature-selection-and-cluster-analysis-1.html#cb150-1"></a>top10 &lt;-<span class="st"> </span>pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">10</span>, avg_logFC)</span>
<span id="cb150-2"><a href="feature-selection-and-cluster-analysis-1.html#cb150-2"></a><span class="co"># setting slim.col.label to TRUE will print just the cluster IDS instead of</span></span>
<span id="cb150-3"><a href="feature-selection-and-cluster-analysis-1.html#cb150-3"></a><span class="co"># every cell name</span></span>
<span id="cb150-4"><a href="feature-selection-and-cluster-analysis-1.html#cb150-4"></a><span class="kw">DoHeatmap</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> top10<span class="op">$</span>gene, <span class="dt">label =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## Warning in DoHeatmap(object = seurat, features = top10$gene, label = TRUE): The following
## features were omitted as they were not found in the scale.data slot for the RNA assay: RHOC,
## RP11-290F20.3, GZMM, CD8A, VPREB3, CD40LG, CD7, PIK3IP1, CD3E, NOSIP, LEF1, PRKCQ-AS1, CD3D,
## LDHB, CCR7, RPS3A</code></pre>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/do_heatmap-1.png" width="672" /></p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="feature-selection-and-cluster-analysis-1.html#cb152-1"></a>new.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Memory CD4 T&quot;</span>, <span class="st">&quot;Naive CD4 T&quot;</span>, <span class="st">&quot;CD14+ Mono&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;CD8 T&quot;</span>, <span class="st">&quot;FCGR3A+ Mono&quot;</span>, <span class="st">&quot;NK&quot;</span>, <span class="st">&quot;DC&quot;</span>, <span class="st">&quot;Mk&quot;</span>)</span>
<span id="cb152-2"><a href="feature-selection-and-cluster-analysis-1.html#cb152-2"></a><span class="kw">names</span>(<span class="dt">x =</span> new.cluster.ids) &lt;-<span class="st"> </span><span class="kw">levels</span>(<span class="dt">x =</span> seurat)</span>
<span id="cb152-3"><a href="feature-selection-and-cluster-analysis-1.html#cb152-3"></a>seurat &lt;-<span class="st"> </span><span class="kw">RenameIdents</span>(<span class="dt">object =</span> seurat, new.cluster.ids)</span></code></pre></div>
<pre><code>## Warning: Cannot find identity NA</code></pre>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="feature-selection-and-cluster-analysis-1.html#cb154-1"></a><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, <span class="dt">reduction =</span> <span class="st">&#39;tsne&#39;</span>, <span class="dt">label =</span> <span class="ot">TRUE</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span><span class="kw">NoLegend</span>()</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/new_names-1.png" width="672" /></p>
</div>
<div id="further-subdivisions-within-cell-types" class="section level4">
<h4><span class="header-section-number">7.2.3.2</span> Further subdivisions within cell types</h4>
<p>If you perturb some of our parameter choices above (for example, setting resolution=0.8 or changing the number of PCs), you might see the CD4 T cells subdivide into two groups. You can explore this subdivision to find markers separating the two T cell subsets. However, before reclustering (which will overwrite <a href="mailto:object@ident" class="email">object@ident</a>), we can stash our renamed identities to be easily recovered later.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="feature-selection-and-cluster-analysis-1.html#cb155-1"></a><span class="co"># First lets stash our identities for later</span></span>
<span id="cb155-2"><a href="feature-selection-and-cluster-analysis-1.html#cb155-2"></a>seurat[[<span class="st">&quot;ClusterNames_0.6&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">Idents</span>(<span class="dt">object =</span> seurat)</span>
<span id="cb155-3"><a href="feature-selection-and-cluster-analysis-1.html#cb155-3"></a><span class="co"># Note that if you set save.snn=T above, you don&#39;t need to recalculate the</span></span>
<span id="cb155-4"><a href="feature-selection-and-cluster-analysis-1.html#cb155-4"></a><span class="co"># SNN, and can simply put: pbmc &lt;- FindClusters(pbmc,resolution = 0.8)</span></span>
<span id="cb155-5"><a href="feature-selection-and-cluster-analysis-1.html#cb155-5"></a>seurat &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb155-6"><a href="feature-selection-and-cluster-analysis-1.html#cb155-6"></a>                       <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, </span>
<span id="cb155-7"><a href="feature-selection-and-cluster-analysis-1.html#cb155-7"></a>                       <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, </span>
<span id="cb155-8"><a href="feature-selection-and-cluster-analysis-1.html#cb155-8"></a>                       <span class="dt">resolution =</span> <span class="fl">0.8</span>)</span></code></pre></div>
<pre><code>## Warning: The following arguments are not used: reduction, dims

## Warning: The following arguments are not used: reduction, dims</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 2655
## Number of edges: 97412
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8274
## Number of communities: 10
## Elapsed time: 0 seconds</code></pre>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="feature-selection-and-cluster-analysis-1.html#cb158-1"></a><span class="kw">set.seed</span>(<span class="dv">2020</span>)</span>
<span id="cb158-2"><a href="feature-selection-and-cluster-analysis-1.html#cb158-2"></a><span class="co">## Warning in BuildSNN(object = object, genes.use = genes.use, reduction.type</span></span>
<span id="cb158-3"><a href="feature-selection-and-cluster-analysis-1.html#cb158-3"></a><span class="co">## = reduction.type, : Build parameters exactly match those of already</span></span>
<span id="cb158-4"><a href="feature-selection-and-cluster-analysis-1.html#cb158-4"></a><span class="co">## computed and stored SNN. To force recalculation, set force.recalc to TRUE.</span></span>
<span id="cb158-5"><a href="feature-selection-and-cluster-analysis-1.html#cb158-5"></a><span class="co"># Demonstration of how to plot two tSNE plots side by side, and how to color</span></span>
<span id="cb158-6"><a href="feature-selection-and-cluster-analysis-1.html#cb158-6"></a><span class="co"># points based on different criteria</span></span>
<span id="cb158-7"><a href="feature-selection-and-cluster-analysis-1.html#cb158-7"></a>plot1 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb158-8"><a href="feature-selection-and-cluster-analysis-1.html#cb158-8"></a>                 <span class="dt">reduction=</span> <span class="st">&quot;tsne&quot;</span>,</span>
<span id="cb158-9"><a href="feature-selection-and-cluster-analysis-1.html#cb158-9"></a>                 <span class="dt">label =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">NoLegend</span>()</span>
<span id="cb158-10"><a href="feature-selection-and-cluster-analysis-1.html#cb158-10"></a></span>
<span id="cb158-11"><a href="feature-selection-and-cluster-analysis-1.html#cb158-11"></a>plot2 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(<span class="dt">object =</span> seurat, </span>
<span id="cb158-12"><a href="feature-selection-and-cluster-analysis-1.html#cb158-12"></a>                  <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>, </span>
<span id="cb158-13"><a href="feature-selection-and-cluster-analysis-1.html#cb158-13"></a>                  <span class="dt">group.by =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>,</span>
<span id="cb158-14"><a href="feature-selection-and-cluster-analysis-1.html#cb158-14"></a>                  <span class="dt">label =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">NoLegend</span>()</span>
<span id="cb158-15"><a href="feature-selection-and-cluster-analysis-1.html#cb158-15"></a></span>
<span id="cb158-16"><a href="feature-selection-and-cluster-analysis-1.html#cb158-16"></a><span class="kw">CombinePlots</span>(<span class="kw">list</span>(plot1, plot2))</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/compare_params-1.png" width="672" /></p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="feature-selection-and-cluster-analysis-1.html#cb159-1"></a><span class="co"># Find discriminating markers</span></span>
<span id="cb159-2"><a href="feature-selection-and-cluster-analysis-1.html#cb159-2"></a>tcell.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>)</span>
<span id="cb159-3"><a href="feature-selection-and-cluster-analysis-1.html#cb159-3"></a></span>
<span id="cb159-4"><a href="feature-selection-and-cluster-analysis-1.html#cb159-4"></a><span class="co"># Most of the markers tend to be expressed in C1 (i.e. S100A4). However, we</span></span>
<span id="cb159-5"><a href="feature-selection-and-cluster-analysis-1.html#cb159-5"></a><span class="co"># can see that CCR7 is upregulated in C0, strongly indicating that we can</span></span>
<span id="cb159-6"><a href="feature-selection-and-cluster-analysis-1.html#cb159-6"></a><span class="co"># differentiate memory from naive CD4 cells.  cols demarcates the color</span></span>
<span id="cb159-7"><a href="feature-selection-and-cluster-analysis-1.html#cb159-7"></a><span class="co"># palette from low to high expression</span></span>
<span id="cb159-8"><a href="feature-selection-and-cluster-analysis-1.html#cb159-8"></a><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> seurat, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;S100A4&quot;</span>, <span class="st">&quot;CCR7&quot;</span>), <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;blue&quot;</span>))</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/tcell_markers-1.png" width="672" /></p>
</div>
</div>
</div>
<div id="feature-selection" class="section level2">
<h2><span class="header-section-number">7.3</span> Feature Selection</h2>
<div id="differential-expression-analysis" class="section level3">
<h3><span class="header-section-number">7.3.1</span> Differential Expression Analysis</h3>
<div id="differential-expression-tests" class="section level4">
<h4><span class="header-section-number">7.3.1.1</span> Differential Expression Tests</h4>
<p>One of the most commonly performed tasks for RNA-seq data is differential gene expression (DE) analysis. Although well-established tools exist for such analysis in bulk RNA-seq data, methods for scRNA-seq data are just emerging. Given the special characteristics of scRNA-seq data, including generally low library sizes, high noise levels and a large fraction of so-called ‘dropout’ events, it is unclear whether DE methods that have been developed for bulk RNA-seq are suitable also for scRNA-seq. Check the help page out for the FindMarkers function by using <code> ?FindMarkers </code></p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="feature-selection-and-cluster-analysis-1.html#cb160-1"></a><span class="co">## Differential expression using t-test</span></span>
<span id="cb160-2"><a href="feature-selection-and-cluster-analysis-1.html#cb160-2"></a><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>, <span class="dt">test.use =</span> <span class="st">&quot;t&quot;</span>)</span></code></pre></div>
<p><strong>Task:</strong> Try to use different test for diffential expression analysis (hint: <code>?FindMarkers</code>)</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="feature-selection-and-cluster-analysis-1.html#cb161-1"></a><span class="co">## Use the help function and run other tests.  Do they find similar markers?</span></span>
<span id="cb161-2"><a href="feature-selection-and-cluster-analysis-1.html#cb161-2"></a><span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>, <span class="dt">test.use =</span> ??)</span></code></pre></div>
</div>
</div>
<div id="check-clusters" class="section level3">
<h3><span class="header-section-number">7.3.2</span> Check Clusters</h3>
<p>How do we test the cell types identified? How do we know how reliable they are?</p>
<p>Use Classifier to predict cell cluster. See how it predicts using hold out data.
<a href="https://satijalab.org/seurat/v3.0/integration.html">reference</a></p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="feature-selection-and-cluster-analysis-1.html#cb162-1"></a><span class="co"># Assign the test object a three level attribute</span></span>
<span id="cb162-2"><a href="feature-selection-and-cluster-analysis-1.html#cb162-2"></a>groups &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">&quot;train&quot;</span>, <span class="st">&quot;test&quot;</span>), <span class="dt">size =</span> <span class="kw">NROW</span>(seurat<span class="op">@</span>meta.data), <span class="dt">replace =</span> <span class="ot">TRUE</span>, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="fl">0.8</span>, <span class="fl">0.2</span>))</span>
<span id="cb162-3"><a href="feature-selection-and-cluster-analysis-1.html#cb162-3"></a><span class="kw">names</span>(groups) &lt;-<span class="st"> </span><span class="kw">colnames</span>(seurat)</span>
<span id="cb162-4"><a href="feature-selection-and-cluster-analysis-1.html#cb162-4"></a>seurat &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> seurat, <span class="dt">metadata =</span> groups, <span class="dt">col.name =</span> <span class="st">&quot;group&quot;</span>)</span>
<span id="cb162-5"><a href="feature-selection-and-cluster-analysis-1.html#cb162-5"></a></span>
<span id="cb162-6"><a href="feature-selection-and-cluster-analysis-1.html#cb162-6"></a><span class="co"># Find Anchors</span></span>
<span id="cb162-7"><a href="feature-selection-and-cluster-analysis-1.html#cb162-7"></a>seurat.list &lt;-<span class="st"> </span><span class="kw">SplitObject</span>(seurat, <span class="dt">split.by =</span> <span class="st">&quot;group&quot;</span>)</span>
<span id="cb162-8"><a href="feature-selection-and-cluster-analysis-1.html#cb162-8"></a>seurat.anchors &lt;-<span class="st"> </span><span class="kw">FindIntegrationAnchors</span>(<span class="dt">object.list =</span> seurat.list, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</span></code></pre></div>
<pre><code>## Computing 2000 integration features</code></pre>
<pre><code>## Scaling features for provided objects</code></pre>
<pre><code>## Finding all pairwise anchors</code></pre>
<pre><code>## Running CCA</code></pre>
<pre><code>## Merging objects</code></pre>
<pre><code>## Finding neighborhoods</code></pre>
<pre><code>## Finding anchors</code></pre>
<pre><code>##  Found 2597 anchors</code></pre>
<pre><code>## Filtering anchors</code></pre>
<pre><code>##  Retained 1400 anchors</code></pre>
<pre><code>## Extracting within-dataset neighbors</code></pre>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="feature-selection-and-cluster-analysis-1.html#cb174-1"></a>seurat.integrated &lt;-<span class="st"> </span><span class="kw">IntegrateData</span>(<span class="dt">anchorset =</span> seurat.anchors, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</span></code></pre></div>
<pre><code>## Merging dataset 1 into 2</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<pre><code>## Warning: Adding a command log without an assay associated with it</code></pre>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="feature-selection-and-cluster-analysis-1.html#cb181-1"></a>seurat.query &lt;-<span class="st"> </span>seurat.list[[<span class="st">&quot;train&quot;</span>]]</span>
<span id="cb181-2"><a href="feature-selection-and-cluster-analysis-1.html#cb181-2"></a>seurat.anchors &lt;-<span class="st"> </span><span class="kw">FindTransferAnchors</span>(<span class="dt">reference =</span> seurat.integrated, </span>
<span id="cb181-3"><a href="feature-selection-and-cluster-analysis-1.html#cb181-3"></a>                                      <span class="dt">query =</span> seurat.query, </span>
<span id="cb181-4"><a href="feature-selection-and-cluster-analysis-1.html#cb181-4"></a>                                      <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</span></code></pre></div>
<pre><code>## Performing PCA on the provided reference using 2000 features as input.</code></pre>
<pre><code>## Projecting PCA</code></pre>
<pre><code>## Finding neighborhoods</code></pre>
<pre><code>## Finding anchors</code></pre>
<pre><code>##  Found 6141 anchors</code></pre>
<pre><code>## Filtering anchors</code></pre>
<pre><code>##  Retained 4729 anchors</code></pre>
<pre><code>## Extracting within-dataset neighbors</code></pre>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="feature-selection-and-cluster-analysis-1.html#cb190-1"></a>predictions &lt;-<span class="st"> </span><span class="kw">TransferData</span>(<span class="dt">anchorset =</span> seurat.anchors, </span>
<span id="cb190-2"><a href="feature-selection-and-cluster-analysis-1.html#cb190-2"></a>                            <span class="dt">refdata =</span> seurat.integrated<span class="op">$</span>ClusterNames_<span class="fl">0.6</span>, </span>
<span id="cb190-3"><a href="feature-selection-and-cluster-analysis-1.html#cb190-3"></a>                            <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>)</span></code></pre></div>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Predicting cell labels</code></pre>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="feature-selection-and-cluster-analysis-1.html#cb194-1"></a>seurat.query &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(seurat.query, <span class="dt">metadata =</span> predictions)</span>
<span id="cb194-2"><a href="feature-selection-and-cluster-analysis-1.html#cb194-2"></a><span class="kw">table</span>(seurat.query<span class="op">@</span>meta.data<span class="op">$</span>ClusterNames_<span class="fl">0.6</span>, seurat.query<span class="op">@</span>meta.data<span class="op">$</span>predicted.id)</span></code></pre></div>
<pre><code>##               
##                  B CD14+ Mono CD8 T  DC FCGR3A+ Mono Memory CD4 T Naive CD4 T  NK
##   Memory CD4 T   0          0     2   0            1          542          11   0
##   Naive CD4 T    0          0     0   0            0           53         334   0
##   CD14+ Mono     0        365     0   0            1            0           0   0
##   B            268          0     0   0            0            0           0   0
##   CD8 T          0          0   225   0            0           12           7   0
##   FCGR3A+ Mono   0          8     0   0          130            0           0   0
##   NK             1          0     7   0            0            0           0 122
##   DC             0          1     0  28            0            0           0   0</code></pre>
</div>
<div id="view-entire-object-structure" class="section level3">
<h3><span class="header-section-number">7.3.3</span> View Entire Object Structure</h3>
<p>Notice all the slots and elements added to the object.</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="feature-selection-and-cluster-analysis-1.html#cb196-1"></a><span class="kw">str</span>(seurat)</span></code></pre></div>
<pre><code>## Formal class &#39;Seurat&#39; [package &quot;Seurat&quot;] with 12 slots
##   ..@ assays      :List of 1
##   .. ..$ RNA:Formal class &#39;Assay&#39; [package &quot;Seurat&quot;] with 8 slots
##   .. .. .. ..@ counts       :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. .. .. .. .. ..@ i       : int [1:2269052] 29 73 80 147 162 183 185 226 228 229 ...
##   .. .. .. .. .. ..@ p       : int [1:2656] 0 779 2131 3260 4220 4741 5522 6304 7094 7626 ...
##   .. .. .. .. .. ..@ Dim     : int [1:2] 13697 2655
##   .. .. .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. .. .. ..$ : chr [1:13697] &quot;AL627309.1&quot; &quot;AP006222.2&quot; &quot;RP11-206L10.2&quot; &quot;RP11-206L10.9&quot; ...
##   .. .. .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..@ x       : num [1:2269052] 1 1 2 1 1 1 1 41 1 1 ...
##   .. .. .. .. .. ..@ factors : list()
##   .. .. .. ..@ data         :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. .. .. .. .. ..@ i       : int [1:2269052] 29 73 80 147 162 183 185 226 228 229 ...
##   .. .. .. .. .. ..@ p       : int [1:2656] 0 779 2131 3260 4220 4741 5522 6304 7094 7626 ...
##   .. .. .. .. .. ..@ Dim     : int [1:2] 13697 2655
##   .. .. .. .. .. ..@ Dimnames:List of 2
##   .. .. .. .. .. .. ..$ : chr [1:13697] &quot;AL627309.1&quot; &quot;AP006222.2&quot; &quot;RP11-206L10.2&quot; &quot;RP11-206L10.9&quot; ...
##   .. .. .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..@ x       : num [1:2269052] 1.64 1.64 2.23 1.64 1.64 ...
##   .. .. .. .. .. ..@ factors : list()
##   .. .. .. ..@ scale.data   : num [1:2000, 1:2655] -0.8423 -0.2675 1.5408 -0.0487 -0.4588 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2000] &quot;ISG15&quot; &quot;CPSF3L&quot; &quot;MRPL20&quot; &quot;ATAD3C&quot; ...
##   .. .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. ..@ key          : chr &quot;rna_&quot;
##   .. .. .. ..@ assay.orig   : NULL
##   .. .. .. ..@ var.features : chr [1:2000] &quot;S100A9&quot; &quot;LYZ&quot; &quot;IGLL5&quot; &quot;GNLY&quot; ...
##   .. .. .. ..@ meta.features:&#39;data.frame&#39;:   13697 obs. of  5 variables:
##   .. .. .. .. ..$ vst.mean                 : num [1:13697] 0.00339 0.00113 0.00188 0.00113 0.00678 ...
##   .. .. .. .. ..$ vst.variance             : num [1:13697] 0.00338 0.00113 0.00188 0.00113 0.00674 ...
##   .. .. .. .. ..$ vst.variance.expected    : num [1:13697] 0.00364 0.00113 0.00196 0.00113 0.00748 ...
##   .. .. .. .. ..$ vst.variance.standardized: num [1:13697] 0.928 0.999 0.961 0.999 0.901 ...
##   .. .. .. .. ..$ vst.variable             : logi [1:13697] FALSE FALSE FALSE FALSE FALSE FALSE ...
##   .. .. .. ..@ misc         : NULL
##   ..@ meta.data   :&#39;data.frame&#39;: 2655 obs. of  13 variables:
##   .. ..$ orig.ident      : Factor w/ 1 level &quot;10X_PBMC&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ nCount_RNA      : num [1:2655] 2419 4903 3147 2639 980 ...
##   .. ..$ nFeature_RNA    : int [1:2655] 779 1352 1129 960 521 781 782 790 532 550 ...
##   .. ..$ percent.mt      : num [1:2655] 3.02 3.79 0.89 1.74 1.22 ...
##   .. ..$ S.Score         : num [1:2655] 0.0825 -0.0276 -0.0124 0.0405 -0.0273 ...
##   .. ..$ G2M.Score       : num [1:2655] -0.0324 -0.0517 0.0833 0.0155 0.0334 ...
##   .. ..$ Phase           : Factor w/ 3 levels &quot;G1&quot;,&quot;G2M&quot;,&quot;S&quot;: 3 1 2 3 2 1 1 2 1 3 ...
##   .. ..$ old.ident       : Factor w/ 1 level &quot;10X_PBMC&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ RNA_snn_res.0.5 : Factor w/ 8 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 2 4 2 6 7 2 5 5 1 6 ...
##   .. ..$ seurat_clusters : Factor w/ 10 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 1 3 1 8 9 1 5 5 2 8 ...
##   .. ..$ ClusterNames_0.6: Factor w/ 8 levels &quot;Memory CD4 T&quot;,..: 2 4 2 6 7 2 5 5 1 6 ...
##   .. ..$ RNA_snn_res.0.8 : Factor w/ 10 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 1 3 1 8 9 1 5 5 2 8 ...
##   .. ..$ group           : chr [1:2655] &quot;test&quot; &quot;test&quot; &quot;train&quot; &quot;train&quot; ...
##   ..@ active.assay: chr &quot;RNA&quot;
##   ..@ active.ident: Factor w/ 10 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 1 3 1 8 9 1 5 5 2 8 ...
##   .. ..- attr(*, &quot;names&quot;)= chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   ..@ graphs      :List of 2
##   .. ..$ RNA_nn :Formal class &#39;Graph&#39; [package &quot;Seurat&quot;] with 7 slots
##   .. .. .. ..@ assay.used: chr &quot;RNA&quot;
##   .. .. .. ..@ i         : int [1:53100] 0 102 167 417 623 1207 1260 1835 2148 2155 ...
##   .. .. .. ..@ p         : int [1:2656] 0 12 32 51 75 86 90 126 153 164 ...
##   .. .. .. ..@ Dim       : int [1:2] 2655 2655
##   .. .. .. ..@ Dimnames  :List of 2
##   .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. ..@ x         : num [1:53100] 1 1 1 1 1 1 1 1 1 1 ...
##   .. .. .. ..@ factors   : list()
##   .. ..$ RNA_snn:Formal class &#39;Graph&#39; [package &quot;Seurat&quot;] with 7 slots
##   .. .. .. ..@ assay.used: chr &quot;RNA&quot;
##   .. .. .. ..@ i         : int [1:197479] 0 84 102 167 205 383 417 457 480 489 ...
##   .. .. .. ..@ p         : int [1:2656] 0 62 115 168 225 270 319 403 479 512 ...
##   .. .. .. ..@ Dim       : int [1:2] 2655 2655
##   .. .. .. ..@ Dimnames  :List of 2
##   .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. ..@ x         : num [1:197479] 1 0.0811 0.1765 0.0811 0.1111 ...
##   .. .. .. ..@ factors   : list()
##   ..@ neighbors   : list()
##   ..@ reductions  :List of 2
##   .. ..$ pca :Formal class &#39;DimReduc&#39; [package &quot;Seurat&quot;] with 9 slots
##   .. .. .. ..@ cell.embeddings           : num [1:2655, 1:50] 5.28 1.22 1.89 -12.14 2.44 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..$ : chr [1:50] &quot;PC_1&quot; &quot;PC_2&quot; &quot;PC_3&quot; &quot;PC_4&quot; ...
##   .. .. .. ..@ feature.loadings          : num [1:2000, 1:50] -0.11749 -0.11826 0.00861 0.01508 -0.0136 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2000] &quot;S100A9&quot; &quot;LYZ&quot; &quot;IGLL5&quot; &quot;GNLY&quot; ...
##   .. .. .. .. .. ..$ : chr [1:50] &quot;PC_1&quot; &quot;PC_2&quot; &quot;PC_3&quot; &quot;PC_4&quot; ...
##   .. .. .. ..@ feature.loadings.projected: num [1:2000, 1:50] -7900 388 -1755 248 -1734 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2000] &quot;ISG15&quot; &quot;CPSF3L&quot; &quot;MRPL20&quot; &quot;ATAD3C&quot; ...
##   .. .. .. .. .. ..$ : chr [1:50] &quot;PC_1&quot; &quot;PC_2&quot; &quot;PC_3&quot; &quot;PC_4&quot; ...
##   .. .. .. ..@ assay.used                : chr &quot;RNA&quot;
##   .. .. .. ..@ global                    : logi FALSE
##   .. .. .. ..@ stdev                     : num [1:50] 6.96 4.45 3.77 3.35 3.15 ...
##   .. .. .. ..@ key                       : chr &quot;PC_&quot;
##   .. .. .. ..@ jackstraw                 :Formal class &#39;JackStrawData&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] 
##   .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] 
##   .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] 
##   .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] 
##   .. .. .. ..@ misc                      :List of 1
##   .. .. .. .. ..$ total.variance: num 1721
##   .. ..$ tsne:Formal class &#39;DimReduc&#39; [package &quot;Seurat&quot;] with 9 slots
##   .. .. .. ..@ cell.embeddings           : num [1:2655, 1:2] 2.08 37.16 15.8 -10.98 22.77 ...
##   .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. .. .. .. .. ..$ : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. .. ..$ : chr [1:2] &quot;tSNE_1&quot; &quot;tSNE_2&quot;
##   .. .. .. ..@ feature.loadings          : num[0 , 0 ] 
##   .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] 
##   .. .. .. ..@ assay.used                : chr &quot;RNA&quot;
##   .. .. .. ..@ global                    : logi TRUE
##   .. .. .. ..@ stdev                     : num(0) 
##   .. .. .. ..@ key                       : chr &quot;tSNE_&quot;
##   .. .. .. ..@ jackstraw                 :Formal class &#39;JackStrawData&#39; [package &quot;Seurat&quot;] with 4 slots
##   .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] 
##   .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] 
##   .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] 
##   .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] 
##   .. .. .. ..@ misc                      : list()
##   ..@ project.name: chr &quot;10X_PBMC&quot;
##   ..@ misc        : list()
##   ..@ version     :Classes &#39;package_version&#39;, &#39;numeric_version&#39;  hidden list of 1
##   .. ..$ : int [1:3] 3 1 2
##   ..@ commands    :List of 8
##   .. ..$ NormalizeData.RNA       :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;NormalizeData.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:49:19&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;NormalizeData(object = seurat, normalization.method = \&quot;LogNormalize\&quot;, &quot; &quot;    scale.factor = 10000)&quot;
##   .. .. .. ..@ params     :List of 5
##   .. .. .. .. ..$ assay               : chr &quot;RNA&quot;
##   .. .. .. .. ..$ normalization.method: chr &quot;LogNormalize&quot;
##   .. .. .. .. ..$ scale.factor        : num 10000
##   .. .. .. .. ..$ margin              : num 1
##   .. .. .. .. ..$ verbose             : logi TRUE
##   .. ..$ FindVariableFeatures.RNA:Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;FindVariableFeatures.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:49:20&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;FindVariableFeatures(object = seurat, mean.function = ExpMean, &quot; &quot;    dispersion.function = LogVMR)&quot;
##   .. .. .. ..@ params     :List of 12
##   .. .. .. .. ..$ assay              : chr &quot;RNA&quot;
##   .. .. .. .. ..$ selection.method   : chr &quot;vst&quot;
##   .. .. .. .. ..$ loess.span         : num 0.3
##   .. .. .. .. ..$ clip.max           : chr &quot;auto&quot;
##   .. .. .. .. ..$ mean.function      :function (x, ...)  
##   .. .. .. .. ..$ dispersion.function:function (x, ...)  
##   .. .. .. .. ..$ num.bin            : num 20
##   .. .. .. .. ..$ binning.method     : chr &quot;equal_width&quot;
##   .. .. .. .. ..$ nfeatures          : num 2000
##   .. .. .. .. ..$ mean.cutoff        : num [1:2] 0.1 8
##   .. .. .. .. ..$ dispersion.cutoff  : num [1:2] 1 Inf
##   .. .. .. .. ..$ verbose            : logi TRUE
##   .. ..$ ScaleData.RNA           :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;ScaleData.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:49:29&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;ScaleData(object = seurat, vars.to.regress = c(\&quot;batch\&quot;, &quot; &quot;    \&quot;percent.mt\&quot;))&quot;
##   .. .. .. ..@ params     :List of 11
##   .. .. .. .. ..$ features          : chr [1:2000] &quot;S100A9&quot; &quot;LYZ&quot; &quot;IGLL5&quot; &quot;GNLY&quot; ...
##   .. .. .. .. ..$ assay             : chr &quot;RNA&quot;
##   .. .. .. .. ..$ vars.to.regress   : chr [1:2] &quot;batch&quot; &quot;percent.mt&quot;
##   .. .. .. .. ..$ model.use         : chr &quot;linear&quot;
##   .. .. .. .. ..$ use.umi           : logi FALSE
##   .. .. .. .. ..$ do.scale          : logi TRUE
##   .. .. .. .. ..$ do.center         : logi TRUE
##   .. .. .. .. ..$ scale.max         : num 10
##   .. .. .. .. ..$ block.size        : num 1000
##   .. .. .. .. ..$ min.cells.to.block: num 2655
##   .. .. .. .. ..$ verbose           : logi TRUE
##   .. ..$ RunPCA.RNA              :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;RunPCA.RNA&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:49:34&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;RunPCA(object = seurat, do.print = TRUE, pcs.print = 1:5, &quot; &quot;    genes.print = 5)&quot;
##   .. .. .. ..@ params     :List of 10
##   .. .. .. .. ..$ assay          : chr &quot;RNA&quot;
##   .. .. .. .. ..$ npcs           : num 50
##   .. .. .. .. ..$ rev.pca        : logi FALSE
##   .. .. .. .. ..$ weight.by.var  : logi TRUE
##   .. .. .. .. ..$ verbose        : logi TRUE
##   .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5
##   .. .. .. .. ..$ nfeatures.print: num 30
##   .. .. .. .. ..$ reduction.name : chr &quot;pca&quot;
##   .. .. .. .. ..$ reduction.key  : chr &quot;PC_&quot;
##   .. .. .. .. ..$ seed.use       : num 42
##   .. ..$ ProjectDim.RNA.pca      :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;ProjectDim.RNA.pca&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:49:35&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr &quot;ProjectDim(object = seurat, reduction = \&quot;pca\&quot;)&quot;
##   .. .. .. ..@ params     :List of 7
##   .. .. .. .. ..$ reduction      : chr &quot;pca&quot;
##   .. .. .. .. ..$ assay          : chr &quot;RNA&quot;
##   .. .. .. .. ..$ dims.print     : int [1:5] 1 2 3 4 5
##   .. .. .. .. ..$ nfeatures.print: num 20
##   .. .. .. .. ..$ overwrite      : logi FALSE
##   .. .. .. .. ..$ do.center      : logi FALSE
##   .. .. .. .. ..$ verbose        : logi TRUE
##   .. ..$ FindNeighbors.RNA.pca   :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;FindNeighbors.RNA.pca&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:49:36&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr &quot;FindNeighbors(object = seurat, dims = 1:10)&quot;
##   .. .. .. ..@ params     :List of 13
##   .. .. .. .. ..$ reduction   : chr &quot;pca&quot;
##   .. .. .. .. ..$ dims        : int [1:10] 1 2 3 4 5 6 7 8 9 10
##   .. .. .. .. ..$ assay       : chr &quot;RNA&quot;
##   .. .. .. .. ..$ k.param     : num 20
##   .. .. .. .. ..$ compute.SNN : logi TRUE
##   .. .. .. .. ..$ prune.SNN   : num 0.0667
##   .. .. .. .. ..$ nn.method   : chr &quot;rann&quot;
##   .. .. .. .. ..$ annoy.metric: chr &quot;euclidean&quot;
##   .. .. .. .. ..$ nn.eps      : num 0
##   .. .. .. .. ..$ verbose     : logi TRUE
##   .. .. .. .. ..$ force.recalc: logi FALSE
##   .. .. .. .. ..$ do.plot     : logi FALSE
##   .. .. .. .. ..$ graph.name  : chr [1:2] &quot;RNA_nn&quot; &quot;RNA_snn&quot;
##   .. ..$ RunTSNE                 :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;RunTSNE&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:49:42&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;RunTSNE(seurat, reduction.use = \&quot;pca\&quot;, dims.use = 1:10, &quot; &quot;    perplexity = 10)&quot;
##   .. .. .. ..@ params     :List of 9
##   .. .. .. .. ..$ reduction     : chr &quot;pca&quot;
##   .. .. .. .. ..$ cells         : chr [1:2655] &quot;AAACATACAACCAC&quot; &quot;AAACATTGAGCTAC&quot; &quot;AAACATTGATCAGC&quot; &quot;AAACCGTGCTTCCG&quot; ...
##   .. .. .. .. ..$ dims          : int [1:5] 1 2 3 4 5
##   .. .. .. .. ..$ seed.use      : num 1
##   .. .. .. .. ..$ tsne.method   : chr &quot;Rtsne&quot;
##   .. .. .. .. ..$ add.iter      : num 0
##   .. .. .. .. ..$ dim.embed     : num 2
##   .. .. .. .. ..$ reduction.name: chr &quot;tsne&quot;
##   .. .. .. .. ..$ reduction.key : chr &quot;tSNE_&quot;
##   .. ..$ FindClusters            :Formal class &#39;SeuratCommand&#39; [package &quot;Seurat&quot;] with 5 slots
##   .. .. .. ..@ name       : chr &quot;FindClusters&quot;
##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: &quot;2020-01-16 15:51:13&quot;
##   .. .. .. ..@ assay.used : chr &quot;RNA&quot;
##   .. .. .. ..@ call.string: chr [1:2] &quot;FindClusters(object = seurat, reduction = \&quot;pca\&quot;, dims = 1:10, &quot; &quot;    resolution = 0.8)&quot;
##   .. .. .. ..@ params     :List of 10
##   .. .. .. .. ..$ graph.name      : chr &quot;RNA_snn&quot;
##   .. .. .. .. ..$ modularity.fxn  : num 1
##   .. .. .. .. ..$ resolution      : num 0.8
##   .. .. .. .. ..$ method          : chr &quot;matrix&quot;
##   .. .. .. .. ..$ algorithm       : num 1
##   .. .. .. .. ..$ n.start         : num 10
##   .. .. .. .. ..$ n.iter          : num 10
##   .. .. .. .. ..$ random.seed     : num 0
##   .. .. .. .. ..$ group.singletons: logi TRUE
##   .. .. .. .. ..$ verbose         : logi TRUE
##   ..@ tools       : list()</code></pre>
<div id="probabilistic-lda" class="section level4">
<h4><span class="header-section-number">7.3.3.1</span> Probabilistic (LDA)</h4>
<p>Another type of clustering we can do is a fuzzy or probablistic clustering. This is where cells are not assigned to specifically only one cluster. They get assigned a score for how much the cells belong to each of the clusters (sometimes called topics). This can be helpful for when your dataset continuous processes and/or cellular states as opposed to distinct cell types.</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="feature-selection-and-cluster-analysis-1.html#cb198-1"></a><span class="kw">data</span>(<span class="st">&quot;MouseDeng2014.FitGoM&quot;</span>)</span>
<span id="cb198-2"><a href="feature-selection-and-cluster-analysis-1.html#cb198-2"></a><span class="kw">names</span>(MouseDeng2014.FitGoM)</span></code></pre></div>
<pre><code>## [1] &quot;clust_3&quot; &quot;clust_6&quot;</code></pre>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="feature-selection-and-cluster-analysis-1.html#cb200-1"></a>omega &lt;-<span class="st"> </span>MouseDeng2014.FitGoM<span class="op">$</span>clust_<span class="dv">6</span><span class="op">$</span>omega</span>
<span id="cb200-2"><a href="feature-selection-and-cluster-analysis-1.html#cb200-2"></a></span>
<span id="cb200-3"><a href="feature-selection-and-cluster-analysis-1.html#cb200-3"></a></span>
<span id="cb200-4"><a href="feature-selection-and-cluster-analysis-1.html#cb200-4"></a>annotation &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb200-5"><a href="feature-selection-and-cluster-analysis-1.html#cb200-5"></a>  <span class="dt">sample_id =</span> <span class="kw">paste0</span>(<span class="st">&quot;X&quot;</span>, <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">NROW</span>(omega))),</span>
<span id="cb200-6"><a href="feature-selection-and-cluster-analysis-1.html#cb200-6"></a>  <span class="dt">tissue_label =</span> <span class="kw">factor</span>(<span class="kw">rownames</span>(omega),</span>
<span id="cb200-7"><a href="feature-selection-and-cluster-analysis-1.html#cb200-7"></a>                        <span class="dt">levels =</span> <span class="kw">rev</span>(<span class="kw">c</span>(<span class="st">&quot;zy&quot;</span>, <span class="st">&quot;early2cell&quot;</span>, <span class="st">&quot;mid2cell&quot;</span>, <span class="st">&quot;late2cell&quot;</span>,</span>
<span id="cb200-8"><a href="feature-selection-and-cluster-analysis-1.html#cb200-8"></a>                                        <span class="st">&quot;4cell&quot;</span>, <span class="st">&quot;8cell&quot;</span>, <span class="st">&quot;16cell&quot;</span>, <span class="st">&quot;earlyblast&quot;</span>, <span class="st">&quot;midblast&quot;</span>, <span class="st">&quot;lateblast&quot;</span>)))</span>
<span id="cb200-9"><a href="feature-selection-and-cluster-analysis-1.html#cb200-9"></a>  )</span>
<span id="cb200-10"><a href="feature-selection-and-cluster-analysis-1.html#cb200-10"></a></span>
<span id="cb200-11"><a href="feature-selection-and-cluster-analysis-1.html#cb200-11"></a><span class="kw">rownames</span>(omega) &lt;-<span class="st"> </span>annotation<span class="op">$</span>sample_id;</span>
<span id="cb200-12"><a href="feature-selection-and-cluster-analysis-1.html#cb200-12"></a></span>
<span id="cb200-13"><a href="feature-selection-and-cluster-analysis-1.html#cb200-13"></a><span class="kw">StructureGGplot</span>(<span class="dt">omega =</span> omega,</span>
<span id="cb200-14"><a href="feature-selection-and-cluster-analysis-1.html#cb200-14"></a>                <span class="dt">annotation =</span> annotation,</span>
<span id="cb200-15"><a href="feature-selection-and-cluster-analysis-1.html#cb200-15"></a>                <span class="dt">palette =</span> RColorBrewer<span class="op">::</span><span class="kw">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Accent&quot;</span>),</span>
<span id="cb200-16"><a href="feature-selection-and-cluster-analysis-1.html#cb200-16"></a>                <span class="dt">yaxis_label =</span> <span class="st">&quot;Amplification batch&quot;</span>,</span>
<span id="cb200-17"><a href="feature-selection-and-cluster-analysis-1.html#cb200-17"></a>                <span class="dt">order_sample =</span> <span class="ot">TRUE</span>,</span>
<span id="cb200-18"><a href="feature-selection-and-cluster-analysis-1.html#cb200-18"></a>                <span class="dt">axis_tick =</span> <span class="kw">list</span>(<span class="dt">axis_ticks_length =</span> <span class="fl">.1</span>,</span>
<span id="cb200-19"><a href="feature-selection-and-cluster-analysis-1.html#cb200-19"></a>                                 <span class="dt">axis_ticks_lwd_y =</span> <span class="fl">.1</span>,</span>
<span id="cb200-20"><a href="feature-selection-and-cluster-analysis-1.html#cb200-20"></a>                                 <span class="dt">axis_ticks_lwd_x =</span> <span class="fl">.1</span>,</span>
<span id="cb200-21"><a href="feature-selection-and-cluster-analysis-1.html#cb200-21"></a>                                 <span class="dt">axis_label_size =</span> <span class="dv">7</span>,</span>
<span id="cb200-22"><a href="feature-selection-and-cluster-analysis-1.html#cb200-22"></a>                                 <span class="dt">axis_label_face =</span> <span class="st">&quot;bold&quot;</span>))</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/example_lda-1.png" width="672" /></p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="feature-selection-and-cluster-analysis-1.html#cb201-1"></a><span class="kw">set.seed</span>(<span class="dv">2020</span>)</span>
<span id="cb201-2"><a href="feature-selection-and-cluster-analysis-1.html#cb201-2"></a></span>
<span id="cb201-3"><a href="feature-selection-and-cluster-analysis-1.html#cb201-3"></a><span class="co">## Preprocessing Steps</span></span>
<span id="cb201-4"><a href="feature-selection-and-cluster-analysis-1.html#cb201-4"></a>pbmc_small &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> pbmc_small, <span class="dt">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="dt">scale.factor =</span> <span class="dv">10000</span>)</span>
<span id="cb201-5"><a href="feature-selection-and-cluster-analysis-1.html#cb201-5"></a>pbmc_small &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(<span class="dt">object =</span> pbmc_small)</span></code></pre></div>
<pre><code>## Warning in irlba(A = t(x = object), nv = npcs, ...): You&#39;re computing too large a percentage
## of total singular values, use a standard svd instead.</code></pre>
<pre><code>## Warning in irlba(A = t(x = object), nv = npcs, ...): did not converge--results might be
## invalid!; try increasing work or maxit</code></pre>
<pre><code>## PC_ 1 
## Positive:  SDPR, PF4, PPBP, TUBB1, CA2, TREML1, MYL9, PGRMC1, RUFY1, PARVB 
##     VDAC3, IGLL5, AKR1C3, CD1C, RP11-290F20.3, GNLY, S100A8, S100A9, HLA-DQA1, HLA-DPB1 
## Negative:  HLA-DPB1, HLA-DQA1, S100A9, S100A8, GNLY, RP11-290F20.3, CD1C, AKR1C3, IGLL5, VDAC3 
##     PARVB, RUFY1, PGRMC1, MYL9, TREML1, CA2, TUBB1, PPBP, PF4, SDPR 
## PC_ 2 
## Positive:  HLA-DPB1, HLA-DQA1, S100A8, S100A9, CD1C, RP11-290F20.3, PARVB, IGLL5, MYL9, SDPR 
##     PPBP, CA2, RUFY1, TREML1, PF4, TUBB1, PGRMC1, VDAC3, AKR1C3, GNLY 
## Negative:  GNLY, AKR1C3, VDAC3, PGRMC1, TUBB1, PF4, TREML1, RUFY1, CA2, PPBP 
##     SDPR, MYL9, IGLL5, PARVB, RP11-290F20.3, CD1C, S100A9, S100A8, HLA-DQA1, HLA-DPB1 
## PC_ 3 
## Positive:  S100A9, S100A8, RP11-290F20.3, AKR1C3, PARVB, GNLY, PPBP, PGRMC1, MYL9, TUBB1 
##     CA2, TREML1, SDPR, VDAC3, PF4, RUFY1, HLA-DPB1, IGLL5, CD1C, HLA-DQA1 
## Negative:  HLA-DQA1, CD1C, IGLL5, HLA-DPB1, RUFY1, PF4, VDAC3, SDPR, TREML1, CA2 
##     TUBB1, MYL9, PGRMC1, PPBP, GNLY, PARVB, AKR1C3, RP11-290F20.3, S100A8, S100A9 
## PC_ 4 
## Positive:  IGLL5, RP11-290F20.3, VDAC3, PPBP, TUBB1, TREML1, PF4, CA2, PARVB, MYL9 
##     SDPR, RUFY1, PGRMC1, S100A9, HLA-DQA1, HLA-DPB1, GNLY, S100A8, AKR1C3, CD1C 
## Negative:  CD1C, AKR1C3, S100A8, GNLY, HLA-DPB1, HLA-DQA1, S100A9, PGRMC1, RUFY1, SDPR 
##     MYL9, PARVB, CA2, PF4, TREML1, TUBB1, PPBP, VDAC3, RP11-290F20.3, IGLL5 
## PC_ 5 
## Positive:  MYL9, PARVB, IGLL5, TREML1, AKR1C3, PGRMC1, HLA-DPB1, S100A9, TUBB1, PF4 
##     SDPR, GNLY, PPBP, S100A8, CA2, HLA-DQA1, CD1C, RUFY1, RP11-290F20.3, VDAC3 
## Negative:  VDAC3, RP11-290F20.3, RUFY1, CD1C, HLA-DQA1, CA2, S100A8, PPBP, GNLY, SDPR 
##     PF4, TUBB1, S100A9, HLA-DPB1, PGRMC1, AKR1C3, TREML1, IGLL5, PARVB, MYL9</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="feature-selection-and-cluster-analysis-1.html#cb205-1"></a>pbmc_small &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> pbmc_small, </span>
<span id="cb205-2"><a href="feature-selection-and-cluster-analysis-1.html#cb205-2"></a>                           <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, </span>
<span id="cb205-3"><a href="feature-selection-and-cluster-analysis-1.html#cb205-3"></a>                           <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, </span>
<span id="cb205-4"><a href="feature-selection-and-cluster-analysis-1.html#cb205-4"></a>                           <span class="dt">resolution =</span> <span class="dv">1</span>, </span>
<span id="cb205-5"><a href="feature-selection-and-cluster-analysis-1.html#cb205-5"></a>                           <span class="dt">print.output =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>## Warning: The following arguments are not used: reduction, dims.use, print.output</code></pre>
<pre><code>## Suggested parameter: dims instead of dims.use; verbose instead of print.output</code></pre>
<pre><code>## Warning: The following arguments are not used: reduction, dims.use, print.output</code></pre>
<pre><code>## Suggested parameter: dims instead of dims.use; verbose instead of print.output</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 80
## Number of edges: 3124
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.1433
## Number of communities: 3
## Elapsed time: 0 seconds</code></pre>
<pre><code>## Warning: Adding a command log without an assay associated with it</code></pre>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="feature-selection-and-cluster-analysis-1.html#cb212-1"></a>pbmc_counts &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(pbmc_small<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>data)</span>
<span id="cb212-2"><a href="feature-selection-and-cluster-analysis-1.html#cb212-2"></a>pbmc_meta &lt;-<span class="st"> </span>pbmc_small<span class="op">@</span>meta.data</span>
<span id="cb212-3"><a href="feature-selection-and-cluster-analysis-1.html#cb212-3"></a>gene_names &lt;-<span class="st"> </span><span class="kw">rownames</span>(pbmc_counts)</span>
<span id="cb212-4"><a href="feature-selection-and-cluster-analysis-1.html#cb212-4"></a></span>
<span id="cb212-5"><a href="feature-selection-and-cluster-analysis-1.html#cb212-5"></a></span>
<span id="cb212-6"><a href="feature-selection-and-cluster-analysis-1.html#cb212-6"></a>pbmc_FitGoM &lt;-<span class="st"> </span><span class="kw">FitGoM</span>(<span class="kw">t</span>(pbmc_counts), <span class="dt">K=</span><span class="dv">4</span>)</span></code></pre></div>
<pre><code>## options not specified: switching to default BIC, other choice is BF for Bayes factor</code></pre>
<pre><code>## Fitting a Grade of Membership model
##             (Taddy M., AISTATS 2012, JMLR 22,
##             http://proceedings.mlr.press/v22/taddy12/taddy12.pdf)</code></pre>
<pre><code>## 
## Estimating on a 80 document collection.
## Fit and Bayes Factor Estimation for K = 4
## log posterior increase: 1006, 69.7, 71, done.
## log BF( 4 ) = 5090.41</code></pre>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="feature-selection-and-cluster-analysis-1.html#cb216-1"></a>omega &lt;-<span class="st"> </span><span class="kw">data.frame</span>(pbmc_FitGoM<span class="op">$</span>fit<span class="op">$</span>omega)</span>
<span id="cb216-2"><a href="feature-selection-and-cluster-analysis-1.html#cb216-2"></a></span>
<span id="cb216-3"><a href="feature-selection-and-cluster-analysis-1.html#cb216-3"></a>annotation &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample_id =</span> <span class="kw">rownames</span>(omega),</span>
<span id="cb216-4"><a href="feature-selection-and-cluster-analysis-1.html#cb216-4"></a>                         <span class="dt">tissue_label =</span> <span class="kw">paste0</span>(<span class="st">&quot;cluster&quot;</span>, pbmc_small<span class="op">@</span>active.ident))</span>
<span id="cb216-5"><a href="feature-selection-and-cluster-analysis-1.html#cb216-5"></a></span>
<span id="cb216-6"><a href="feature-selection-and-cluster-analysis-1.html#cb216-6"></a><span class="kw">colnames</span>(omega) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;topic&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)</span>
<span id="cb216-7"><a href="feature-selection-and-cluster-analysis-1.html#cb216-7"></a><span class="kw">rownames</span>(omega) &lt;-<span class="st"> </span>annotation<span class="op">$</span>sample_id;</span>
<span id="cb216-8"><a href="feature-selection-and-cluster-analysis-1.html#cb216-8"></a></span>
<span id="cb216-9"><a href="feature-selection-and-cluster-analysis-1.html#cb216-9"></a><span class="kw">StructureGGplot</span>(<span class="dt">omega =</span> omega,</span>
<span id="cb216-10"><a href="feature-selection-and-cluster-analysis-1.html#cb216-10"></a>                <span class="dt">annotation =</span> annotation,</span>
<span id="cb216-11"><a href="feature-selection-and-cluster-analysis-1.html#cb216-11"></a>                <span class="dt">palette =</span> RColorBrewer<span class="op">::</span><span class="kw">brewer.pal</span>(<span class="dv">4</span>, <span class="st">&quot;Dark2&quot;</span>),</span>
<span id="cb216-12"><a href="feature-selection-and-cluster-analysis-1.html#cb216-12"></a>                <span class="dt">yaxis_label =</span> <span class="st">&quot;Cells&quot;</span>,</span>
<span id="cb216-13"><a href="feature-selection-and-cluster-analysis-1.html#cb216-13"></a>                <span class="dt">order_sample =</span> <span class="ot">TRUE</span>,</span>
<span id="cb216-14"><a href="feature-selection-and-cluster-analysis-1.html#cb216-14"></a>                <span class="dt">axis_tick =</span> <span class="kw">list</span>(<span class="dt">axis_ticks_length =</span> <span class="fl">.1</span>,</span>
<span id="cb216-15"><a href="feature-selection-and-cluster-analysis-1.html#cb216-15"></a>                                 <span class="dt">axis_ticks_lwd_y =</span> <span class="fl">.1</span>,</span>
<span id="cb216-16"><a href="feature-selection-and-cluster-analysis-1.html#cb216-16"></a>                                 <span class="dt">axis_ticks_lwd_x =</span> <span class="fl">.1</span>,</span>
<span id="cb216-17"><a href="feature-selection-and-cluster-analysis-1.html#cb216-17"></a>                                 <span class="dt">axis_label_size =</span> <span class="dv">7</span>,</span>
<span id="cb216-18"><a href="feature-selection-and-cluster-analysis-1.html#cb216-18"></a>                                 <span class="dt">axis_label_face =</span> <span class="st">&quot;bold&quot;</span>))</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/pbmc_lda-1.png" width="672" /></p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="feature-selection-and-cluster-analysis-1.html#cb217-1"></a><span class="co"># ## Add Topic Scores to Meta Data Part of the Seurat Object</span></span>
<span id="cb217-2"><a href="feature-selection-and-cluster-analysis-1.html#cb217-2"></a>pbmc_small &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(pbmc_small, omega)</span></code></pre></div>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="feature-selection-and-cluster-analysis-1.html#cb218-1"></a>pbmc_small<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb218-2"><a href="feature-selection-and-cluster-analysis-1.html#cb218-2"></a><span class="st">  </span><span class="kw">group_by</span>(RNA_snn_res<span class="fl">.1</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb218-3"><a href="feature-selection-and-cluster-analysis-1.html#cb218-3"></a><span class="st">  </span><span class="kw">summarise</span>(<span class="dt">topic1 =</span> <span class="kw">mean</span>(topic1),</span>
<span id="cb218-4"><a href="feature-selection-and-cluster-analysis-1.html#cb218-4"></a>            <span class="dt">topic2 =</span> <span class="kw">mean</span>(topic2),</span>
<span id="cb218-5"><a href="feature-selection-and-cluster-analysis-1.html#cb218-5"></a>            <span class="dt">topic3 =</span> <span class="kw">mean</span>(topic3),</span>
<span id="cb218-6"><a href="feature-selection-and-cluster-analysis-1.html#cb218-6"></a>            <span class="dt">topic4 =</span> <span class="kw">mean</span>(topic4))</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   RNA_snn_res.1 topic1 topic2 topic3 topic4
##   &lt;fct&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
## 1 0              0.281 0.290  0.221  0.207 
## 2 1              0.913 0.0253 0.0170 0.0450
## 3 2              0.595 0.0663 0.194  0.145</code></pre>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="feature-selection-and-cluster-analysis-1.html#cb220-1"></a><span class="co">## ggplot object, you can add layers</span></span>
<span id="cb220-2"><a href="feature-selection-and-cluster-analysis-1.html#cb220-2"></a>p1 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(pbmc_small, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Resolution 1&quot;</span>) <span class="co">## return ggplot object</span></span>
<span id="cb220-3"><a href="feature-selection-and-cluster-analysis-1.html#cb220-3"></a>p1</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/tsne_plot_pbmc-1.png" width="672" /></p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="feature-selection-and-cluster-analysis-1.html#cb221-1"></a>p2 &lt;-<span class="st"> </span><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc_small, </span>
<span id="cb221-2"><a href="feature-selection-and-cluster-analysis-1.html#cb221-2"></a>                  <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;topic1&quot;</span>, <span class="st">&quot;topic2&quot;</span>, <span class="st">&quot;topic3&quot;</span>, <span class="st">&quot;topic4&quot;</span>), </span>
<span id="cb221-3"><a href="feature-selection-and-cluster-analysis-1.html#cb221-3"></a>                  <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;blue&quot;</span>), </span>
<span id="cb221-4"><a href="feature-selection-and-cluster-analysis-1.html#cb221-4"></a>                  <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>) <span class="co">## return ggplot object</span></span>
<span id="cb221-5"><a href="feature-selection-and-cluster-analysis-1.html#cb221-5"></a></span>
<span id="cb221-6"><a href="feature-selection-and-cluster-analysis-1.html#cb221-6"></a>p2</span></code></pre></div>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/topic_plot-1.png" width="672" /></p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="feature-selection-and-cluster-analysis-1.html#cb222-1"></a><span class="kw">CombinePlots</span>(<span class="kw">list</span>(p1, p2))</span></code></pre></div>
<pre><code>## Warning: Graphs cannot be vertically aligned unless the axis parameter is set. Placing graphs
## unaligned.</code></pre>
<p><img src="KrumlovSingleCellWorkshop2020_files/figure-html/combine_topic_plot-1.png" width="672" /></p>
</div>
<div id="extract-top-feature" class="section level4">
<h4><span class="header-section-number">7.3.3.2</span> Extract Top Feature</h4>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="feature-selection-and-cluster-analysis-1.html#cb224-1"></a>theta_mat &lt;-<span class="st"> </span>pbmc_FitGoM<span class="op">$</span>fit<span class="op">$</span>theta</span>
<span id="cb224-2"><a href="feature-selection-and-cluster-analysis-1.html#cb224-2"></a></span>
<span id="cb224-3"><a href="feature-selection-and-cluster-analysis-1.html#cb224-3"></a>top_features &lt;-<span class="st"> </span><span class="kw">ExtractTopFeatures</span>(theta_mat, </span>
<span id="cb224-4"><a href="feature-selection-and-cluster-analysis-1.html#cb224-4"></a>                                   <span class="dt">top_features=</span><span class="dv">100</span>,</span>
<span id="cb224-5"><a href="feature-selection-and-cluster-analysis-1.html#cb224-5"></a>                                   <span class="dt">method=</span><span class="st">&quot;poisson&quot;</span>, </span>
<span id="cb224-6"><a href="feature-selection-and-cluster-analysis-1.html#cb224-6"></a>                                   <span class="dt">options=</span><span class="st">&quot;min&quot;</span>)</span>
<span id="cb224-7"><a href="feature-selection-and-cluster-analysis-1.html#cb224-7"></a></span>
<span id="cb224-8"><a href="feature-selection-and-cluster-analysis-1.html#cb224-8"></a>gene_list &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, </span>
<span id="cb224-9"><a href="feature-selection-and-cluster-analysis-1.html#cb224-9"></a>                     <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">dim</span>(top_features<span class="op">$</span>indices)[<span class="dv">1</span>],</span>
<span id="cb224-10"><a href="feature-selection-and-cluster-analysis-1.html#cb224-10"></a>                            <span class="cf">function</span>(x) gene_names[top_features<span class="op">$</span>indices[x,]]))</span></code></pre></div>
<p>We tabulate the top <code>5</code> genes for these <code>4</code> topics</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="feature-selection-and-cluster-analysis-1.html#cb225-1"></a>out_table &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, <span class="cf">function</span>(i) <span class="kw">toString</span>(gene_list[i,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])))</span>
<span id="cb225-2"><a href="feature-selection-and-cluster-analysis-1.html#cb225-2"></a><span class="kw">rownames</span>(out_table) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Topic&quot;</span>, <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>))</span>
<span id="cb225-3"><a href="feature-selection-and-cluster-analysis-1.html#cb225-3"></a>out_table</span></code></pre></div>
<pre><code>##         [,1]                                  
## Topic 1 &quot;GRN, HLA-DMB, IFI30, LY86, PF4&quot;      
## Topic 2 &quot;GNLY, GZMB, PRF1, GZMH, CCL4&quot;        
## Topic 3 &quot;GZMK, HNRNPH1, SATB1, SIT1, CRBN&quot;    
## Topic 4 &quot;PIK3IP1, TNFAIP8, THYN1, SP100, CCR7&quot;</code></pre>
</div>
</div>
<div id="practice-visualizingembedding" class="section level3">
<h3><span class="header-section-number">7.3.4</span> Practice Visualizing/Embedding</h3>
<div id="tsne" class="section level4">
<h4><span class="header-section-number">7.3.4.1</span> tSNE</h4>
<p>Change the parameter settings for tSNE</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="feature-selection-and-cluster-analysis-1.html#cb227-1"></a><span class="kw">RunTSNE</span>()</span></code></pre></div>
</div>
<div id="umap" class="section level4">
<h4><span class="header-section-number">7.3.4.2</span> UMAP</h4>
<p>Change the parameter settings for UMAP</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="feature-selection-and-cluster-analysis-1.html#cb228-1"></a><span class="kw">RunUMAP</span>()</span></code></pre></div>
</div>
</div>
</div>
<div id="other-options-for-analysis" class="section level2">
<h2><span class="header-section-number">7.4</span> Other Options For Analysis</h2>
<ul>
<li><a href="https://satijalab.org/seurat/vignettes.html">More Seurat Vignettes</a></li>
<li><a href="https://broadinstitute.github.io/2019_scWorkshop/">Single Cell Analysis Workshop</a></li>
<li><a href="https://scrnaseq-course.cog.sanger.ac.uk/website/index.html">Hemberg Lab Course</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment</a></li>
<li><a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">Scanpy</a></li>
<li><a href="https://github.com/klarman-cell-observatory/pegasus/blob/master/notebooks/Pegasus%20Tutorial%20in%20Workshop.ipynb">Pegasus</a></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="identifying-cell-populations-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["KrumlovSingleCellWorkshop2020.pdf", "KrumlovSingleCellWorkshop2020.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
