<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/" />
  <meta property="og:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />
  
  <meta name="github-repo" content="broadinstitute/KrumlovSingleCellWorkship2020" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  
  <meta name="twitter:image" content="https://broadinstitute.github.io/KrumlovSingleCellWorkship2020/images/cover.png" />

<meta name="author" content="Kirk Gosik" />


<meta name="date" content="2020-01-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="intro.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#introduction-1"><i class="fa fa-check"></i><b>1.5.1</b> Introduction</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#transcriptome-quantification"><i class="fa fa-check"></i><b>1.5.2</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#expression-qc-and-normalization"><i class="fa fa-check"></i><b>1.5.3</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#data-wrangling-scrnaseq"><i class="fa fa-check"></i><b>1.5.4</b> Data Wrangling scRNAseq</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#identifying-cell-populations"><i class="fa fa-check"></i><b>1.5.5</b> Identifying Cell Populations</a></li>
<li class="chapter" data-level="1.5.6" data-path="index.html"><a href="index.html#feature-selection-and-cluster-analysis"><i class="fa fa-check"></i><b>1.5.6</b> Feature Selection and Cluster Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="transcriptome-quantification-1.html"><a href="transcriptome-quantification-1.html"><i class="fa fa-check"></i><b>3</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="4" data-path="expression-qc-and-normalization-1.html"><a href="expression-qc-and-normalization-1.html"><i class="fa fa-check"></i><b>4</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#goal"><i class="fa fa-check"></i><b>5.1</b> Goal</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#introduction-2"><i class="fa fa-check"></i><b>5.2</b> Introduction</a><ul>
<li class="chapter" data-level="5.2.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#load-necessary-packages"><i class="fa fa-check"></i><b>5.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="5.2.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>5.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="5.2.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>5.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="5.2.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>5.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="5.2.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>5.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>5.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="5.3.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>5.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="5.3.2" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>5.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#beginning-with-seurat-httpsatijalab.orgseurat"><i class="fa fa-check"></i><b>5.4</b> Beginning with Seurat: <span>http://satijalab.org/seurat/</span></a><ul>
<li class="chapter" data-level="5.4.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>5.4.1</b> Creating a seurat object</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>5.5</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>5.6</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="5.6.1" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>5.6.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>5.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling-scrnaseq-1.html"><a href="data-wrangling-scrnaseq-1.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>5.8</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html"><i class="fa fa-check"></i><b>6</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="6.1" data-path="identifying-cell-populations-1.html"><a href="identifying-cell-populations-1.html#google-slides"><i class="fa fa-check"></i><b>6.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html"><i class="fa fa-check"></i><b>7</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#abstract"><i class="fa fa-check"></i><b>7.1</b> Abstract</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#seurat-tutorial"><i class="fa fa-check"></i><b>7.2</b> Seurat Tutorial</a><ul>
<li class="chapter" data-level="7.2.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#preprocessing-steps"><i class="fa fa-check"></i><b>7.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="7.2.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>7.2.2</b> Start of Identifying Cell Types</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#feature-selection"><i class="fa fa-check"></i><b>7.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="7.3.1" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#differential-expression-analysis"><i class="fa fa-check"></i><b>7.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="7.3.2" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#dimensionality-reduction"><i class="fa fa-check"></i><b>7.3.2</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="7.3.3" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#independent-components-analysis-ica"><i class="fa fa-check"></i><b>7.3.3</b> Independent Components Analysis (ICA)</a></li>
<li class="chapter" data-level="7.3.4" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#clustering"><i class="fa fa-check"></i><b>7.3.4</b> Clustering</a></li>
<li class="chapter" data-level="7.3.5" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#check-clusters"><i class="fa fa-check"></i><b>7.3.5</b> Check Clusters</a></li>
<li class="chapter" data-level="7.3.6" data-path="feature-selection-and-cluster-analysis-1.html"><a href="feature-selection-and-cluster-analysis-1.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>7.3.6</b> Practice Visualizing/Embedding</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</h1>
<p class="author"><em>Kirk Gosik</em></p>
<p class="date"><em>01/16/2020</em></p>
</div>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<div id="course-overview" class="section level2">
<h2><span class="header-section-number">1.1</span> COURSE OVERVIEW</h2>
<p>In recent years single cell RNA-seq (scRNA-seq) has become widely used for transcriptome analysis in many areas of biology. In contrast to bulk RNA-seq, scRNA-seq provides quantitative measurements of the expression of every gene in a single cell. However, to analyze scRNA-seq data, novel methods are required and some of the underlying assumptions for the methods developed for bulk RNA-seq experiments are no longer valid. In this course we will cover all steps of the scRNA-seq processing, starting from the raw reads coming off the sequencer. The course includes common analysis strategies, using state-of-the-art methods and we also discuss the central biological questions that can be addressed using scRNA-seq.</p>
</div>
<div id="targeted-audience-assumed-background" class="section level2">
<h2><span class="header-section-number">1.2</span> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</h2>
<p>This course is aimed at researchers and technical workers who are or will be analyzing scRNA-seq data. The material is suitable both for experimentalists who want to learn more about data-analysis as well as computational biologists who want to learn about scRNASeq methods. Examples demonstrated in this course can be applied to any experimental protocol or biological system.</p>
<p>The requirements for this course are:
1. Working knowledge of unix (managing files, running programs)
2. Programming experience in R (writing a function, basic I/O operations, variable types, using packages). Bioconductor experience is a plus.
3. Familiarity with next-generation sequencing data and its analyses (using alignment and quantification tools for bulk sequencing data)</p>
</div>
<div id="course-format" class="section level2">
<h2><span class="header-section-number">1.3</span> COURSE FORMAT</h2>
<p>The course will be delivered over the course of five days. Each day will include a lecture and laboratory component. The lecture will introduce the topics of discussion and the laboratory sessions will be focused on practical hands-on analysis of scRNA-seq data. These sessions will involve a combination of both mirroring exercises with the instructor to demonstrate a skill as well as applying these skills on your own to complete individual exercises. After and during each exercise, interpretation of results will be discussed as a group. Computing will be done using a combination of tools installed on the attendees laptop computer and web resources accessed via web browser.</p>
</div>
<div id="getting-started" class="section level2">
<h2><span class="header-section-number">1.4</span> Getting Started</h2>
</div>
<div id="session-content" class="section level2">
<h2><span class="header-section-number">1.5</span> SESSION CONTENT</h2>
<div id="introduction-1" class="section level3">
<h3><span class="header-section-number">1.5.1</span> Introduction</h3>
<ul>
<li>Overview of course</li>
<li>General introduction: HCA/KCO overview</li>
<li>Comparison of Bulk and single cell RNA-Seq</li>
<li>Overview of available scRNA-seq technologies (10x) and experimental protocols</li>
<li>scRNA-Seq experimental design and analysis workflow?</li>
</ul>
</div>
<div id="transcriptome-quantification" class="section level3">
<h3><span class="header-section-number">1.5.2</span> Transcriptome Quantification</h3>
<ul>
<li>Read &amp; UMI counting (Kallisto alignment-free pseudocounts as well), how RSEM works (length dependence, sequencing depth, multimapping reads), CellRanger (dropest), bustools</li>
<li>10x barcode structure and links to Perturb-seq</li>
<li>Gene length &amp; coverage</li>
<li>Gene expression units (count data Smart-seq2 counts or 10x UMIs vs expression data)</li>
<li>Some R overview slides, <a href="https://r4ds.had.co.nz/" class="uri">https://r4ds.had.co.nz/</a></li>
</ul>
</div>
<div id="expression-qc-and-normalization" class="section level3">
<h3><span class="header-section-number">1.5.3</span> Expression QC and Normalization</h3>
<ul>
<li>What CellRanger does for quality filtering</li>
<li>PBMC data</li>
<li>Normalisation methods <a href="https://www.nature.com/articles/nmeth.4292" class="uri">https://www.nature.com/articles/nmeth.4292</a></li>
<li>Doublets, empty droplets</li>
<li>Barcode swapping</li>
<li>Regression with technical covariates</li>
<li>What about imputation?</li>
</ul>
</div>
<div id="data-wrangling-scrnaseq" class="section level3">
<h3><span class="header-section-number">1.5.4</span> Data Wrangling scRNAseq</h3>
<ul>
<li>Data structures and file formats for single-cell data</li>
<li>Quality control of cells and genes (doublets, ambient, empty drops)</li>
<li>Data exploration: violin plots…</li>
<li>Introducing Seurat object
<ul>
<li>Genes</li>
<li>House keeping genes</li>
<li>Mitochondrial genes (never used these ones)</li>
<li>Filter - Do we remove both cells and genes here?</li>
<li>Normalize (introduce more options, other than log transform?)</li>
<li>Find variable genes (Is it a first reduction? Why the binning?)</li>
<li>Scaling</li>
<li>Regression</li>
<li>Heatmap of desired genes?</li>
<li>Sigantures?</li>
</ul></li>
</ul>
</div>
<div id="identifying-cell-populations" class="section level3">
<h3><span class="header-section-number">1.5.5</span> Identifying Cell Populations</h3>
<ul>
<li>Feature selection</li>
<li>Dimensionality reduction</li>
<li>Clustering and assigning identity (Louvain, NMF, topic models, variational autoencoder)</li>
<li>Differential expression tests</li>
</ul>
</div>
<div id="feature-selection-and-cluster-analysis" class="section level3">
<h3><span class="header-section-number">1.5.6</span> Feature Selection and Cluster Analysis</h3>
<ul>
<li>Parameters and clustering</li>
<li>Comparison of feature selection methods</li>
</ul>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="intro.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["KrumlovSingleCellWorkshop2020.pdf", "KrumlovSingleCellWorkshop2020.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
